--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -167,7 +_,7 @@
     private boolean addBranchingBlocks(final BlockPos fromPos) {
         BlockState fromState = this.level.getBlockState(fromPos);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             if (direction.getAxis() != this.pushDirection.getAxis()) {
                 BlockPos neighbourPos = fromPos.relative(direction);
                 BlockState neighbourState = this.level.getBlockState(neighbourPos);
