--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -126,7 +_,7 @@
     }
 
     private boolean getNeighborSignal(final SignalGetter level, final BlockPos pos, final Direction pushDirection) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             if (direction != pushDirection && level.hasSignal(pos.relative(direction), direction)) {
                 return true;
             }
@@ -138,7 +_,7 @@
 
         BlockPos above = pos.above();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             if (direction != Direction.DOWN && level.hasSignal(above.relative(direction), direction)) {
                 return true;
             }
