--- a/net/minecraft/world/entity/ai/goal/BreathAirGoal.java
+++ b/net/minecraft/world/entity/ai/goal/BreathAirGoal.java
@@ -74,6 +_,6 @@
 
     private boolean givesAir(final LevelReader level, final BlockPos pos) {
         BlockState state = level.getBlockState(pos);
-        return (level.getFluidState(pos).isEmpty() || state.is(Blocks.BUBBLE_COLUMN)) && state.isPathfindable(PathComputationType.LAND);
+        return (state.getFluidState().isEmpty() || state.is(Blocks.BUBBLE_COLUMN)) && state.isPathfindable(PathComputationType.LAND); // Pluto - Decrease chunk/block lookups
     }
 }
