--- a/net/minecraft/world/entity/ai/goal/ClimbOnTopOfPowderSnowGoal.java
+++ b/net/minecraft/world/entity/ai/goal/ClimbOnTopOfPowderSnowGoal.java
@@ -23,9 +_,10 @@
     public boolean canUse() {
         boolean inPowderSnow = this.mob.wasInPowderSnow || this.mob.isInPowderSnow;
         if (inPowderSnow && this.mob.is(EntityTypeTags.POWDER_SNOW_WALKABLE_MOBS)) {
-            BlockPos above = this.mob.blockPosition().above();
-            BlockState aboveBlockState = this.level.getBlockState(above);
-            return aboveBlockState.is(Blocks.POWDER_SNOW) || aboveBlockState.getCollisionShape(this.level, above) == Shapes.empty();
+            // Pluto start - Memory allocation optimisations
+            BlockState aboveBlockState = this.level.pluto$getBlockState(this.mob.blockPosition().getX(), this.mob.blockPosition().getY() + 1, this.mob.blockPosition().getZ());
+            return aboveBlockState.is(Blocks.POWDER_SNOW) || aboveBlockState.getCollisionShape(this.level, this.mob.blockPosition().above()) == Shapes.empty();
+            // Pluto end - Memory allocation optimisations
         } else {
             return false;
         }
