--- a/net/minecraft/world/entity/ai/goal/DolphinJumpGoal.java
+++ b/net/minecraft/world/entity/ai/goal/DolphinJumpGoal.java
@@ -42,7 +_,10 @@
 
     private boolean waterIsClear(final BlockPos dolphinPos, final int stepX, final int stepZ, final int currentStep) {
         BlockPos nextPos = dolphinPos.offset(stepX * currentStep, 0, stepZ * currentStep);
-        return this.dolphin.level().getFluidState(nextPos).is(FluidTags.WATER) && !this.dolphin.level().getBlockState(nextPos).blocksMotion();
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.block.state.BlockState state = this.dolphin.level().getBlockState(nextPos);
+        return state.getFluidState().is(FluidTags.WATER) && !state.blocksMotion();
+        // Pluto end - Decrease chunk/block lookups
     }
 
     private boolean surfaceIsClear(final BlockPos dolphinPos, final int stepX, final int stepZ, final int currentStep) {
