--- a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
@@ -95,11 +_,9 @@
     public PathType getPathType(final PathfindingContext context, final int x, final int y, final int z) {
         PathType blockPathType = context.getPathTypeFromState(x, y, z);
         if (blockPathType == PathType.WATER) {
-            BlockPos.MutableBlockPos reusablePos = new BlockPos.MutableBlockPos();
 
-            for (Direction direction : Direction.values()) {
-                reusablePos.set(x, y, z).move(direction);
-                PathType pathType = context.getPathTypeFromState(reusablePos.getX(), reusablePos.getY(), reusablePos.getZ());
+            for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+                PathType pathType = context.getPathTypeFromState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ()); // Pluto - Memory allocation optimisations
                 if (pathType == PathType.BLOCKED) {
                     return PathType.WATER_BORDER;
                 }
