--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -51,7 +_,7 @@
         int count = 0;
         Map<Direction, Node> nodes = Maps.newEnumMap(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             Node node = this.findAcceptedNode(pos.x + direction.getStepX(), pos.y + direction.getStepY(), pos.z + direction.getStepZ());
             nodes.put(direction, node);
             if (this.isNodeValid(node)) {
@@ -59,7 +_,7 @@
             }
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Direction secondDirection = direction.getClockWise();
             if (hasMalus(nodes.get(direction)) && hasMalus(nodes.get(secondDirection))) {
                 Node diagonalNode = this.findAcceptedNode(
