--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -129,7 +_,7 @@
 
         double posHeight = this.getFloorLevel(new BlockPos(pos.x, pos.y, pos.z));
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Node node = this.findAcceptedNode(
                 pos.x + direction.getStepX(), pos.y, pos.z + direction.getStepZ(), jumpSize, posHeight, direction, blockPathTypeCurrent
             );
@@ -139,7 +_,7 @@
             }
         }
 
-        for (Direction directionx : Direction.Plane.HORIZONTAL) {
+        for (Direction directionx : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Direction secondDirection = directionx.getClockWise();
             if (this.isDiagonalValid(pos, this.reusableNeighbors[directionx.get2DDataValue()], this.reusableNeighbors[secondDirection.get2DDataValue()])) {
                 Node diagonalNode = this.findAcceptedNode(
