--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -100,6 +_,18 @@
     public boolean needsDecoration;
     // CraftBukkit end
 
+    // Pufferfish start - instead of using a random every time the chunk is ticked, define when lightning strikes preemptively
+    private int lightningTick;
+    // shouldDoLightning compiles down to 29 bytes, which with the default of 35 byte inlining should guarantee an inline
+    public final boolean shouldDoLightning(net.minecraft.util.RandomSource random) {
+        if (this.lightningTick-- <= 0) {
+            this.lightningTick = random.nextInt(this.level.spigotConfig.thunderChance) << 1;
+            return true;
+        }
+        return false;
+    }
+    // Pufferfish end
+
     // Paper start
     boolean loadedTicketLevel;
     // Paper end
@@ -162,6 +_,8 @@
             }
         }
 
+        this.lightningTick = new java.util.Random().nextInt(100000) << 1; // Pufferfish - initialize lightning tick
+
         this.postLoad = postLoad;
         this.blockTicks = blockTicks;
         this.fluidTicks = fluidTicks;
@@ -472,6 +_,48 @@
     public BlockEntity getBlockEntity(BlockPos pos) {
         return this.getBlockEntity(pos, LevelChunk.EntityCreationType.CHECK);
     }
+
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(int x, int y, int z) {
+        return getBlockStateFinal(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(long packedPos) {
+        return getBlockStateFinal(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(int x, int y, int z) {
+        return getBlockStateFinal(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(long packedPos) {
+        return getBlockStateFinal(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        return getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        return getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(int x, int y, int z) {
+        return getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(long packedPos) {
+        return getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+    // Pluto end - Memory allocation optimisations
 
     @Nullable
     public BlockEntity getBlockEntity(BlockPos pos, LevelChunk.EntityCreationType creationType) {
