--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -421,6 +_,12 @@
                         state.onPlace(this.level, pos, blockState, isMoving);
                     }
 
+                    // Paper start - Some blocks may break here
+                    if (!section.getBlockState(i, i1, i2).is(block)) {
+                        return null;
+                    }
+                    // Paper end - Some blocks may break here
+
                     if (state.hasBlockEntity()) {
                         BlockEntity blockEntity = this.getBlockEntity(pos, LevelChunk.EntityCreationType.CHECK);
                         if (blockEntity != null && !blockEntity.isValidBlockState(state)) {
