--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -84,6 +_,8 @@
         public String getType() {
             return "<null>";
         }
+
+        @Override public long getChunkPos() { return ChunkPos.ZERO.longKey(); } // Pluto - Optimise block entity ticking
     };
     private final Map<BlockPos, LevelChunk.RebindableTickingBlockEntityWrapper> tickersInLevel = Maps.newHashMap();
     private boolean loaded;
@@ -99,6 +_,18 @@
     public boolean needsDecoration;
     // CraftBukkit end
 
+    // Pufferfish start - instead of using a random every time the chunk is ticked, define when lightning strikes preemptively
+    private int lightningTick;
+    // shouldDoLightning compiles down to 29 bytes, which with the default of 35 byte inlining should guarantee an inline
+    public final boolean shouldDoLightning(net.minecraft.util.RandomSource random) {
+        if (this.lightningTick-- <= 0) {
+            this.lightningTick = random.nextInt(this.level.spigotConfig.thunderChance) << 1;
+            return true;
+        }
+        return false;
+    }
+    // Pufferfish end
+
     // Paper start
     boolean loadedTicketLevel;
     // Paper end
@@ -161,6 +_,8 @@
             }
         }
 
+        this.lightningTick = new java.util.Random().nextInt(100000) << 1; // Pufferfish - initialize lightning tick
+
         this.postLoad = postLoad;
         this.blockTicks = blockTicks;
         this.fluidTicks = fluidTicks;
@@ -466,6 +_,23 @@
         return this.getBlockEntity(pos, LevelChunk.EntityCreationType.CHECK);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return getBlockStateFinal(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        return getBlockStateFinal(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        return getFluidState(x, y, z);
+    }
+    // Pluto end - Memory allocation optimisations
+
     public @Nullable BlockEntity getBlockEntity(final BlockPos pos, final LevelChunk.EntityCreationType creationType) {
         // CraftBukkit start
         BlockEntity blockEntity = this.level.capturedBlockEntities.get(pos);
@@ -522,6 +_,30 @@
             );
     }
 
+    // Pluto start - Optimise block entity ticking
+    int pluto$currentTick = -1;
+    boolean pluto$wholeChunkIsTickable = false;
+    boolean pluto$isTicking(final BlockPos pos) {
+        if (!this.level.plutoConfig().useOptimisedBlockEntityTicking) return this.isTicking(pos);
+
+        if (pluto$currentTick == net.minecraft.server.MinecraftServer.currentTick)
+            return this.pluto$wholeChunkIsTickable || pluto$optimisedIsTicking(pos); // Fallback if the whole chunk isn't ticking.
+
+        this.pluto$currentTick = net.minecraft.server.MinecraftServer.currentTick;
+        this.pluto$wholeChunkIsTickable = this.getFullStatus().isOrAfter(FullChunkStatus.BLOCK_TICKING)
+            && this.level.getWorldBorder().isWithinBounds(pos)
+            && this.level.areEntitiesLoaded(this.chunkPos.longKey());
+        return this.pluto$wholeChunkIsTickable || pluto$optimisedIsTicking(pos); // Fallback if the whole chunk isn't ticking.
+    }
+
+    boolean pluto$optimisedIsTicking(final BlockPos pos) {
+        // Check the chunk status before the world border
+        return this.getFullStatus().isOrAfter(FullChunkStatus.BLOCK_TICKING)
+            && this.level.getWorldBorder().isWithinBounds(pos)
+            && this.level.areEntitiesLoaded(this.chunkPos.longKey());
+    }
+    // Pluto end - Optimise block entity ticking
+
     @Override
     public void setBlockEntity(final BlockEntity blockEntity) {
         BlockPos pos = blockEntity.getBlockPos();
@@ -932,17 +_,19 @@
         private final T blockEntity;
         private final BlockEntityTicker<T> ticker;
         private boolean loggedInvalidBlockState;
+        private final long chunkKey; @Override public long getChunkPos() { return chunkKey; } // Pluto - Optimise block entity ticking
 
         private BoundTickingBlockEntity(final T blockEntity, final BlockEntityTicker<T> ticker) {
             this.blockEntity = blockEntity;
             this.ticker = ticker;
+            this.chunkKey = ChunkPos.pack(blockEntity.getBlockPos()); // Pluto - Optimise block entity ticking
         }
 
         @Override
         public void tick() {
             if (!this.blockEntity.isRemoved() && this.blockEntity.hasLevel()) {
                 BlockPos pos = this.blockEntity.getBlockPos();
-                if (LevelChunk.this.isTicking(pos)) {
+                if (LevelChunk.this.pluto$isTicking(pos)) { // Pluto - Optimise block entity ticking
                     try {
                         ProfilerFiller profiler = Profiler.get();
                         profiler.push(this::getType);
@@ -1046,6 +_,8 @@
         public String toString() {
             return this.ticker + " <wrapped>";
         }
+
+        @Override public long getChunkPos() { return this.ticker.getChunkPos(); } // Pluto - Optimise block entity ticking
     }
 
     @FunctionalInterface
