--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -197,6 +_,27 @@
         return this.blockEntities.get(pos);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        return getBlockState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        if (this.isOutsideBuildHeight(y))
+            return Fluids.EMPTY.defaultFluidState();
+
+        LevelChunkSection section = this.getSection(this.getSectionIndex(y));
+        return section.hasOnlyAir() ? Fluids.EMPTY.defaultFluidState() : section.getFluidState(x & 15, y & 15, z & 15);
+    }
+    // Pluto end - Memory allocation optimisations
+
     public Map<BlockPos, BlockEntity> getBlockEntities() {
         return this.blockEntities;
     }
