--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -197,6 +_,52 @@
         return this.blockEntities.get(pos);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(int x, int y, int z) {
+        return getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(long packedPos) {
+        return getBlockState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(int x, int y, int z) {
+        return getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(long packedPos) {
+        return getBlockState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        if (this.isOutsideBuildHeight(y))
+            return Fluids.EMPTY.defaultFluidState();
+
+        LevelChunkSection section = this.getSection(this.getSectionIndex(y));
+        return section.hasOnlyAir() ? Fluids.EMPTY.defaultFluidState() : section.getFluidState(x & 15, y & 15, z & 15);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        return pluto$getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(int x, int y, int z) {
+        return pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(long packedPos) {
+        return pluto$getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+    // Pluto end - Memory allocation optimisations
+
     public Map<BlockPos, BlockEntity> getBlockEntities() {
         return this.blockEntities;
     }
