--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -224,10 +_,12 @@
         return 0.4F;
     }
 
+    private int behaviorTick = 0; // Pufferfish
     @Override
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("allayBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish
         this.getBrain().tick(level, this);
         profilerFiller.pop();
         profilerFiller.push("allayActivityUpdate");
@@ -396,6 +_,7 @@
     @Override
     public void updateDynamicGameEventListener(BiConsumer<DynamicGameEventListener<?>, ServerLevel> listenerConsumer) {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (serverLevel.plutoConfig().disableAllayGameEventListening) return; // Pluto - Allay Configuration
             listenerConsumer.accept(this.dynamicVibrationListener, serverLevel);
             listenerConsumer.accept(this.dynamicJukeboxListener, serverLevel);
         }
@@ -502,7 +_,7 @@
     }
 
     public boolean canDuplicate() {
-        return this.entityData.get(DATA_CAN_DUPLICATE);
+        return !this.level().plutoConfig().disableAllayDuplication && this.entityData.get(Allay.DATA_CAN_DUPLICATE); // Pluto - Allay configuration
     }
 
     private void removeInteractionItem(Player player, ItemStack stack) {
