--- a/net/minecraft/world/phys/Vec3.java
+++ b/net/minecraft/world/phys/Vec3.java
@@ -78,6 +_,12 @@
         this(vec.getX(), vec.getY(), vec.getZ());
     }
 
+    // Pluto start - Memory allocation optimisations
+    public long asBlockPosLong() {
+        return net.minecraft.core.BlockPos.asLong(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    // Pluto end - Memory allocation optimisations
+
     public Vec3 vectorTo(final Vec3 vec) {
         return new Vec3(vec.x - this.x, vec.y - this.y, vec.z - this.z);
     }
