--- a/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
@@ -29,7 +_,33 @@
 public class BlockStateListPopulator extends DummyLevelAccessor {
 
     private final LevelAccessor level;
-    private final Map<BlockPos, CapturedBlock> blocks = new LinkedHashMap<>();
+    // Pluto start - Memory allocation optimisations
+    private final it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<CapturedBlock> blocks = new it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<>();
+
+    @Override
+    public net.minecraft.world.level.block.state.BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        CapturedBlock block = this.blocks.get(BlockPos.asLong(x, y, z));
+        return (block != null) ? block.state() : this.level.pluto$getBlockState(x, y, z);
+    }
+
+    @Override
+    public net.minecraft.world.level.block.state.BlockState pluto$getBlockState(final long packedPos) {
+        CapturedBlock block = this.blocks.get(packedPos);
+        return (block != null) ? block.state() : this.level.pluto$getBlockState(packedPos);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        CapturedBlock block = this.blocks.get(BlockPos.asLong(x, y, z));
+        return (block != null) ? block.state().getFluidState() : this.level.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final long packedPos) {
+        CapturedBlock block = this.blocks.get(packedPos);
+        return (block != null) ? block.state().getFluidState() : this.level.pluto$getFluidState(packedPos);
+    }
+    // Pluto end - Memory allocation optimisations
 
     private List<CraftBlockState> snapshots;
 
@@ -39,19 +_,19 @@
 
     @Override
     public net.minecraft.world.level.block.state.BlockState getBlockState(BlockPos pos) {
-        CapturedBlock block = this.blocks.get(pos);
+        CapturedBlock block = this.blocks.get(pos.asLong()); // Pluto - Memory allocation optimisations
         return block != null ? block.state() : this.level.getBlockState(pos);
     }
 
     @Override
     public FluidState getFluidState(BlockPos pos) {
-        CapturedBlock block = this.blocks.get(pos);
+        CapturedBlock block = this.blocks.get(pos.asLong()); // Pluto - Memory allocation optimisations
         return block != null ? block.state().getFluidState() : this.level.getFluidState(pos);
     }
 
     @Override
     public BlockEntity getBlockEntity(BlockPos pos) {
-        CapturedBlock block = this.blocks.get(pos);
+        CapturedBlock block = this.blocks.get(pos.asLong()); // Pluto - Memory allocation optimisations
         return block != null ? block.blockEntity() : this.level.getBlockEntity(pos);
     }
 
@@ -59,7 +_,10 @@
     public boolean setBlock(BlockPos pos, net.minecraft.world.level.block.state.BlockState blockState, @Block.UpdateFlags int updateFlags, int updateLimit) {
         pos = pos.immutable();
         // remove first to keep last updated order
-        this.blocks.remove(pos);
+        // Pluto start - Memory allocation optimisations
+        final long posLong = pos.asLong();
+        this.blocks.remove(posLong);
+        // Pluto end - Memory allocation optimisations
 
         final BlockEntity newBlockEntity;
         if (blockState.getBlock() instanceof EntityBlock entityBlock) {
@@ -75,7 +_,7 @@
             newBlockEntity = null;
         }
 
-        this.blocks.put(pos, new CapturedBlock(blockState, updateFlags, newBlockEntity));
+        this.blocks.put(posLong, new CapturedBlock(blockState, updateFlags, newBlockEntity)); // Pluto - Memory allocation optimisations
         return true;
     }
 
@@ -96,10 +_,10 @@
     }
 
     private void iterateSnapshots(Consumer<CraftBlockState> callback) {
-        for (Map.Entry<BlockPos, CapturedBlock> entry : this.blocks.entrySet()) {
+        for (final it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap.Entry<CapturedBlock> entry : this.blocks.long2ObjectEntrySet()) { // Pluto - Memory allocation optimisations
             CapturedBlock block = entry.getValue();
             CraftBlockState snapshot = CraftBlockStates.getBlockState(
-                this.getMinecraftWorld().getWorld(), entry.getKey(), block.state(), block.blockEntity()
+                this.getMinecraftWorld().getWorld(), BlockPos.of(entry.getLongKey()), block.state(), block.blockEntity() // Pluto - Memory allocation optimisations
             );
             snapshot.setFlags(block.flags());
             snapshot.setWorldHandle(this.level);
