--- a/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/BlockStateListPopulator.java
@@ -24,36 +_,67 @@
 
 public class BlockStateListPopulator extends DummyGeneratorAccess {
     private final LevelAccessor world;
-    private final Map<BlockPos, net.minecraft.world.level.block.state.BlockState> dataMap = new HashMap<>();
-    private final Map<BlockPos, BlockEntity> entityMap = new HashMap<>();
-    private final LinkedHashMap<BlockPos, CraftBlockState> blocks;
+    // Pluto start - Memory allocation optimisations
+    private final Map<Long, net.minecraft.world.level.block.state.BlockState> dataMap = new HashMap<>();
+    private final Map<Long, BlockEntity> entityMap = new HashMap<>();
+    private final LinkedHashMap<Long, CraftBlockState> blocks;
+    // Pluto end - Memory allocation optimisations
 
     public BlockStateListPopulator(LevelAccessor world) {
         this(world, new LinkedHashMap<>());
     }
 
-    private BlockStateListPopulator(LevelAccessor world, LinkedHashMap<BlockPos, CraftBlockState> blocks) {
+    private BlockStateListPopulator(LevelAccessor world, LinkedHashMap<Long, CraftBlockState> blocks) { // Pluto - Memory allocation optimisations
         this.world = world;
         this.blocks = blocks;
     }
 
     @Override
     public net.minecraft.world.level.block.state.BlockState getBlockState(BlockPos pos) {
-        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(pos);
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(pos.asLong()); // Pluto - Memory allocation optimisations
         return (state != null) ? state : this.world.getBlockState(pos);
     }
 
     @Override
     public FluidState getFluidState(BlockPos pos) {
-        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(pos);
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(pos.asLong()); // Pluto - Memory allocation optimisations
         return (state != null) ? state.getFluidState() : this.world.getFluidState(pos);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public net.minecraft.world.level.block.state.BlockState pluto$getBlockState(int x, int y, int z) {
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(BlockPos.asLong(x, y, z));
+        return (state != null) ? state : this.world.pluto$getBlockState(x, y, z);
+    }
+
+    @Override
+    public net.minecraft.world.level.block.state.BlockState pluto$getBlockState(long packedPos) {
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(packedPos);
+        return (state != null) ? state : this.world.pluto$getBlockState(packedPos);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(BlockPos.asLong(x, y, z));
+        return (state != null) ? state.getFluidState() : this.world.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        net.minecraft.world.level.block.state.BlockState state = this.dataMap.get(packedPos);
+        return (state != null) ? state.getFluidState() : this.world.pluto$getFluidState(packedPos);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     public BlockEntity getBlockEntity(BlockPos pos) {
         // The contains is important to check for null values
-        if (this.entityMap.containsKey(pos)) {
-            return this.entityMap.get(pos);
+        // Pluto start - Memory allocation optimisations
+        long posLong = pos.asLong();
+        if (this.entityMap.containsKey(posLong)) {
+            return this.entityMap.get(posLong);
+            // Pluto end - Memory allocation optimisations
         }
 
         return this.world.getBlockEntity(pos);
@@ -63,13 +_,16 @@
     public boolean setBlock(BlockPos pos, net.minecraft.world.level.block.state.BlockState state, int flags, int recursionLeft) {
         pos = pos.immutable();
         // remove first to keep insertion order
-        this.blocks.remove(pos);
+        // Pluto start - Memory allocation optimisations
+        long posLong = pos.asLong();
+        this.blocks.remove(posLong);
 
-        this.dataMap.put(pos, state);
+        this.dataMap.put(posLong, state);
         if (state.hasBlockEntity()) {
-            this.entityMap.put(pos, ((EntityBlock) state.getBlock()).newBlockEntity(pos, state));
+            this.entityMap.put(posLong, ((EntityBlock) state.getBlock()).newBlockEntity(pos, state));
         } else {
-            this.entityMap.put(pos, null);
+            this.entityMap.put(posLong, null);
+            // Pluto end - Memory allocation optimisations
         }
 
         // use 'this' to ensure that the block state is the correct TileState
@@ -77,7 +_,7 @@
         snapshot.setFlags(flags);
         // set world handle to ensure that updated calls are done to the world and not to this populator
         snapshot.setWorldHandle(this.world);
-        this.blocks.put(pos, snapshot);
+        this.blocks.put(posLong, snapshot); // Pluto - Memory allocation optimisations
         return true;
     }
 
