--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
@@ -764,6 +_,48 @@
         return this.delegate.getFluidIfLoaded(pos);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public boolean pluto$isEmptyBlock(final long packedPos) {
+        return this.delegate.pluto$isEmptyBlock(packedPos);
+    }
+
+    @Override
+    public boolean pluto$isEmptyBlock(final int x, final int y, final int z) {
+        return this.delegate.pluto$isEmptyBlock(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return this.delegate.pluto$getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(final long packedPos) {
+        return this.delegate.pluto$getBlockState(packedPos);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        return this.delegate.pluto$getBlockStateIfLoaded(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final long packedPos) {
+        return this.delegate.pluto$getBlockStateIfLoaded(packedPos);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        return this.delegate.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final long packedPos) {
+        return this.delegate.pluto$getFluidState(packedPos);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Nullable
     @Override
     public ChunkAccess getChunkIfLoadedImmediately(final int x, final int z) {
