--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
@@ -764,6 +_,68 @@
         return this.delegate.getFluidIfLoaded(pos);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public int pluto$getDirectSignal(int x, int y, int z, Direction direction) {
+        return this.delegate.pluto$getDirectSignal(x, y, z, direction);
+    }
+
+    @Override
+    public int pluto$getControlInputSignal(int x, int y, int z, Direction direction, boolean diodesOnly) {
+        return this.delegate.pluto$getControlInputSignal(x, y, z, direction, diodesOnly);
+    }
+
+    @Override
+    public boolean pluto$isEmptyBlock(long packedPos) {
+        return this.delegate.pluto$isEmptyBlock(packedPos);
+    }
+
+    @Override
+    public boolean pluto$isEmptyBlock(int x, int y, int z) {
+        return this.delegate.pluto$isEmptyBlock(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(int x, int y, int z) {
+        return this.delegate.pluto$getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(long packedPos) {
+        return this.delegate.pluto$getBlockState(packedPos);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(int x, int y, int z) {
+        return this.delegate.pluto$getBlockStateIfLoaded(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(long packedPos) {
+        return this.delegate.pluto$getBlockStateIfLoaded(packedPos);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        return this.delegate.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        return this.delegate.pluto$getFluidState(packedPos);
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(int x, int y, int z) {
+        return this.delegate.pluto$getFluidStateIfLoaded(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidStateIfLoaded(long packedPos) {
+        return this.delegate.pluto$getFluidStateIfLoaded(packedPos);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Nullable
     @Override
     public ChunkAccess getChunkIfLoadedImmediately(final int x, final int z) {
