--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -40,6 +_,7 @@
 }
 
 dependencies {
+    api("io.sentry:sentry:5.4.0") // Pufferfish
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:33.3.1-jre")
     api("com.google.code.gson:gson:2.11.0")
@@ -90,7 +_,7 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("../paper-api/src/generated/java").asFile.toPath() // Pufferfish
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -100,6 +_,20 @@
     main {
         java {
             srcDir(generatedDir)
+            // Pufferfish start
+            srcDir(file("../paper-api/src/main/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/main/resources"))
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
+            // Pufferfish end
         }
     }
 }
@@ -147,6 +_,12 @@
     }
 }
 
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
         into("META-INF/maven/${project.group}/${project.name}")
@@ -166,7 +_,10 @@
 
 tasks.withType<Javadoc>().configureEach {
     val options = options as StandardJavadocDocletOptions
-    options.overview = "src/main/javadoc/overview.html"
+    // Pufferfish start
+    options.addStringOption("-add-modules", "jdk.incubator.vector")
+    options.overview = "../paper-api/src/main/javadoc/overview.html"
+    // Pufferfish end
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -199,11 +_,11 @@
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../paper-api/src/main/javadoc").withPropertyName("javadoc-sourceset") // Pufferfish
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../paper-api/src/main/javadoc") { // Pufferfish
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
