--- a/net/minecraft/world/entity/projectile/AbstractThrownPotion.java
+++ b/net/minecraft/world/entity/projectile/AbstractThrownPotion.java
@@ -60,7 +_,7 @@
                 this.dowseFire(blockPos1);
                 this.dowseFire(blockPos1.relative(direction.getOpposite()));
 
-                for (Direction direction1 : Direction.Plane.HORIZONTAL) {
+                for (Direction direction1 : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                     this.dowseFire(blockPos1.relative(direction1));
                 }
             }
