--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -346,7 +_,7 @@
                 double fishX = this.getX() + angleSin * this.timeUntilHooked * 0.1F;
                 double fishY = Mth.floor(this.getY()) + 1.0F;
                 double fishZ = this.getZ() + angleCos * this.timeUntilHooked * 0.1F;
-                BlockState splashBlockState = serverLevel.getBlockState(BlockPos.containing(fishX, fishY - 1.0, fishZ));
+                BlockState splashBlockState = serverLevel.pluto$getBlockState(fishX, fishY - 1.0, fishZ); // Pluto - Memory allocation optimisations
                 if (splashBlockState.is(Blocks.WATER)) {
                     if (this.random.nextFloat() < 0.15F) {
                         serverLevel.sendParticles(ParticleTypes.BUBBLE, fishX, fishY - 0.1F, fishZ, 1, angleSin, 0.1, angleCos, 0.0);
@@ -400,7 +_,7 @@
                 double fishX = this.getX() + Mth.sin(angle) * dist * 0.1;
                 double fishY = Mth.floor(this.getY()) + 1.0F;
                 double fishZ = this.getZ() + Mth.cos(angle) * dist * 0.1;
-                BlockState splashBlockState = serverLevel.getBlockState(BlockPos.containing(fishX, fishY - 1.0, fishZ));
+                BlockState splashBlockState = serverLevel.pluto$getBlockState(fishX, fishY - 1.0, fishZ); // Pluto - Memory allocation optimisations
                 if (splashBlockState.is(Blocks.WATER)) {
                     serverLevel.sendParticles(ParticleTypes.SPLASH, fishX, fishY, fishZ, 2 + this.random.nextInt(2), 0.1F, 0.0, 0.1F, 0.0);
                 }
