--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -819,6 +_,14 @@
             return;
         }
         // Paper end - entity despawn time limit
+
+        // Pluto start - TTL setting for entities from spawners
+        if (type != EntityType.PLAYER && spawnedViaMobSpawner && type.spawnerTTL >= 0 && this.tickCount >= type.spawnerTTL) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            return;
+        }
+        // Pluto end - TTL setting for entities from spawners
+
         this.baseTick();
     }
 
@@ -1276,6 +_,7 @@
             this.movementThisTick.add(new Entity.Movement(oldPosition, position));
             List<Entity.Movement> list = List.copyOf(this.movementThisTick);
             this.movementThisTick.clear();
+            if (this.level().plutoConfig().entityInsideBlockCheck > 0 && (this.level().plutoConfig().entityInsideBlockCheck == 1 || this.tickCount % this.level().plutoConfig().entityInsideBlockCheck == 0)) // Pluto - Configuration for entity inside block check frequency
             this.checkInsideBlocks(list, this.blocksInside);
             boolean flag = Iterables.any(this.blocksInside, state -> state.is(BlockTags.FIRE) || state.is(Blocks.LAVA));
             this.blocksInside.clear();
@@ -1343,7 +_,7 @@
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level().plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
@@ -1854,10 +_,26 @@
         return this.wasTouchingWater;
     }
 
+    private int lastRainCheck = -1; private boolean lastRainCheckResult = false; // Pluto - Optimise rain check
     public boolean isInRain() {
+        // Pluto start - Optimise rain check
+        if (this.tickCount - this.lastRainCheck < this.level().plutoConfig().entityRainCheckRate) {
+            return this.lastRainCheckResult;
+        }
+
+        this.lastRainCheck = this.tickCount;
+
+        // The nether & end can't have rain
+        // Both dimensions have skylight set to false.
+        if (!this.level().dimensionType().hasSkyLight()) {
+            this.lastRainCheckResult = false;
+            return false;
+        }
+
         BlockPos blockPos = this.blockPosition();
-        return this.level().isRainingAt(blockPos)
-            || this.level().isRainingAt(BlockPos.containing(blockPos.getX(), this.getBoundingBox().maxY, blockPos.getZ()));
+        this.lastRainCheckResult = this.level().isRainingAt(blockPos) || this.level().isRainingAt(BlockPos.containing((double) blockPos.getX(), this.getBoundingBox().maxY, (double) blockPos.getZ()));
+        return this.lastRainCheckResult;
+        // Pluto end - Optimise rain check
     }
 
     public boolean isInBubbleColumn() {
@@ -1898,6 +_,13 @@
         this.fluidHeight.clear();
         this.updateInWaterStateAndDoWaterCurrentPushing();
         double d = this.level().dimensionType().ultraWarm() ? 0.007 : 0.0023333333333333335;
+
+        // Pluto start - Optimise entity fluid pushing
+        if (level.plutoConfig().useFluidPushingOptimisation) {
+            return this.isInWater() || this.updateFluidHeightAndDoFluidPushing(FluidTags.LAVA, d);
+        }
+        // Pluto end - Optimise entity fluid pushing
+
         boolean flag = this.updateFluidHeightAndDoFluidPushing(FluidTags.LAVA, d);
         return this.isInWater() || flag;
     }
@@ -5074,6 +_,12 @@
             return this.save;
         }
     }
+
+    // Pluto start - Add method to obtain current entity id
+    public static int currentEntityId() {
+        return ENTITY_COUNTER.get();
+    }
+    // Pluto end - Add method to obtain current entity id
 
     // Paper start - Expose entity id counter
     public static int nextEntityId() {
