--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -696,6 +_,14 @@
             return;
         }
         // Paper end - entity despawn time limit
+
+        // Pluto start - TTL setting for entities from spawners
+        if (type != EntityType.PLAYER && spawnedViaMobSpawner && type.spawnerTTL >= 0 && this.tickCount >= type.spawnerTTL) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            return;
+        }
+        // Pluto end - TTL setting for entities from spawners
+
         this.baseTick();
     }
 
@@ -722,7 +_,7 @@
         }
 
         if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick
-        if (this.canSpawnSprintParticle()) {
+        if (!this.level.plutoConfig().disableSprintParticles && this.canSpawnSprintParticle()) { // Pluto - Option to disable sprint particles | More efficient to exit early to avoid canSpawnSprintParticle
             this.spawnSprintParticle();
         }
 
@@ -1179,6 +_,7 @@
 
             boolean isOnFire = this.isOnFire();
             boolean isFreezing = this.isFreezing();
+            if (this.level().plutoConfig().entityInsideBlockCheck > 0 && (this.level().plutoConfig().entityInsideBlockCheck == 1 || this.tickCount % this.level().plutoConfig().entityInsideBlockCheck == 0)) // Pluto - Configuration for entity inside block check frequency
             this.checkInsideBlocks(movements, this.insideEffectCollector);
             this.insideEffectCollector.applyAndClear(this);
             if (this.isInRain()) {
@@ -1255,7 +_,7 @@
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level().plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
@@ -1709,10 +_,25 @@
         return this.wasTouchingWater;
     }
 
+    private int lastRainCheck = -1; private boolean lastRainCheckResult = false; // Pluto - Optimise rain check
     public boolean isInRain() {
+        // Pluto start - Optimise rain check
+        if (this.tickCount - this.lastRainCheck < this.level().plutoConfig().entityRainCheckRate) {
+            return this.lastRainCheckResult;
+        }
+
+        this.lastRainCheck = this.tickCount;
+
+        // The nether & end can't have rain
+        // Both dimensions have skylight set to false.
+        if (!this.level().dimensionType().hasSkyLight()) {
+            this.lastRainCheckResult = false;
+            return false;
+        }
         BlockPos blockPos = this.blockPosition();
-        return this.level().isRainingAt(blockPos)
-            || this.level().isRainingAt(BlockPos.containing(blockPos.getX(), this.getBoundingBox().maxY, blockPos.getZ()));
+        this.lastRainCheckResult = this.level().isRainingAt(blockPos) || this.level().isRainingAt(BlockPos.containing(blockPos.getX(), this.getBoundingBox().maxY, blockPos.getZ()));
+        return this.lastRainCheckResult;
+        // Pluto end - Optimise rain check
     }
 
     public boolean isInWaterOrRain() {
@@ -1741,6 +_,13 @@
         this.fluidHeight.clear();
         this.updateInWaterStateAndDoWaterCurrentPushing();
         double d = this.level().dimensionType().ultraWarm() ? 0.007 : 0.0023333333333333335;
+
+        // Pluto start - Optimise entity fluid pushing
+        if (level.plutoConfig().useFluidPushingOptimisation) {
+            return this.isInWater() || this.updateFluidHeightAndDoFluidPushing(FluidTags.LAVA, d);
+        }
+        // Pluto end - Optimise entity fluid pushing
+
         boolean flag = this.updateFluidHeightAndDoFluidPushing(FluidTags.LAVA, d);
         return this.isInWater() || flag;
     }
@@ -4777,6 +_,43 @@
             return this.save;
         }
     }
+
+    // Pluto start - Add method to obtain current entity id
+    public static int currentEntityId() {
+        return ENTITY_COUNTER.get();
+    }
+    // Pluto end - Add method to obtain current entity id
+
+    // Pluto start - Entities search for hoppers
+    @Nullable
+    public net.minecraft.world.level.block.entity.HopperBlockEntity findHopper() {
+        int originX = this.getBlockX();
+        int originY = this.getBlockY();
+        int originZ = this.getBlockZ();
+
+        BlockPos.MutableBlockPos adjacentPos = new BlockPos.MutableBlockPos();
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(originX >> 4, originZ >> 4);
+        if (chunk == null) return null;
+
+        for (int y = originY; y > originY - 2; --y) {
+            adjacentPos.set(originX, y, originZ);
+            net.minecraft.world.level.block.entity.HopperBlockEntity hopper = chunk.getBlockEntity(adjacentPos) instanceof net.minecraft.world.level.block.entity.HopperBlockEntity hopperBlockEntity ? hopperBlockEntity : null;
+            if (hopper == null) continue;
+            return hopper;
+        }
+        return null;
+    }
+
+    public boolean foundHopper() {
+        if (!this.level().plutoConfig().entitiesSearchForHoppers || tickCount % this.level().plutoConfig().ticksBetweenEntitiesSearchForHoppers != 0) return false;
+        net.minecraft.world.level.block.entity.HopperBlockEntity hopper = findHopper();
+        return hopper != null && hopper.canAcceptItems() && acceptItem(hopper);
+    }
+
+    public boolean acceptItem(net.minecraft.world.level.block.entity.HopperBlockEntity hopper) {
+        return false;
+    }
+    // Pluto end - Entities search for hoppers
 
     // Paper start - Expose entity id counter
     public static int nextEntityId() {
