--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -75,6 +_,8 @@
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
     private final boolean trackDeltas;
+    public boolean dabEnabled = false; // Pufferfish
+    public int spawnerTTL = -1; // Pluto - TTL setting for entities from spawners
 
     public static Identifier getKey(final EntityType<?> type) {
         return BuiltInRegistries.ENTITY_TYPE.getKey(type);
