--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -180,7 +_,7 @@
             this.applyEffectsFromBlocks();
             float friction = 0.98F;
             if (this.onGround()) {
-                friction = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                friction = this.level().pluto$getBlockState(this.pluto$getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F; // Pluto - Memory allocation optimisations
             }
 
             this.setDeltaMovement(this.getDeltaMovement().scale(friction));
@@ -239,6 +_,13 @@
     public BlockPos getBlockPosBelowThatAffectsMyMovement() {
         return this.getOnPos(0.999999F);
     }
+
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public long pluto$getBlockPosBelowThatAffectsMyMovement() {
+        return this.pluto$getOnPos((float) DELTA_AFFECTED_BY_BLOCKS_BELOW_1_0);
+    }
+    // Pluto end - Memory allocation optimisations
 
     private void scanForMerges() {
         if (this.level() instanceof ServerLevel) {
