--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -154,7 +_,7 @@
             this.applyEffectsFromBlocks();
             float f = 0.98F;
             if (this.onGround()) {
-                f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                f = this.level().pluto$getBlockState(this.pluto$getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F; // Pluto - Memory efficient block lookups
             }
 
             this.setDeltaMovement(this.getDeltaMovement().scale(f));
@@ -211,8 +_,15 @@
 
     @Override
     public BlockPos getBlockPosBelowThatAffectsMyMovement() {
-        return this.getOnPos(0.999999F);
-    }
+        return this.getOnPos(0.999999F); // Pluto - Memory efficient block lookups | Diff on change
+    }
+
+    // Pluto start - Memory efficient block lookups
+    @Override
+    public long pluto$getBlockPosBelowThatAffectsMyMovement() {
+        return this.pluto$getOnPos(0.999999F);
+    }
+    // Pluto end - Memory efficient block lookups
 
     private void scanForMerges() {
         if (this.level() instanceof ServerLevel) {
