--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -434,7 +_,7 @@
         profiler.push("livingEntityBaseTick");
         if (this.isAlive() && this.level() instanceof ServerLevel level) {
             boolean isPlayer = this instanceof Player;
-            if (this.isInWall()) {
+            if (shouldCheckForSuffocation() && this.isInWall()) { // Pufferfish - optimize suffocation
                 this.hurtServer(level, this.damageSources().inWall(), 1.0F);
             } else if (isPlayer && !level.getWorldBorder().isWithinBounds(this.getBoundingBox())) {
                 double dist = level.getWorldBorder().getDistanceToBorder(this) + level.getWorldBorder().getSafeZone();
@@ -447,7 +_,7 @@
             }
 
             if (this.isEyeInFluid(FluidTags.WATER)
-                && !level.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
+                && !level.pluto$getBlockState(Mth.floor(this.getX()), Mth.floor(this.getEyeY()), Mth.floor(this.getZ())).is(Blocks.BUBBLE_COLUMN)) { // Pluto - Memory allocation optimisations
                 boolean canDrownInWater = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
                     && (!isPlayer || !((Player)this).getAbilities().invulnerable);
@@ -2948,6 +_,11 @@
         }
     }
 
+    // Pluto start - Expose some sound related functions
+    public float pluto$GetSoundVolume() { return getSoundVolume(); }
+    public float pluto$GetVoicePitch() { return getVoicePitch(); }
+    // Pluto end - Expose some sound related functions
+
     public float getSoundVolume() {
         return 1.0F;
     }
@@ -3103,15 +_,17 @@
     }
 
     private void travelInAir(final Vec3 input) {
-        BlockPos posBelow = this.getBlockPosBelowThatAffectsMyMovement();
-        float blockFriction = this.onGround() ? this.level().getBlockState(posBelow).getBlock().getFriction() : 1.0F;
+        // Pluto start - Memory allocation optimisations
+        long blockPosBelowThatAffectsMyMovement = this.pluto$getBlockPosBelowThatAffectsMyMovement();
+        float blockFriction = this.onGround() ? this.level().pluto$getBlockState(blockPosBelowThatAffectsMyMovement).getBlock().getFriction() : 1.0F;
+        // Pluto end - Memory allocation optimisations
         float friction = blockFriction * 0.91F;
         Vec3 movement = this.handleRelativeFrictionAndCalculateMovement(input, blockFriction);
         double movementY = movement.y;
         MobEffectInstance levitationEffect = this.getEffect(MobEffects.LEVITATION);
         if (levitationEffect != null) {
             movementY += (0.05 * (levitationEffect.getAmplifier() + 1) - movement.y) * 0.2;
-        } else if (!this.level().isClientSide() || this.level().hasChunkAt(posBelow)) {
+        } else if (!this.level().isClientSide() || this.level().hasChunkAt(BlockPos.getX(blockPosBelowThatAffectsMyMovement) >> 4, BlockPos.getZ(blockPosBelowThatAffectsMyMovement) >> 4)) { // Pluto - Memory allocation optimisations
             movementY -= this.getEffectiveGravity();
         } else if (this.getY() > this.level().getMinY()) {
             movementY = -0.1;
@@ -3795,6 +_,7 @@
             this.checkAutoSpinAttack(beforeTravelBox, this.getBoundingBox());
         }
 
+        if (this.level().plutoConfig().entityPushingModulo > 0 && (this.level().plutoConfig().entityPushingModulo == 1 || this.tickCount % this.level().plutoConfig().entityPushingModulo == 0)) // Pluto - Configuration for entity pushing frequency
         this.pushEntities();
         profiler.pop();
         // Paper start - Add EntityMoveEvent
@@ -3885,6 +_,7 @@
             return;
         }
         // Paper end - don't run getEntities if we're not going to use its result
+        if (this.level().plutoConfig().onlyPlayersPushEntities && !(this instanceof ServerPlayer)) return; // Pluto - Option to make only players push entities
         List<Entity> pushableEntities = this.level().getPushableEntities(this, this.getBoundingBox());
         if (!pushableEntities.isEmpty()) {
             if (this.level() instanceof ServerLevel serverLevel) {
@@ -4496,19 +_,27 @@
         double zo = this.getZ();
         double y = yy;
         boolean ok = false;
-        BlockPos pos = BlockPos.containing(xx, yy, zz);
         Level level = this.level();
-        if (level.hasChunkAt(pos)) {
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(Mth.floor(xo) >> 4, Mth.floor(zo) >> 4);
+        if (chunk != null) {
+            // Pluto start - Memory allocation optimisations
+            final int x2 = Mth.floor(xo);
+            int y2 = Mth.floor(y);
+            final int z2 = Mth.floor(zo);
+            // Pluto end - Memory allocation optimisations
+            // Pluto end - Decrease chunk/block lookups
             boolean landed = false;
 
-            while (!landed && pos.getY() > level.getMinY()) {
-                BlockPos below = pos.below();
-                BlockState state = level.getBlockState(below);
+            // Pluto start - Memory allocation optimisations
+            while (!landed && y2 > level.getMinY()) {
+                BlockState state = chunk.getBlockState(x2, y2 - 1, z2); // Pluto - Decrease chunk/block lookups
+                // Pluto end - Memory allocation optimisations
                 if (state.blocksMotion()) {
                     landed = true;
                 } else {
                     y--;
-                    pos = below;
+                    y2--; // Pluto - Memory allocation optimisations
                 }
             }
 
@@ -4650,8 +_,7 @@
                 Direction facing = state.getValue(BedBlock.FACING);
                 this.level().setBlock(bedPosition, state.setValue(BedBlock.OCCUPIED, false), Block.UPDATE_ALL);
                 Vec3 standUp = BedBlock.findStandUpPosition(this.getType(), this.level(), bedPosition, facing, this.getYRot()).orElseGet(() -> {
-                    BlockPos above = bedPosition.above();
-                    return new Vec3(above.getX() + 0.5, above.getY() + 0.1, above.getZ() + 0.5);
+                    return new Vec3(bedPosition.getX() + 0.5, bedPosition.getY() + 1.1, bedPosition.getZ() + 0.5); // Pluto - Memory allocation optimisations
                 });
                 Vec3 lookDirection = Vec3.atBottomCenterOf(bedPosition).subtract(standUp).normalize();
                 float yaw = (float)Mth.wrapDegrees(Mth.atan2(lookDirection.z, lookDirection.x) * 180.0F / (float)Math.PI - 90.0);
@@ -4675,6 +_,19 @@
     public boolean isInWall() {
         return !this.isSleeping() && super.isInWall();
     }
+
+    // Pufferfish start - optimize suffocation
+    public boolean couldPossiblyBeHurt(float amount) {
+        if ((float) this.invulnerableTime > (float) this.invulnerableDuration / 2.0F && amount <= this.lastHurt) {
+            return false;
+        }
+        return true;
+    }
+
+    public boolean shouldCheckForSuffocation() {
+        return !gg.pufferfish.pufferfish.PufferfishConfig.enableSuffocationOptimization || (tickCount % 10 == 0 && couldPossiblyBeHurt(1.0F));
+    }
+    // Pufferfish end
 
     public ItemStack getProjectile(final ItemStack heldWeapon) {
         return ItemStack.EMPTY;
