--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -215,15 +_,17 @@
         return this.lookControl;
     }
 
+    int _pufferfish_inactiveTickDisableCounter = 0; // Pufferfish - throttle inactive goal selector ticking
     // Paper start
     @Override
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
-        if (this.goalSelector.inactiveTick()) {
+        boolean isThrottled = gg.pufferfish.pufferfish.PufferfishConfig.throttleInactiveGoalSelectorTick && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, true) && !isThrottled) { // Pufferfish - pass activated priroity // Pufferfish - throttle inactive goal selector ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priority // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
         }
     }
@@ -580,9 +_,8 @@
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
             float br = this.getLightLevelDependentMagicValue();
-            BlockPos roundedPos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
             boolean isInNonBurnableBlock = this.isInWaterOrRain() || this.isInPowderSnow || this.wasInPowderSnow;
-            if (br > 0.5F && this.random.nextFloat() * 30.0F < (br - 0.4F) * 2.0F && !isInNonBurnableBlock && this.level().canSeeSky(roundedPos)) {
+            if (br > 0.5F && this.random.nextFloat() * 30.0F < (br - 0.4F) * 2.0F && !isInNonBurnableBlock && this.level().canSeeSky(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()))) { // Pluto - Memory allocation optimisations
                 return true;
             }
         }
@@ -772,12 +_,41 @@
         return this.isPassenger() || this.isLeashed();
     }
 
+    // Pluto start - Avoid looping over all players in a world per entity for despawns
+    private @Nullable Entity pluto$getNearestPlayer(final Entity entity, final double range, final @Nullable Predicate<Entity> predicate) {
+        // Fallback if this entity is somehow not from a server level or if the world has a small amount of players.
+        if (this.level().players().size() <= 10 || !(this.level() instanceof ServerLevel serverLevel)) return this.level().findNearbyPlayer(entity, range, predicate);
+
+        // Copied from EntityGetter#findNearbyPlayer(double, double, double, double, Predicate), but using moonrise's nearby players.
+        double x = entity.getX();
+        double y = entity.getY();
+        double z = entity.getZ();
+        double best = -1.0;
+        Player result = null;
+
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> players = serverLevel.moonrise$getNearbyPlayers().getPlayers(entity.blockPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+        if (players == null) return null;
+
+        for (final net.minecraft.server.level.ServerPlayer player : players) {
+            if (predicate == null || predicate.test(player)) {
+                double dist = player.distanceToSqr(x, y, z);
+                if ((range < 0.0 || dist < range * range) && (best == -1.0 || dist < best)) {
+                    best = dist;
+                    result = player;
+                }
+            }
+        }
+
+        return result;
+    }
+    // Pluto end - Avoid looping over all players in a world per entity for despawns
+
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && !this.getType().isAllowedInPeaceful()) {
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+            Entity player = this.level().plutoConfig().useFasterNearbyPlayerDespawnCheck ? pluto$getNearestPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING) : this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API // Pluto - Avoid looping over all players in a world per entity for despawns
             if (player != null) {
                 // Paper start - Configurable despawn distances
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
@@ -825,16 +_,20 @@
         int idBasedTickCount = this.tickCount + this.getId();
         if (idBasedTickCount % 2 != 0 && this.tickCount > 1) {
             profiler.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tickRunningGoals(false);
             profiler.pop();
             profiler.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tickRunningGoals(false);
             profiler.pop();
         } else {
             profiler.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
             profiler.pop();
             profiler.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
             profiler.pop();
         }
@@ -989,6 +_,8 @@
                 return player.getVehicle() == Mob.this || player.isWithinEntityInteractionRange(Mob.this, 4.0);
             }
 
+            @Override public boolean stillValid() { return !Mob.this.isRemoved(); } // Pluto - Hoppers cache containers
+
             // Paper start
             private final List<org.bukkit.entity.HumanEntity> viewers = new java.util.ArrayList<>();
             private int maxStackSize = MAX_STACK;
@@ -1025,6 +_,13 @@
                 }
                 return null;
             }
+
+            // Pluto start - Allow toggling snapshots for certain blocks
+            @Override
+            public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
+                return getOwner();
+            }
+            // Pluto end - Allow toggling snapshots for certain blocks
 
             @Override
             public void setMaxStackSize(final int size) {
