--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -217,14 +_,16 @@
         return this.lookControl;
     }
 
+    int _pufferfish_inactiveTickDisableCounter = 0; // Pufferfish - throttle inactive goal selector ticking
     // Paper start
     @Override
     public void inactiveTick() {
         super.inactiveTick();
-        if (this.goalSelector.inactiveTick()) {
+        boolean isThrottled = gg.pufferfish.pufferfish.PufferfishConfig.throttleInactiveGoalSelectorTick && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, true) && !isThrottled) { // Pufferfish - pass activated priroity // Pufferfish - throttle inactive goal selector ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priority // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
         }
     }
@@ -578,12 +_,11 @@
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
             float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue();
-            BlockPos blockPos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
             boolean flag = this.isInWater() || this.isInPowderSnow || this.wasInPowderSnow;
             if (lightLevelDependentMagicValue > 0.5F
                 && this.random.nextFloat() * 30.0F < (lightLevelDependentMagicValue - 0.4F) * 2.0F
                 && !flag
-                && this.level().canSeeSky(blockPos)) {
+                && this.level().canSeeSky(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()))) { // Pluto - Memory allocation optimisations
                 return true;
             }
         }
@@ -810,16 +_,20 @@
         int i = this.tickCount + this.getId();
         if (i % 2 != 0 && this.tickCount > 1) {
             profilerFiller.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tickRunningGoals(false);
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tickRunningGoals(false);
             profilerFiller.pop();
         } else {
             profilerFiller.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
             profilerFiller.pop();
         }
@@ -978,6 +_,8 @@
                 return player.getVehicle() == Mob.this || player.isWithinEntityInteractionRange(Mob.this, 4.0);
             }
 
+            @Override public boolean stillValid() { return !Mob.this.isRemoved(); } // Pluto - Hoppers cache containers
+
             // Paper start
             private final List<org.bukkit.entity.HumanEntity> viewers = new java.util.ArrayList<>();
             private int maxStackSize = MAX_STACK;
@@ -1014,6 +_,13 @@
                 }
                 return null;
             }
+
+            // Pluto start - Allow toggling snapshots for certain blocks
+            @Override
+            public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
+                return getOwner();
+            }
+            // Pluto end - Allow toggling snapshots for certain blocks
 
             @Override
             public void setMaxStackSize(final int size) {
