--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -217,14 +_,16 @@
         return this.lookControl;
     }
 
+    int _pufferfish_inactiveTickDisableCounter = 0; // Pufferfish - throttle inactive goal selector ticking
     // Paper start
     @Override
     public void inactiveTick() {
         super.inactiveTick();
-        if (this.goalSelector.inactiveTick()) {
+        boolean isThrottled = gg.pufferfish.pufferfish.PufferfishConfig.throttleInactiveGoalSelectorTick && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, true) && !isThrottled) { // Pufferfish - pass activated priroity // Pufferfish - throttle inactive goal selector ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priority // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
         }
     }
@@ -578,12 +_,11 @@
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
             float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue();
-            BlockPos blockPos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
             boolean flag = this.isInWaterOrRain() || this.isInPowderSnow || this.wasInPowderSnow;
             if (lightLevelDependentMagicValue > 0.5F
                 && this.random.nextFloat() * 30.0F < (lightLevelDependentMagicValue - 0.4F) * 2.0F
                 && !flag
-                && this.level().canSeeSky(blockPos)) {
+                && this.level().canSeeSky(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()))) { // Pluto - Memory allocation optimisations
                 return true;
             }
         }
@@ -757,12 +_,38 @@
     }
     // Paper end - allow changing despawnInPeaceful
 
+    // Pluto start - Avoid looping over all players in a world per entity for despawns
+    private @Nullable Entity pluto$getNearestPlayer(Entity entity, double maxDistance, @Nullable Predicate<Entity> predicate) {
+        // Fallback if this entity is somehow not from a server level or if the world has a small amount of players.
+        if (this.level().players().size() <= 10 || !(this.level() instanceof ServerLevel serverLevel)) return this.level().findNearbyPlayer(entity, maxDistance, predicate);
+
+        // Copied from EntityGetter#findNearbyPlayer(double, double, double, double, Predicate), but using moonrise's nearby players.
+        double x = entity.getX();
+        double y = entity.getY();
+        double z = entity.getZ();
+        double d = -1.0;
+        Player player = null;
+
+        for (final net.minecraft.server.level.ServerPlayer player1 : serverLevel.moonrise$getNearbyPlayers().getPlayers(entity.blockPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE)) {
+            if (predicate == null || predicate.test(player1)) {
+                double d1 = player1.distanceToSqr(x, y, z);
+                if ((maxDistance < 0.0 || d1 < maxDistance * maxDistance) && (d == -1.0 || d1 < d)) {
+                    d = d1;
+                    player = player1;
+                }
+            }
+        }
+
+        return player;
+    }
+    // Pluto end - Avoid looping over all players in a world per entity for despawns
+
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity nearestPlayer = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+            Entity nearestPlayer = this.level().plutoConfig().useFasterNearbyPlayerDespawnCheck ? pluto$getNearestPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING) : this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API // Pluto - Avoid looping over all players in a world per entity for despawns
             if (nearestPlayer != null) {
                 // Paper start - Configurable despawn distances
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
@@ -810,16 +_,20 @@
         int i = this.tickCount + this.getId();
         if (i % 2 != 0 && this.tickCount > 1) {
             profilerFiller.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tickRunningGoals(false);
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tickRunningGoals(false);
             profilerFiller.pop();
         } else {
             profilerFiller.push("targetSelector");
+            if (!this.level().plutoConfig().disableTargetSelector && this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.targetSelector.tick();
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
+            if (!this.level().plutoConfig().disableGoalSelector && this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking // Pluto - Option to disable entity target/goal selectors
             this.goalSelector.tick();
             profilerFiller.pop();
         }
@@ -978,6 +_,8 @@
                 return player.getVehicle() == Mob.this || player.isWithinEntityInteractionRange(Mob.this, 4.0);
             }
 
+            @Override public boolean stillValid() { return !Mob.this.isRemoved(); } // Pluto - Hoppers cache containers
+
             // Paper start
             private final List<org.bukkit.entity.HumanEntity> viewers = new java.util.ArrayList<>();
             private int maxStackSize = MAX_STACK;
@@ -1014,6 +_,13 @@
                 }
                 return null;
             }
+
+            // Pluto start - Allow toggling snapshots for certain blocks
+            @Override
+            public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
+                return getOwner();
+            }
+            // Pluto end - Allow toggling snapshots for certain blocks
 
             @Override
             public void setMaxStackSize(final int size) {
