--- a/net/minecraft/world/entity/SpawnPlacementTypes.java
+++ b/net/minecraft/world/entity/SpawnPlacementTypes.java
@@ -12,8 +_,10 @@
     SpawnPlacementType NO_RESTRICTIONS = (level, blockPos, type) -> true;
     SpawnPlacementType IN_WATER = (level, blockPos, type) -> {
         if (type != null && level.getWorldBorder().isWithinBounds(blockPos)) {
-            BlockPos above = blockPos.above();
-            return level.getFluidState(blockPos).is(FluidTags.WATER) && !level.getBlockState(above).isRedstoneConductor(level, above);
+            // Pluto start - Memory allocation optimisations
+            BlockPos above;
+            return level.getFluidState(blockPos).is(FluidTags.WATER) && !level.getBlockState(above = blockPos.above()).isRedstoneConductor(level, above);
+            // Pluto end - Memory allocation optimisations
         } else {
             return false;
         }
@@ -25,12 +_,11 @@
         @Override
         public boolean isSpawnPositionOk(final LevelReader level, final BlockPos blockPos, final @Nullable EntityType<?> type) {
             if (type != null && level.getWorldBorder().isWithinBounds(blockPos)) {
-                BlockPos above = blockPos.above();
                 BlockPos below = blockPos.below();
                 BlockState belowState = level.getBlockState(below);
                 return belowState.isValidSpawn(level, below, type)
                     && this.isValidEmptySpawnBlock(level, blockPos, type)
-                    && this.isValidEmptySpawnBlock(level, above, type);
+                    && this.isValidEmptySpawnBlock(level, blockPos.above(), type); // Pluto - Memory allocation optimisations
             } else {
                 return false;
             }
