--- a/net/minecraft/world/entity/animal/frog/FrogAi.java
+++ b/net/minecraft/world/entity/animal/frog/FrogAi.java
@@ -211,14 +_,16 @@
 
     private static <E extends Mob> boolean isAcceptableLandingSpot(final E body, final BlockPos targetPos) {
         Level level = body.level();
-        BlockPos below = targetPos.below();
-        if (level.getFluidState(targetPos).isEmpty() && level.getFluidState(below).isEmpty() && level.getFluidState(targetPos.above()).isEmpty()) {
-            BlockState bs = level.getBlockState(targetPos);
-            BlockState bsBelow = level.getBlockState(below);
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.ChunkAccess chunk = level.getChunk(targetPos);
+        BlockState bs = chunk.getBlockState(targetPos); // Moved up
+        BlockState bsBelow; // Moved up
+        if (bs.getFluidState().isEmpty() && (bsBelow = chunk.getBlockState(targetPos.getX(), targetPos.getY() - 1, targetPos.getZ())).getFluidState().isEmpty() && chunk.pluto$getFluidState(targetPos.getX(), targetPos.getY() + 1, targetPos.getZ()).isEmpty()) { // Pluto - Memory allocation optimisations
+            // Pluto end - Decrease chunk/block lookups
             if (!bs.is(BlockTags.FROG_PREFER_JUMP_TO) && !bsBelow.is(BlockTags.FROG_PREFER_JUMP_TO)) {
                 PathfindingContext context = new PathfindingContext(body.level(), body);
                 PathType pathType = WalkNodeEvaluator.getPathTypeStatic(context, targetPos.mutable());
-                PathType pathTypeBelow = WalkNodeEvaluator.getPathTypeStatic(context, below.mutable());
+                PathType pathTypeBelow = WalkNodeEvaluator.getPathTypeStatic(context, targetPos.below().mutable()); // Pluto - Memory allocation optimisations
                 return pathType == PathType.TRAPDOOR
                     || bs.isAir() && pathTypeBelow == PathType.TRAPDOOR
                     || LongJumpToRandomPos.defaultAcceptableLandingSpot(body, targetPos);
