--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -455,8 +_,10 @@
                     }
 
                     if (nextPos != null) {
-                        BlockPos next = BlockPos.containing(nextPos);
-                        if (!level.getFluidState(next).is(FluidTags.WATER) || !level.getBlockState(next).isPathfindable(PathComputationType.WATER)) {
+                        // Pluto start - Decrease chunk/block lookups
+                        net.minecraft.world.level.block.state.BlockState state = level.pluto$getBlockState(nextPos.asBlockPosLong()); // Pluto - Memory allocation optimisations
+                        if (!state.getFluidState().is(FluidTags.WATER) || !state.isPathfindable(PathComputationType.WATER)) {
+                            // Pluto end - Decrease chunk/block lookups
                             nextPos = DefaultRandomPos.getPosTowards(this.dolphin, 8, 5, treasurePos, (float) (Math.PI / 2));
                         }
                     }
