--- a/net/minecraft/world/level/block/AzaleaBlock.java
+++ b/net/minecraft/world/level/block/AzaleaBlock.java
@@ -45,7 +_,7 @@
             return false;
         } else {
             int minHeight = TreeGrower.AZALEA.getMinimumHeight(serverLevel).orElse(0);
-            return level.isInsideBuildHeight(pos.above(minHeight + 2)) && level.getFluidState(pos.above()).isEmpty();
+            return level.isInsideBuildHeight(pos.above(minHeight + 2)) && level.pluto$getFluidState(pos.getX(), pos. getY() + 1, pos.getZ()).isEmpty(); // Pluto - Memory allocation optimisations
         }
     }
 
