--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -81,11 +_,17 @@
 
     @Override
     public @Nullable BlockState getStateForPlacement(final BlockPlaceContext context) {
-        FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
+        // Pluto start - Decrease chunk/block lookups
+        final BlockPos clickedPos = context.getClickedPos();
+        net.minecraft.world.level.chunk.LevelChunk chunk = context.getLevel().getChunkIfLoaded(clickedPos);
+        if (chunk == null) return null;
+
+        FluidState fluidState = chunk.getFluidState(clickedPos);
+        // Pluto end - Decrease chunk/block lookups
         if (!fluidState.isEmpty()) {
             return null;
         } else {
-            BlockState belowState = context.getLevel().getBlockState(context.getClickedPos().below());
+            BlockState belowState = chunk.getBlockState(clickedPos.getX(), clickedPos.getY() - 1, clickedPos.getZ()); // Pluto - Decrease chunk/block lookups
             if (belowState.is(BlockTags.SUPPORTS_BAMBOO)) {
                 if (belowState.is(Blocks.BAMBOO_SAPLING)) {
                     return this.defaultBlockState().setValue(AGE, 0);
@@ -93,7 +_,7 @@
                     int age = belowState.getValue(AGE) > 0 ? 1 : 0;
                     return this.defaultBlockState().setValue(AGE, age);
                 } else {
-                    BlockState aboveState = context.getLevel().getBlockState(context.getClickedPos().above());
+                    BlockState aboveState = chunk.getBlockState(clickedPos.getX(), clickedPos.getY() + 1, clickedPos.getZ()); // Pluto - Decrease chunk/block lookups
                     return aboveState.is(Blocks.BAMBOO)
                         ? this.defaultBlockState().setValue(AGE, aboveState.getValue(AGE))
                         : Blocks.BAMBOO_SAPLING.defaultBlockState();
@@ -119,7 +_,7 @@
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (state.getValue(STAGE) == 0) {
-            if (random.nextFloat() < (level.spigotConfig.bambooModifier / (100.0F * 3)) && level.isEmptyBlock(pos.above()) && level.getRawBrightness(pos.above(), 0) >= 9) { // Spigot - SPIGOT-7159: Better modifier resolution
+            if (random.nextFloat() < (level.spigotConfig.bambooModifier / (100.0F * 3)) && level.pluto$isEmptyBlock(pos.getX(), pos.getY() + 1, pos.getZ()) && (level.plutoConfig().bambooStalkIgnoreLightLevel || level.getRawBrightness(pos.above(), 0) >= 9)) { // Spigot - SPIGOT-7159: Better modifier resolution // Pluto - Toggle for light lookups on certain flora // Pluto - Memory allocation optimisations
                 int height = this.getHeightBelowUpToMax(level, pos) + 1;
                 if (height < level.paperConfig().maxGrowthHeight.bamboo.max) { // Paper - Configurable cactus/bamboo/reed growth height
                     this.growBamboo(state, level, pos, random, height);
@@ -130,7 +_,7 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
-        return level.getBlockState(pos.below()).is(BlockTags.SUPPORTS_BAMBOO);
+        return level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()).is(BlockTags.SUPPORTS_BAMBOO); // Pluto - Memory allocation optimisations
     }
 
     @Override
@@ -159,7 +_,7 @@
         int heightBelow = this.getHeightBelowUpToMax(level, pos);
         BlockPos growthPos = pos.above(heightAbove + 1);
         return heightAbove + heightBelow + 1 < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max // Paper - Configurable cactus/bamboo/reed growth height
-            && level.getBlockState(pos.above(heightAbove)).getValue(STAGE) != 1
+            && level.pluto$getBlockState(pos.getX(), pos.getY() + heightAbove, pos.getZ()).getValue(STAGE) != 1
             && level.isInsideBuildHeight(growthPos)
             && level.isEmptyBlock(growthPos);
     }
@@ -176,24 +_,32 @@
         int totalHeight = heightAbove + heightBelow + 1;
         int newBamboo = 1 + random.nextInt(2);
 
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
         for (int i = 0; i < newBamboo; i++) {
-            BlockPos topPos = pos.above(heightAbove);
-            BlockState topState = level.getBlockState(topPos);
-            BlockPos growthPos = topPos.above();
-            if (totalHeight >= level.paperConfig().maxGrowthHeight.bamboo.max || !topState.is(Blocks.BAMBOO) || topState.getValue(STAGE) == 1 || !level.isEmptyBlock(growthPos) || level.isOutsideBuildHeight(growthPos)) { // CraftBukkit - If the BlockSpreadEvent was cancelled, we have no bamboo here // Paper - Configurable cactus/bamboo/reed growth height
+            final int y1 = y + heightAbove;
+            BlockState topState = level.pluto$getBlockState(x, y1, z);
+            if (i >= level.paperConfig().maxGrowthHeight.bamboo.max || !topState.is(Blocks.BAMBOO) || topState.getValue(BambooStalkBlock.STAGE) == 1 || !level.pluto$isEmptyBlock(x, y1 + 1, z)) { // CraftBukkit - If the BlockSpreadEvent was cancelled, we have no bamboo here // Paper - Configurable cactus/bamboo/reed growth height
+                // Pluto end - Memory allocation optimisations
                 return;
             }
 
-            this.growBamboo(topState, level, topPos, random, totalHeight);
+            this.growBamboo(topState, level, new BlockPos(x, y1, z), random, i); // Pluto - Memory allocation optimisations
             heightAbove++;
             totalHeight++;
         }
     }
 
     protected void growBamboo(final BlockState state, final Level level, final BlockPos pos, final RandomSource random, final int height) {
-        BlockState belowState = level.getBlockState(pos.below());
-        BlockPos twoBelowPos = pos.below(2);
-        BlockState twoBelowState = level.getBlockState(twoBelowPos);
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+
+        BlockState belowState = chunk.getBlockState(pos.getX(), pos.getY() - 1, pos.getZ());
+        BlockState twoBelowState = chunk.getBlockState(pos.getX(), pos.getY() - 2, pos.getZ());
+        // Pluto end - Decrease chunk/block lookups
         BambooLeaves leaves = BambooLeaves.NONE;
         boolean shouldUpdateOthers = false; // CraftBukkit
         if (height >= 1) {
@@ -218,7 +_,7 @@
         if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, pos.above(), this.defaultBlockState().setValue(AGE, age).setValue(LEAVES, leaves).setValue(STAGE, stage), Block.UPDATE_ALL)) {
             if (shouldUpdateOthers) {
                 level.setBlock(pos.below(), belowState.setValue(LEAVES, BambooLeaves.SMALL), Block.UPDATE_ALL);
-                level.setBlock(twoBelowPos, twoBelowState.setValue(LEAVES, BambooLeaves.NONE), Block.UPDATE_ALL);
+                level.setBlock(pos.below(2), twoBelowState.setValue(BambooStalkBlock.LEAVES, BambooLeaves.NONE), Block.UPDATE_ALL); // Pluto - Decrease chunk/block lookups
             }
         }
         // CraftBukkit end
@@ -227,7 +_,12 @@
     protected int getHeightAboveUpToMax(final BlockGetter level, final BlockPos pos) {
         int height = 0;
 
-        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.getBlockState(pos.above(height + 1)).is(Blocks.BAMBOO)) { // Paper - Configurable cactus/bamboo/reed growth height
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.pluto$getBlockState(x, y + (height + 1), z).is(Blocks.BAMBOO)) { // Paper - Configurable cactus/bamboo/reed growth height
+            // Pluto end - Memory allocation optimisations
             height++;
         }
 
@@ -237,7 +_,12 @@
     protected int getHeightBelowUpToMax(final BlockGetter level, final BlockPos pos) {
         int height = 0;
 
-        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.getBlockState(pos.below(height + 1)).is(Blocks.BAMBOO)) { // Paper - Configurable cactus/bamboo/reed growth height
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.pluto$getBlockState(x, y - (height + 1), z).is(Blocks.BAMBOO)) { // Paper - Configurable cactus/bamboo/reed growth height
+            // Pluto end - Memory allocation optimisations
             height++;
         }
 
