--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -155,6 +_,18 @@
         return direction == Direction.UP ? this.getSignalForState(state) : Redstone.SIGNAL_MIN;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction direction) {
+        return this.getSignalForState(blockState);
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction side) {
+        return side == Direction.UP ? this.getSignalForState(blockState) : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected boolean isSignalSource(final BlockState state) {
         return true;
