--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -43,8 +_,7 @@
     }
 
     public static BlockState getState(BlockGetter level, BlockPos pos) {
-        BlockPos blockPos = pos.below();
-        BlockState blockState = level.getBlockState(blockPos);
+        BlockState blockState = level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()); // Pluto - Memory allocation optimisations
         return SoulFireBlock.canSurviveOnBlock(blockState) ? Blocks.SOUL_FIRE.defaultBlockState() : ((FireBlock)Blocks.FIRE).getStateForPlacement(level, pos);
     }
 
@@ -164,7 +_,7 @@
     @Override
     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston, net.minecraft.world.item.context.UseOnContext context) { // CraftBukkit - context
         if (!oldState.is(state.getBlock())) {
-            if (inPortalDimension(level)) {
+            if (!level.plutoConfig().disableNetherPortalGeneration && inPortalDimension(level)) { // Pluto - Toggle for nether portal generation
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X);
                 if (optional.isPresent()) {
                     optional.get().createPortalBlocks(level, (context == null) ? null : context.getPlayer()); // CraftBukkit - player
@@ -204,11 +_,15 @@
         if (!inPortalDimension(level)) {
             return false;
         } else {
-            BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
+            // Pluto start - Memory allocation optimisations
+            final int x = pos.getX();
+            final int y = pos.getY();
+            final int z = pos.getZ();
+            // Pluto end - Memory allocation optimisations
             boolean flag = false;
 
-            for (Direction direction1 : Direction.values()) {
-                if (level.getBlockState(mutableBlockPos.set(pos).move(direction1)).is(Blocks.OBSIDIAN)) {
+            for (Direction direction1 : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+                if (level.pluto$getBlockState(x + direction1.getStepX(), y + direction1.getStepY(), z + direction1.getStepZ()).is(Blocks.OBSIDIAN)) { // Pluto - Memory allocation optimisations
                     flag = true;
                     break;
                 }
