--- a/net/minecraft/world/level/block/BrushableBlock.java
+++ b/net/minecraft/world/level/block/BrushableBlock.java
@@ -85,7 +_,7 @@
             brushableBlockEntity.checkReset(level);
         }
 
-        if (FallingBlock.isFree(level.getBlockState(pos.below())) && pos.getY() >= level.getMinY()) {
+        if (FallingBlock.isFree(level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ())) && pos.getY() >= level.getMinY()) { // Pluto - Memory allocation optimisations
             FallingBlockEntity entity = FallingBlockEntity.fall(level, pos, state);
             entity.disableDrop();
         }
