--- a/net/minecraft/world/level/block/BubbleColumnBlock.java
+++ b/net/minecraft/world/level/block/BubbleColumnBlock.java
@@ -60,7 +_,7 @@
     ) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (isPrecise) {
-            BlockState stateAbove = level.getBlockState(pos.above());
+            BlockState stateAbove = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
             boolean nothingAbove = stateAbove.getCollisionShape(level, pos).isEmpty() && stateAbove.getFluidState().isEmpty();
             if (nothingAbove) {
                 entity.onAboveBubbleColumn(state.getValue(DRAG_DOWN), pos);
@@ -72,7 +_,7 @@
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        updateColumn(this, level, pos, state, level.getBlockState(pos.below()));
+        updateColumn(this, level, pos, state, level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ())); // Pluto - Memory allocation optimisations
     }
 
     @Override
@@ -188,7 +_,7 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
-        BlockState belowState = level.getBlockState(pos.below());
+        BlockState belowState = level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()); // Pluto - Memory allocation optimisations
         return belowState.is(this) || belowState.is(BlockTags.ENABLES_BUBBLE_COLUMN_PUSH_UP) || belowState.is(BlockTags.ENABLES_BUBBLE_COLUMN_DRAG_DOWN);
     }
 
