--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -27,8 +_,10 @@
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (random.nextInt(5) == 0) {
             Direction growDirection = DIRECTIONS[random.nextInt(DIRECTIONS.length)];
-            BlockPos growPos = pos.relative(growDirection);
-            BlockState relativeState = level.getBlockState(growPos);
+            // Pluto start - Memory allocation optimisations
+            BlockPos growPos;
+            BlockState relativeState = level.pluto$getBlockState(pos.getX() + growDirection.getStepX(), pos.getY() + growDirection.getStepY(), pos.getZ() + growDirection.getStepZ());
+            // Pluto end - Memory allocation optimisations
             Block nextStage = null;
             if (canClusterGrowAtState(relativeState)) {
                 nextStage = Blocks.SMALL_AMETHYST_BUD;
@@ -45,6 +_,7 @@
                     .setValue(AmethystClusterBlock.FACING, growDirection)
                     .setValue(AmethystClusterBlock.WATERLOGGED, relativeState.getFluidState().is(Fluids.WATER));
                 // Paper start - Have Amethyst throw both spread and grow events
+                growPos = pos.relative(growDirection); // Pluto - Memory allocation optimisations | Moved down
                 if (nextStage == Blocks.SMALL_AMETHYST_BUD) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, growPos, targetState, Block.UPDATE_ALL); // CraftBukkit
                 } else {
