--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -269,16 +_,24 @@
     }
 
     public static boolean isSmokeyPos(final Level level, final BlockPos pos) {
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return false;
+
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        // Pluto end - Decrease chunk/block lookups
+
         for (int i = 1; i <= 5; i++) {
-            BlockPos posToCheck = pos.below(i);
-            BlockState blockState = level.getBlockState(posToCheck);
+            BlockState blockState = chunk.getBlockState(x, y - i, z); // Pluto - Decrease chunk/block lookups
             if (isLitCampfire(blockState)) {
                 return true;
             }
 
             boolean smokeBlocked = Shapes.joinIsNotEmpty(SHAPE_VIRTUAL_POST, blockState.getCollisionShape(level, pos, CollisionContext.empty()), BooleanOp.AND);
             if (smokeBlocked) {
-                BlockState belowState = level.getBlockState(posToCheck.below());
+                BlockState belowState = chunk.getBlockState(x, y - (i + 1), z); // Pluto - Decrease chunk/block lookups
                 return isLitCampfire(belowState);
             }
         }
