--- a/net/minecraft/world/level/block/CandleCakeBlock.java
+++ b/net/minecraft/world/level/block/CandleCakeBlock.java
@@ -125,7 +_,7 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        return level.getBlockState(pos.below()).isSolid();
+        return level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()).isSolid(); // Pluto - Memory allocation optimisations
     }
 
     @Override
