--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -62,20 +_,25 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        BlockPos above = pos.above();
-        if (level.isEmptyBlock(above) && above.getY() <= level.getMaxY()) {
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockPos above;
+        if (level.pluto$isEmptyBlock(x, y + 1, z) && (y + 1) <= level.getMaxY()) {
+            // Pluto end - Memory allocation optimisations
             int currentAge = state.getValue(AGE);
             if (currentAge < 5) {
                 boolean growUpwards = false;
                 boolean pillarOnSupportBlock = false;
-                BlockState belowState = level.getBlockState(pos.below());
+                BlockState belowState = level.pluto$getBlockState(x, y - 1, z); // Pluto - Memory allocation optimisations
                 if (belowState.is(BlockTags.SUPPORTS_CHORUS_FLOWER)) {
                     growUpwards = true;
                 } else if (belowState.is(this.plant)) {
                     int height = 1;
 
                     for (int i = 0; i < 4; i++) {
-                        BlockState testState = level.getBlockState(pos.below(height + 1));
+                        BlockState testState = level.pluto$getBlockState(x, y - (height + 1), z); // Pluto - Memory allocation optimisations
                         if (!testState.is(this.plant)) {
                             if (testState.is(BlockTags.SUPPORTS_CHORUS_FLOWER)) {
                                 pillarOnSupportBlock = true;
@@ -93,8 +_,10 @@
                     growUpwards = true;
                 }
 
-                if (growUpwards && allNeighborsEmpty(level, above, null) && level.isEmptyBlock(pos.above(2))) {
-                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, above, this.defaultBlockState().setValue(AGE, currentAge), Block.UPDATE_CLIENTS)) { // CraftBukkit - add event
+                // Pluto start - Memory allocation optimisations
+                if (growUpwards && allNeighborsEmpty(level, x, y + 1, z, null) && level.pluto$isEmptyBlock(x, y + 2, z)) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, above = pos.above(), this.defaultBlockState().setValue(AGE, currentAge), Block.UPDATE_CLIENTS)) { // CraftBukkit - add event
+                    // Pluto end - Memory allocation optimisations
                     level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), Block.UPDATE_CLIENTS);
                     this.placeGrownFlower(level, above, currentAge);
                     } // CraftBukkit
@@ -108,9 +_,16 @@
 
                     for (int i = 0; i < numBranchAttempts; i++) {
                         Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(random);
-                        BlockPos target = pos.relative(direction);
-                        if (level.isEmptyBlock(target) && level.isEmptyBlock(target.below()) && allNeighborsEmpty(level, target, direction.getOpposite())) {
-                            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, target, this.defaultBlockState().setValue(AGE, currentAge + 1), Block.UPDATE_CLIENTS)) { // CraftBukkit - add event
+                        // Pluto start - Memory allocation optimisations
+                        final int x1 = x + direction.getStepX();
+                        final int y1 = y + direction.getStepY();
+                        final int z1 = z + direction.getStepZ();
+                        BlockPos target;
+                        if (level.pluto$isEmptyBlock(x1, y1, z1)
+                            && level.pluto$isEmptyBlock(x1, y1 - 1, z1)
+                            && allNeighborsEmpty(level, x1, y1, z1, direction.getOpposite())) {
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, target = pos.relative(direction), this.defaultBlockState().setValue(AGE, currentAge + 1), Block.UPDATE_CLIENTS)) { // CraftBukkit - add event
+                            // Pluto end - Memory allocation optimisations
                             this.placeGrownFlower(level, target, currentAge + 1);
                             createdBranch = true;
                             } // CraftBukkit
@@ -143,9 +_,11 @@
         level.levelEvent(LevelEvent.SOUND_CHORUS_DEATH, pos, 0);
     }
 
-    private static boolean allNeighborsEmpty(final LevelReader level, final BlockPos pos, final @Nullable Direction ignore) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
-            if (direction != ignore && !level.isEmptyBlock(pos.relative(direction))) {
+    // Pluto start - Memory allocation optimisations
+    private static boolean allNeighborsEmpty(LevelReader level, int x, int y, int z, @Nullable Direction excludingSide) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+            if (direction != excludingSide && !level.pluto$isEmptyBlock(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ())) {
+                // Pluto end - Memory allocation optimisations
                 return false;
             }
         }
@@ -173,7 +_,12 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
-        BlockState belowState = level.getBlockState(pos.below());
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockState belowState = level.pluto$getBlockState(x, y - 1, z);
+        // Pluto end - Memory allocation optimisations
         if (!belowState.is(this.plant) && !belowState.is(BlockTags.SUPPORTS_CHORUS_FLOWER)) {
             if (!belowState.isAir()) {
                 return false;
@@ -181,8 +_,8 @@
 
             boolean oneNeighbor = false;
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
-                BlockState neighbor = level.getBlockState(pos.relative(direction));
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+                BlockState neighbor = level.pluto$getBlockState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ()); // Pluto - Memory allocation optimisations
                 if (neighbor.is(this.plant)) {
                     if (oneNeighbor) {
                         return false;
@@ -219,12 +_,17 @@
             height++;
         }
 
+        // Pluto start - Memory allocation optimisations
+        final int x = current.getX();
+        final int y = current.getY();
+        final int z = current.getZ();
+        // Pluto end - Memory allocation optimisations
         for (int i = 0; i < height; i++) {
-            BlockPos target = current.above(i + 1);
-            if (!allNeighborsEmpty(level, target, null)) {
+            if (!allNeighborsEmpty(level, x, y + (i + 1), z, null)) { // Pluto - Memory allocation optimisations
                 return;
             }
 
+            BlockPos target = current.above(i + 1); // Pluto - Memory allocation optimisations | Moved down
             level.setBlock(target, ChorusPlantBlock.getStateWithConnections(level, target, chorus.defaultBlockState()), Block.UPDATE_CLIENTS);
             level.setBlock(target.below(), ChorusPlantBlock.getStateWithConnections(level, target.below(), chorus.defaultBlockState()), Block.UPDATE_CLIENTS);
         }
@@ -238,12 +_,17 @@
 
             for (int i = 0; i < stems; i++) {
                 Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(random);
-                BlockPos target = current.above(height).relative(direction);
-                if (Math.abs(target.getX() - startPos.getX()) < maxHorizontalSpread
-                    && Math.abs(target.getZ() - startPos.getZ()) < maxHorizontalSpread
-                    && level.isEmptyBlock(target)
-                    && level.isEmptyBlock(target.below())
-                    && allNeighborsEmpty(level, target, direction.getOpposite())) {
+                // Pluto start - Memory allocation optimisations
+                final int x1 = x + direction.getStepX();
+                final int y1 = (y + i) + direction.getStepY();
+                final int z1 = z + direction.getStepZ();
+                if (Math.abs(x1 - startPos.getX()) < maxHorizontalSpread
+                    && Math.abs(z1 - startPos.getZ()) < maxHorizontalSpread
+                    && level.pluto$isEmptyBlock(x1, y1, z1)
+                    && level.pluto$isEmptyBlock(x1, y1 - 1, z1)
+                    && allNeighborsEmpty(level, x1, y1, z1, direction.getOpposite())) {
+                    BlockPos target = new BlockPos(x1, y1, z1); // Moved down
+                    // Pluto end - Memory allocation optimisations
                     placedStem = true;
                     level.setBlock(target, ChorusPlantBlock.getStateWithConnections(level, target, chorus.defaultBlockState()), Block.UPDATE_CLIENTS);
                     level.setBlock(
