--- a/net/minecraft/world/level/block/ChorusPlantBlock.java
+++ b/net/minecraft/world/level/block/ChorusPlantBlock.java
@@ -44,12 +_,17 @@
     }
 
     public static BlockState getStateWithConnections(final BlockGetter level, final BlockPos pos, final BlockState defaultState) {
-        BlockState down = level.getBlockState(pos.below());
-        BlockState up = level.getBlockState(pos.above());
-        BlockState north = level.getBlockState(pos.north());
-        BlockState east = level.getBlockState(pos.east());
-        BlockState south = level.getBlockState(pos.south());
-        BlockState west = level.getBlockState(pos.west());
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockState down = level.pluto$getBlockState(x, y - 1, z);
+        BlockState up = level.pluto$getBlockState(x, y + 1, z);
+        BlockState north = level.pluto$getBlockState(x, y, z - 1);
+        BlockState east = level.pluto$getBlockState(x + 1, y, z);
+        BlockState south = level.pluto$getBlockState(x, y, z + 1);
+        BlockState west = level.pluto$getBlockState(x - 1, y, z);
+        // Pluto end - Memory allocation optimisations
         Block block = defaultState.getBlock();
         return defaultState.trySetValue(DOWN, down.is(block) || down.is(Blocks.CHORUS_FLOWER) || down.is(BlockTags.SUPPORTS_CHORUS_PLANT))
             .trySetValue(UP, up.is(block) || up.is(Blocks.CHORUS_FLOWER))
@@ -93,18 +_,27 @@
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
         if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableChorusPlantUpdates) return true; // Paper - add option to disable block updates
-        BlockState belowState = level.getBlockState(pos.below());
-        boolean blockAboveOrBelow = !level.getBlockState(pos.above()).isAir() && !belowState.isAir();
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockState belowState = level.pluto$getBlockState(x, y - 1, z);
+        boolean blockAboveOrBelow = !level.pluto$getBlockState(x, y + 1, z).isAir() && !belowState.isAir();
+        // Pluto end - Memory allocation optimisations
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
-            BlockPos neighborPos = pos.relative(direction);
-            BlockState neighborState = level.getBlockState(neighborPos);
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+            // Pluto start - Memory allocation optimisations
+            final int x1 = x + direction.getStepX();
+            final int y1 = y + direction.getStepY();
+            final int z1 = z + direction.getStepZ();
+            // Pluto end - Memory allocation optimisations
+            BlockState neighborState = level.pluto$getBlockState(x1, y1, z1); // Pluto - Memory allocation optimisations
             if (neighborState.is(this)) {
                 if (blockAboveOrBelow) {
                     return false;
                 }
 
-                BlockState below = level.getBlockState(neighborPos.below());
+                BlockState below = level.pluto$getBlockState(x1, y1 - 1, z1); // Pluto - Memory allocation optimisations
                 if (below.is(this) || below.is(BlockTags.SUPPORTS_CHORUS_PLANT)) {
                     return true;
                 }
