--- a/net/minecraft/world/level/block/ComparatorBlock.java
+++ b/net/minecraft/world/level/block/ComparatorBlock.java
@@ -64,6 +_,13 @@
             : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getOutputSignal(BlockGetter level, int x, int y, int z, BlockState state) {
+        return level.getBlockEntity(new BlockPos(x, y, z)) instanceof ComparatorBlockEntity comparatorBlockEntity ? comparatorBlockEntity.getOutputSignal() : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getOutputSignal(final BlockGetter level, final BlockPos pos, final BlockState state) {
         BlockEntity blockEntity = level.getBlockEntity(pos);
