--- a/net/minecraft/world/level/block/ComparatorBlock.java
+++ b/net/minecraft/world/level/block/ComparatorBlock.java
@@ -63,6 +_,13 @@
             : super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getOutputSignal(BlockGetter level, int x, int y, int z, BlockState state) {
+        return level.getBlockEntity(new BlockPos(x, y, z)) instanceof ComparatorBlockEntity comparatorBlockEntity ? comparatorBlockEntity.getOutputSignal() : 0;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getOutputSignal(BlockGetter level, BlockPos pos, BlockState state) {
         BlockEntity blockEntity = level.getBlockEntity(pos);
