--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -73,7 +_,7 @@
         boolean touchesLiquid = false;
         BlockPos.MutableBlockPos testPos = pos.mutable();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             BlockState blockState = level.getBlockState(testPos);
             if (direction != Direction.DOWN || canSolidify(blockState)) {
                 testPos.setWithOffset(pos, direction);
