--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -65,8 +_,13 @@
     }
 
     protected boolean scanForWater(final BlockGetter level, final BlockPos blockPos) {
-        for (Direction direction : Direction.values()) {
-            FluidState fluidState = level.getFluidState(blockPos.relative(direction));
+        // Pluto start - Memory allocation optimisations
+        final int x = blockPos.getX();
+        final int y = blockPos.getY();
+        final int z = blockPos.getZ();
+        // Pluto end - Memory allocation optimisations
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+            FluidState fluidState = level.pluto$getFluidState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ()); // Pluto - Memory allocation optimisations
             if (fluidState.is(FluidTags.WATER)) {
                 return true;
             }
