--- a/net/minecraft/world/level/block/CrafterBlock.java
+++ b/net/minecraft/world/level/block/CrafterBlock.java
@@ -201,10 +_,10 @@
             if (containerAt instanceof net.minecraft.world.CompoundContainer compoundContainer) {
                 destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
             } else {
-                destinationInventory = containerAt.getOwner().getInventory();
+                destinationInventory = containerAt.getOwner(level.plutoConfig().crafterDoSnapshotForTarget).getInventory(); // Pluto - Allow toggling snapshots for certain blocks
             }
 
-            org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(crafter.getOwner().getInventory(), oitemstack, destinationInventory, true);
+            org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(crafter.getOwner(level.plutoConfig().crafterDoSnapshotForSelf).getInventory(), oitemstack, destinationInventory, true); // Pluto - Allow toggling snapshots for certain blocks
             level.getCraftServer().getPluginManager().callEvent(event);
             itemStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
             while (!itemStack.isEmpty()) {
@@ -229,10 +_,10 @@
             if (containerAt instanceof net.minecraft.world.CompoundContainer compoundContainer) {
                 destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
             } else {
-                destinationInventory = containerAt.getOwner().getInventory();
+                destinationInventory = containerAt.getOwner(level.plutoConfig().crafterDoSnapshotForTarget).getInventory(); // Pluto - Allow toggling snapshots for certain blocks
             }
 
-            org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(crafter.getOwner().getInventory(), oitemstack, destinationInventory, true);
+            org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(crafter.getOwner(level.plutoConfig().crafterDoSnapshotForSelf).getInventory(), oitemstack, destinationInventory, true); // Pluto - Allow toggling snapshots for certain blocks
             level.getCraftServer().getPluginManager().callEvent(event);
             itemStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
             while (!itemStack.isEmpty()) {
