--- a/net/minecraft/world/level/block/CreakingHeartBlock.java
+++ b/net/minecraft/world/level/block/CreakingHeartBlock.java
@@ -116,8 +_,13 @@
     public static boolean hasRequiredLogs(final BlockState state, final LevelReader level, final BlockPos pos) {
         Direction.Axis axis = state.getValue(AXIS);
 
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
         for (Direction dir : axis.getDirections()) {
-            BlockState neigbour = level.getBlockState(pos.relative(dir));
+            BlockState neigbour = level.pluto$getBlockState(x + dir.getStepX(), y + dir.getStepY(), z + dir.getStepZ());
+            // Pluto end - Memory allocation optimisations
             if (!neigbour.is(BlockTags.PALE_OAK_LOGS) || neigbour.getValue(AXIS) != axis) {
                 return false;
             }
@@ -127,9 +_,13 @@
     }
 
     private static boolean isSurroundedByLogs(final LevelAccessor level, final BlockPos pos) {
-        for (Direction dir : Direction.values()) {
-            BlockPos neighbourPos = pos.relative(dir);
-            BlockState neighbourState = level.getBlockState(neighbourPos);
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        for (Direction dir : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+            BlockState neighbourState = level.pluto$getBlockState(x + dir.getStepX(), y + dir.getStepY(), z + dir.getStepZ());
+            // Pluto end - Memory allocation optimisations
             if (!neighbourState.is(BlockTags.PALE_OAK_LOGS)) {
                 return false;
             }
