--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -76,10 +_,10 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.getRawBrightness(pos, 0) >= 9) {
+        if (level.plutoConfig().cropIgnoreLightLevel || level.getRawBrightness(pos, 0) >= 9) { // Pluto - Toggle for light lookups on certain flora
             int age = this.getAge(state);
             if (age < this.getMaxAge()) {
-                float growthSpeed = getGrowthSpeed(this, level, pos);
+                float growthSpeed = level.plutoConfig().useStaticGrowthSpeedCrops ? level.plutoConfig().staticGrowthSpeedCrops : getGrowthSpeed(this, level, pos); // Pluto - Static growth speeds
                 // Spigot start
                 int modifier = 100;
                 if (this == Blocks.BEETROOTS) {
@@ -115,12 +_,17 @@
 
     protected static float getGrowthSpeed(Block block, BlockGetter level, BlockPos pos) {
         float f = 1.0F;
-        BlockPos blockPos = pos.below();
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int yBelow = y - 1;
+        final int z = pos.getZ();
+        // Pluto end - Memory allocation optimisations
 
         for (int i = -1; i <= 1; i++) {
             for (int i1 = -1; i1 <= 1; i1++) {
                 float f1 = 0.0F;
-                BlockState blockState = level.getBlockState(blockPos.offset(i, 0, i1));
+                BlockState blockState = level.pluto$getBlockState(x + i, yBelow, z + i1); // Pluto - Memory allocation optimisations
                 if (blockState.is(Blocks.FARMLAND)) {
                     f1 = 1.0F;
                     if (blockState.getValue(FarmBlock.MOISTURE) > 0) {
@@ -136,19 +_,23 @@
             }
         }
 
-        BlockPos blockPos1 = pos.north();
-        BlockPos blockPos2 = pos.south();
-        BlockPos blockPos3 = pos.west();
-        BlockPos blockPos4 = pos.east();
-        boolean flag = level.getBlockState(blockPos3).is(block) || level.getBlockState(blockPos4).is(block);
-        boolean flag1 = level.getBlockState(blockPos1).is(block) || level.getBlockState(blockPos2).is(block);
+        // Pluto start - Memory allocation optimisations
+        final int zNorth = z - 1;
+        final int zSouth = z + 1;
+        final int xWest = x - 1;
+        final int xEast = x + 1;
+        boolean flag = level.pluto$getBlockState(xWest, y, z).is(block) || level.pluto$getBlockState(xEast, y, z).is(block);
+        boolean flag1 = level.pluto$getBlockState(x, y, zNorth).is(block) || level.pluto$getBlockState(x, y, zSouth).is(block);
+        // Pluto end - Memory allocation optimisations
         if (flag && flag1) {
             f /= 2.0F;
         } else {
-            boolean flag2 = level.getBlockState(blockPos3.north()).is(block)
-                || level.getBlockState(blockPos4.north()).is(block)
-                || level.getBlockState(blockPos4.south()).is(block)
-                || level.getBlockState(blockPos3.south()).is(block);
+            // Pluto start - Memory allocation optimisations
+            boolean flag2 = level.pluto$getBlockState(xWest, y, z - 1).is(block)
+                || level.pluto$getBlockState(xEast, y, z - 1).is(block)
+                || level.pluto$getBlockState(xEast, y, z + 1).is(block)
+                || level.pluto$getBlockState(xWest, y, z + 1).is(block);
+            // Pluto end - Memory allocation optimisations
             if (flag2) {
                 f /= 2.0F;
             }
@@ -163,7 +_,7 @@
     }
 
     protected static boolean hasSufficientLight(LevelReader level, BlockPos pos) {
-        return level.getRawBrightness(pos, 0) >= 8;
+        return (level instanceof ServerLevel serverLevel && serverLevel.plutoConfig().cropIgnoreLightLevel) || level.getRawBrightness(pos, 0) >= 8; // Pluto - Toggle for light lookups on certain flora
     }
 
     @Override
