--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -77,10 +_,10 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        if (level.getRawBrightness(pos, 0) >= 9) {
+        if (level.plutoConfig().cropIgnoreLightLevel || level.getRawBrightness(pos, 0) >= 9) { // Pluto - Toggle for light lookups on certain flora
             int age = this.getAge(state);
             if (age < this.getMaxAge()) {
-                float growthSpeed = getGrowthSpeed(this, level, pos);
+                float growthSpeed = level.plutoConfig().useStaticGrowthSpeedCrops ? level.plutoConfig().staticGrowthSpeedCrops : getGrowthSpeed(this, level, pos); // Pluto - Static growth speeds
                 // Spigot start
                 int modifier = 100;
                 if (this == Blocks.BEETROOTS) {
@@ -116,12 +_,17 @@
 
     protected static float getGrowthSpeed(final Block type, final BlockGetter level, final BlockPos pos) {
         float speed = 1.0F;
-        BlockPos below = pos.below();
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int yBelow = y - 1;
+        final int z = pos.getZ();
+        // Pluto end - Memory allocation optimisations
 
         for (int xx = -1; xx <= 1; xx++) {
             for (int zz = -1; zz <= 1; zz++) {
                 float blockSpeed = 0.0F;
-                BlockState blockState = level.getBlockState(below.offset(xx, 0, zz));
+                BlockState blockState = level.pluto$getBlockState(x + xx, yBelow, z + zz); // Pluto - Memory allocation optimisations
                 if (blockState.is(BlockTags.GROWS_CROPS)) {
                     blockSpeed = 1.0F;
                     if (blockState.getValueOrElse(FarmlandBlock.MOISTURE, 0) > 0) {
@@ -137,19 +_,23 @@
             }
         }
 
-        BlockPos north = pos.north();
-        BlockPos south = pos.south();
-        BlockPos west = pos.west();
-        BlockPos east = pos.east();
-        boolean horizontal = level.getBlockState(west).is(type) || level.getBlockState(east).is(type);
-        boolean vertical = level.getBlockState(north).is(type) || level.getBlockState(south).is(type);
+        // Pluto start - Memory allocation optimisations
+        final int zNorth = z - 1;
+        final int zSouth = z + 1;
+        final int xWest = x - 1;
+        final int xEast = x + 1;
+        boolean horizontal = level.pluto$getBlockState(xWest, y, z).is(type) || level.pluto$getBlockState(xEast, y, z).is(type);
+        boolean vertical = level.pluto$getBlockState(x, y, zNorth).is(type) || level.pluto$getBlockState(x, y, zSouth).is(type);
+        // Pluto end - Memory allocation optimisations
         if (horizontal && vertical) {
             speed /= 2.0F;
         } else {
-            boolean diagonal = level.getBlockState(west.north()).is(type)
-                || level.getBlockState(east.north()).is(type)
-                || level.getBlockState(east.south()).is(type)
-                || level.getBlockState(west.south()).is(type);
+            // Pluto start - Memory allocation optimisations
+            boolean diagonal = level.pluto$getBlockState(xWest, y, z - 1).is(type)
+                || level.pluto$getBlockState(xEast, y, z - 1).is(type)
+                || level.pluto$getBlockState(xEast, y, z + 1).is(type)
+                || level.pluto$getBlockState(xWest, y, z + 1).is(type);
+            // Pluto end - Memory allocation optimisations
             if (diagonal) {
                 speed /= 2.0F;
             }
@@ -164,7 +_,7 @@
     }
 
     protected static boolean hasSufficientLight(final LevelReader level, final BlockPos pos) {
-        return level.getRawBrightness(pos, 0) >= 8;
+        return (level instanceof ServerLevel serverLevel && serverLevel.plutoConfig().cropIgnoreLightLevel) || level.getRawBrightness(pos, 0) >= 8; // Pluto - Toggle for light lookups on certain flora
     }
 
     @Override
