--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -86,6 +_,22 @@
         }
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction direction) {
+        return blockState.getValue(POWERED) ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction side) {
+        if (!blockState.getValue(POWERED)) {
+            return Redstone.SIGNAL_MIN;
+        } else {
+            return side == Direction.UP ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+        }
+    }
+    // Pluto end - Memory allocation optimisations
+
     private void checkPressed(final Level level, final BlockPos pos, final BlockState state) {
         if (this.canSurvive(state, level, pos)) {
             if (!state.is(this)) { return; } // Paper - Fix some rails connecting improperly
