--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -89,7 +_,7 @@
                 dropResources(state, level, pos, blockEntity);
                 level.removeBlock(pos, false);
 
-                for (Direction direction : Direction.values()) {
+                for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
                     level.updateNeighborsAt(pos.relative(direction), this);
                 }
             }
@@ -210,7 +_,7 @@
 
     public boolean shouldPrioritize(final BlockGetter level, final BlockPos pos, final BlockState state) {
         Direction direction = state.getValue(FACING).getOpposite();
-        BlockState oppositeState = level.getBlockState(pos.relative(direction));
+        BlockState oppositeState = level.pluto$getBlockState(pos.getX() + direction.getStepX(), pos.getY() + direction.getStepY(), pos.getZ() + direction.getStepZ()); // Pluto - Memory allocation optimisations
         return isDiode(oppositeState) && oppositeState.getValue(FACING) != direction;
     }
 
