--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -81,6 +_,26 @@
         return blockState.getSignal(blockAccess, pos, side);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        return blockState.pluto$getSignal(blockAccess, x, y, z, side);
+    }
+
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        if (!blockState.getValue(POWERED)) {
+            return 0;
+        } else {
+            return blockState.getValue(FACING) == side ? this.pluto$getOutputSignal(blockAccess, x, y, z, blockState) : 0;
+        }
+    }
+
+    protected int pluto$getOutputSignal(BlockGetter level, int x, int y, int z, BlockState state) {
+        return 15;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getSignal(BlockState blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
         if (!blockState.getValue(POWERED)) {
@@ -99,7 +_,7 @@
             dropResources(state, level, pos, blockEntity);
             level.removeBlock(pos, false);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
         }
@@ -148,8 +_,10 @@
         Direction counterClockWise = direction.getCounterClockWise();
         boolean flag = this.sideInputDiodesOnly();
         return Math.max(
-            level.getControlInputSignal(pos.relative(clockWise), clockWise, flag),
-            level.getControlInputSignal(pos.relative(counterClockWise), counterClockWise, flag)
+            // Pluto start - Memory allocation optimisations
+            level.pluto$getControlInputSignal(pos.getX() + clockWise.getStepX(), pos.getY() + clockWise.getStepY(), pos.getZ() + clockWise.getStepZ(), clockWise, flag),
+            level.pluto$getControlInputSignal(pos.getX() + counterClockWise.getStepX(), pos.getY() + counterClockWise.getStepY(), pos.getZ() + counterClockWise.getStepZ(), counterClockWise, flag)
+            // Pluto end - Memory allocation optimisations
         );
     }
 
@@ -204,7 +_,7 @@
 
     public boolean shouldPrioritize(BlockGetter level, BlockPos pos, BlockState state) {
         Direction opposite = state.getValue(FACING).getOpposite();
-        BlockState blockState = level.getBlockState(pos.relative(opposite));
+        BlockState blockState = level.pluto$getBlockState(pos.getX() + opposite.getStepX(), pos.getY() + opposite.getStepY(), pos.getZ() + opposite.getStepZ()); // Pluto - Memory allocation optimisations
         return isDiode(blockState) && blockState.getValue(FACING) != opposite;
     }
 
