--- a/net/minecraft/world/level/block/DirtPathBlock.java
+++ b/net/minecraft/world/level/block/DirtPathBlock.java
@@ -70,7 +_,7 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
-        BlockState aboveState = level.getBlockState(pos.above());
+        BlockState aboveState = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
         return !aboveState.isSolid() || aboveState.getBlock() instanceof FenceGateBlock;
     }
 
