--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -64,7 +_,7 @@
     public @Nullable BlockState getStateForPlacement(final BlockPlaceContext context) {
         BlockPos pos = context.getClickedPos();
         Level level = context.getLevel();
-        return pos.getY() < level.getMaxY() && level.getBlockState(pos.above()).canBeReplaced(context) ? super.getStateForPlacement(context) : null;
+        return pos.getY() < level.getMaxY() && level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()).canBeReplaced(context) ? super.getStateForPlacement(context) : null; // Pluto - Memory allocation optimisations
     }
 
     @Override
@@ -78,7 +_,7 @@
         if (state.getValue(HALF) != DoubleBlockHalf.UPPER) {
             return super.canSurvive(state, level, pos);
         } else {
-            BlockState belowState = level.getBlockState(pos.below());
+            BlockState belowState = level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()); // Pluto - Memory allocation optimisations
             return belowState.is(this) && belowState.getValue(HALF) == DoubleBlockHalf.LOWER;
         }
     }
@@ -127,10 +_,10 @@
         // CraftBukkit end
         DoubleBlockHalf part = state.getValue(HALF);
         if (part == DoubleBlockHalf.UPPER) {
-            BlockPos bottomPos = pos.below();
-            BlockState bottomState = level.getBlockState(bottomPos);
+            BlockState bottomState = level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()); // Pluto - Memory allocation optimisations
             if (bottomState.is(state.getBlock()) && bottomState.getValue(HALF) == DoubleBlockHalf.LOWER) {
                 BlockState blockState = bottomState.getFluidState().is(Fluids.WATER) ? Blocks.WATER.defaultBlockState() : Blocks.AIR.defaultBlockState();
+                BlockPos bottomPos = pos.below(); // Pluto - Memory allocation optimisations
                 level.setBlock(bottomPos, blockState, Block.UPDATE_ALL | Block.UPDATE_SUPPRESS_DROPS);
                 level.levelEvent(player, LevelEvent.PARTICLES_DESTROY_BLOCK, bottomPos, Block.getId(bottomState));
             }
