--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -65,6 +_,19 @@
                         if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(level, pos, itemStack, slot)) return; // Paper - Add BlockPreDispenseEvent
                         remaining = DISPENSE_BEHAVIOUR.dispense(source, itemStack);
                     } else {
+                        // Pluto start - Toggle inventory item move event for droppers
+                        if (level.plutoConfig().disableDropperInventoryMoveEvent) {
+                            remaining = HopperBlockEntity.addItem(blockEntity, into, itemStack.copyWithCount(1), direction.getOpposite());
+                            if (remaining.isEmpty()) {
+                                remaining = itemStack.copy();
+                                remaining.shrink(1);
+                            } else {
+                                remaining = itemStack.copy();
+                            }
+                            blockEntity.setItem(slot, remaining);
+                            return;
+                        }
+                        // Pluto end - Toggle inventory item move event for droppers
                         // CraftBukkit start - Fire event when pushing items into other inventories
                         org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack.copyWithCount(1));
 
@@ -73,10 +_,10 @@
                         if (into instanceof net.minecraft.world.CompoundContainer compoundContainer) {
                             destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
                         } else {
-                            destinationInventory = into.getOwner().getInventory();
+                            destinationInventory = into.getOwner(level.plutoConfig().dropperDoSnapshotForTarget).getInventory(); // Pluto - Allow toggling snapshots for certain blocks
                         }
 
-                        org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(blockEntity.getOwner().getInventory(), oitemstack, destinationInventory, true);
+                        org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(blockEntity.getOwner(level.plutoConfig().dropperDoSnapshotForSelf).getInventory(), oitemstack, destinationInventory, true); // Pluto - Allow toggling snapshots for certain blocks
                         if (!event.callEvent()) {
                             return;
                         }
