--- a/net/minecraft/world/level/block/FallingBlock.java
+++ b/net/minecraft/world/level/block/FallingBlock.java
@@ -47,7 +_,7 @@
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (isFree(level.getBlockState(pos.below())) && pos.getY() >= level.getMinY()) {
+        if (isFree(level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ())) && pos.getY() >= level.getMinY()) { // Pluto - Memory efficient block lookups
             FallingBlockEntity fallingBlockEntity = FallingBlockEntity.fall(level, pos, state);
             this.falling(fallingBlockEntity);
         }
