--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -104,6 +_,13 @@
             : Redstone.SIGNAL_MIN;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return level.getBlockEntity(new BlockPos(x, y, z)) instanceof JukeboxBlockEntity jukeboxBlockEntity && jukeboxBlockEntity.getSongPlayer().isPlaying() ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected boolean hasAnalogOutputSignal(final BlockState state) {
         return true;
