--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -118,11 +_,14 @@
 
     private static BlockState updateDistance(final BlockState state, final LevelAccessor level, final BlockPos pos) {
         int newDistance = 7;
-        BlockPos.MutableBlockPos neighborPos = new BlockPos.MutableBlockPos();
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
 
-        for (Direction direction : Direction.values()) {
-            neighborPos.setWithOffset(pos, direction);
-            newDistance = Math.min(newDistance, getDistanceAt(level.getBlockState(neighborPos)) + 1);
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+            newDistance = Math.min(newDistance, getDistanceAt(level.pluto$getBlockState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ())) + 1);
+            // Pluto end - Memory allocation optimisations
             if (newDistance == 1) {
                 break;
             }
