--- a/net/minecraft/world/level/block/LeverBlock.java
+++ b/net/minecraft/world/level/block/LeverBlock.java
@@ -142,6 +_,18 @@
         return blockState.getValue(POWERED) && getConnectedDirection(blockState) == side ? 15 : 0;
     }
 
+    // Pluto start - Memory efficient block lookups
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction direction) {
+        return blockState.getValue(POWERED) ? 15 : 0;
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        return blockState.getValue(POWERED) && getConnectedDirection(blockState) == side ? 15 : 0;
+    }
+    // Pluto end - Memory efficient block lookups
+
     @Override
     protected boolean isSignalSource(BlockState state) {
         return true;
