--- a/net/minecraft/world/level/block/MossyCarpetBlock.java
+++ b/net/minecraft/world/level/block/MossyCarpetBlock.java
@@ -129,7 +_,7 @@
         BlockState belowState = null;
         createSides |= state.getValue(BASE);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             EnumProperty<WallSide> property = getPropertyForFace(direction);
             WallSide side = canSupportAtFace(level, pos, direction) ? (createSides ? WallSide.LOW : state.getValue(property)) : WallSide.NONE;
             if (side == WallSide.LOW) {
@@ -194,7 +_,7 @@
             BlockState noCarpetBaseState = Blocks.PALE_MOSS_CARPET.defaultBlockState().setValue(BASE, false);
             BlockState aboveState = getUpdatedState(noCarpetBaseState, level, pos.above(), true);
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 EnumProperty<WallSide> property = getPropertyForFace(direction);
                 if (aboveState.getValue(property) != WallSide.NONE && !sideSurvivalTest.getAsBoolean()) {
                     aboveState = aboveState.setValue(property, WallSide.NONE);
