--- a/net/minecraft/world/level/block/NyliumBlock.java
+++ b/net/minecraft/world/level/block/NyliumBlock.java
@@ -30,8 +_,7 @@
     }
 
     private static boolean canBeNylium(BlockState state, LevelReader level, BlockPos pos) {
-        BlockPos blockPos = pos.above();
-        BlockState blockState = level.getBlockState(blockPos);
+        BlockState blockState = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
         int lightBlockInto = LightEngine.getLightBlockInto(state, blockState, Direction.UP, blockState.getLightBlock());
         return lightBlockInto < 15;
     }
