--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -111,6 +_,18 @@
         return state.getSignal(level, pos, side);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        return blockState.pluto$getSignal(blockAccess, x, y, z, side);
+    }
+
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction side) {
+        return blockState.getValue(POWERED) && blockState.getValue(FACING) == side ? 15 : 0;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
         return state.getValue(POWERED) && state.getValue(FACING) == side ? 15 : 0;
