--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -112,6 +_,18 @@
         return state.getSignal(level, pos, direction);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getDirectSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.pluto$getSignal(level, x, y, z, direction);
+    }
+
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.getValue(POWERED) && state.getValue(FACING) == direction ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
         return state.getValue(POWERED) && state.getValue(FACING) == direction ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
