--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -157,7 +_,7 @@
     private BlockState getMissingConnections(BlockGetter level, BlockState state, BlockPos pos) {
         boolean flag = !level.getBlockState(pos.above()).isRedstoneConductor(level, pos);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (!state.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected()) {
                 RedstoneSide connectingSide = this.getConnectingSide(level, pos, direction, flag);
                 state = state.setValue(PROPERTY_BY_DIRECTION.get(direction), connectingSide);
@@ -210,7 +_,7 @@
     protected void updateIndirectNeighbourShapes(BlockState state, LevelAccessor level, BlockPos pos, int flags, int recursionLeft) {
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             RedstoneSide redstoneSide = state.getValue(PROPERTY_BY_DIRECTION.get(direction));
             if (redstoneSide != RedstoneSide.NONE && !level.getBlockState(mutableBlockPos.setWithOffset(pos, direction)).is(this)) {
                 mutableBlockPos.move(Direction.DOWN);
@@ -235,12 +_,17 @@
     }
 
     private RedstoneSide getConnectingSide(BlockGetter level, BlockPos pos, Direction direction, boolean nonNormalCubeAbove) {
-        BlockPos blockPos = pos.relative(direction);
-        BlockState blockState = level.getBlockState(blockPos);
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX() + direction.getStepX();
+        final int y = pos.getY() + direction.getStepY();
+        final int z = pos.getZ() + direction.getStepZ();
+        BlockPos blockPos = null;
+        BlockState blockState = level.pluto$getBlockState(x, y, z);
         if (nonNormalCubeAbove) {
-            boolean flag = blockState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, blockPos, blockState);
-            if (flag && shouldConnectTo(level.getBlockState(blockPos.above()))) {
-                if (blockState.isFaceSturdy(level, blockPos, direction.getOpposite())) {
+            boolean flag = blockState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, blockPos = new BlockPos(x, y, z), blockState);
+            if (flag && shouldConnectTo(level.pluto$getBlockState(x, y + 1, z))) {// Pluto end - Memory allocation optimisations
+                if (blockState.isFaceSturdy(level, blockPos == null ? new BlockPos(x, y, z) : blockPos, direction.getOpposite())) {
+                    // Pluto end - Memory allocation optimisations
                     return RedstoneSide.UP;
                 }
 
@@ -249,7 +_,7 @@
         }
 
         return !shouldConnectTo(blockState, direction)
-                && (blockState.isRedstoneConductor(level, blockPos) || !shouldConnectTo(level.getBlockState(blockPos.below())))
+                && (blockState.isRedstoneConductor(level, blockPos == null ? new BlockPos(x, y, z) : blockPos) || !shouldConnectTo(level.pluto$getBlockState(x, y - 1, z))) // Pluto - Memory allocation optimisations
             ? RedstoneSide.NONE
             : RedstoneSide.SIDE;
     }
@@ -338,7 +_,7 @@
         if (level.getBlockState(pos).is(this)) {
             level.updateNeighborsAt(pos, this);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
         }
@@ -355,7 +_,7 @@
             }
             // Paper end - optimize redstone
 
-            for (Direction direction : Direction.Plane.VERTICAL) {
+            for (Direction direction : Direction.Plane.VERTICAL.faces) { // Pluto - Expose Direction$Plane's faces
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
@@ -366,7 +_,7 @@
     @Override
     protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel level, BlockPos pos, boolean movedByPiston) {
         if (!movedByPiston) {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
@@ -382,11 +_,11 @@
     }
 
     private void updateNeighborsOfNeighboringWires(Level level, BlockPos pos) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             this.checkCornerChangeAt(level, pos.relative(direction));
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos blockPos = pos.relative(direction);
             if (level.getBlockState(blockPos).isRedstoneConductor(level, blockPos)) {
                 this.checkCornerChangeAt(level, blockPos.above());
@@ -425,6 +_,76 @@
         return !this.shouldSignal ? 0 : blockState.getSignal(blockAccess, pos, side);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        return !this.shouldSignal ? 0 : blockState.pluto$getSignal(blockAccess, x, y, z, side);
+    }
+
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        if (this.shouldSignal && side != Direction.DOWN) {
+            int powerValue = blockState.getValue(POWER);
+            if (powerValue == 0) {
+                return 0;
+            } else {
+                return side != Direction.UP
+                        && !this.pluto$getConnectionState(blockAccess, blockState, x, y, z).getValue(PROPERTY_BY_DIRECTION.get(side.getOpposite())).isConnected()
+                        ? 0
+                        : powerValue;
+            }
+        } else {
+            return 0;
+        }
+    }
+
+    private BlockState pluto$getConnectionState(BlockGetter level, BlockState state, int x, int y, int z) {
+        boolean isDot = isDot(state);
+        state = this.pluto$getMissingConnections(level, this.defaultBlockState().setValue(POWER, state.getValue(POWER)), x, y, z);
+        if (isDot && isDot(state)) {
+            return state;
+        } else {
+            boolean isConnected = state.getValue(NORTH).isConnected();
+            boolean isConnected1 = state.getValue(SOUTH).isConnected();
+            boolean isConnected2 = state.getValue(EAST).isConnected();
+            boolean isConnected3 = state.getValue(WEST).isConnected();
+            boolean flag = !isConnected && !isConnected1;
+            boolean flag1 = !isConnected2 && !isConnected3;
+            if (!isConnected3 && flag) {
+                state = state.setValue(WEST, RedstoneSide.SIDE);
+            }
+
+            if (!isConnected2 && flag) {
+                state = state.setValue(EAST, RedstoneSide.SIDE);
+            }
+
+            if (!isConnected && flag1) {
+                state = state.setValue(NORTH, RedstoneSide.SIDE);
+            }
+
+            if (!isConnected1 && flag1) {
+                state = state.setValue(SOUTH, RedstoneSide.SIDE);
+            }
+
+            return state;
+        }
+    }
+
+    private BlockState pluto$getMissingConnections(BlockGetter level, BlockState state, int x, int y, int z) {
+        BlockPos pos = new BlockPos(x, y, z); // TODO: Look into replacing this by using integers.
+        boolean flag = !level.pluto$getBlockState(x, y + 1, z).isRedstoneConductor(level, pos);
+
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+            if (!state.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected()) {
+                RedstoneSide connectingSide = this.getConnectingSide(level, pos, direction, flag);
+                state = state.setValue(PROPERTY_BY_DIRECTION.get(direction), connectingSide);
+            }
+        }
+
+        return state;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected int getSignal(BlockState blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
         if (this.shouldSignal && side != Direction.DOWN) {
@@ -484,7 +_,7 @@
     public void animateTick(BlockState state, Level level, BlockPos pos, RandomSource random) {
         int powerValue = state.getValue(POWER);
         if (powerValue != 0) {
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 RedstoneSide redstoneSide = state.getValue(PROPERTY_BY_DIRECTION.get(direction));
                 switch (redstoneSide) {
                     case UP:
@@ -563,7 +_,7 @@
     private void updatesOnShapeChange(Level level, BlockPos pos, BlockState oldState, BlockState newState) {
         Orientation orientation = ExperimentalRedstoneUtils.initialOrientation(level, null, Direction.UP);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos blockPos = pos.relative(direction);
             if (oldState.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected() != newState.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected()
                 && level.getBlockState(blockPos).isRedstoneConductor(level, blockPos)) {
