--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -139,9 +_,13 @@
         double offset = 0.5625;
         RandomSource random = level.getRandom();
 
-        for (Direction direction : Direction.values()) {
-            BlockPos relative = pos.relative(direction);
-            if (!level.getBlockState(relative).isSolidRender()) {
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+            if (!level.pluto$getBlockState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ()).isSolidRender()) {
+                // Pluto end - Memory allocation optimisations
                 Direction.Axis axis = direction.getAxis();
                 double dx = axis == Direction.Axis.X ? 0.5 + 0.5625 * direction.getStepX() : random.nextFloat();
                 double dy = axis == Direction.Axis.Y ? 0.5 + 0.5625 * direction.getStepY() : random.nextFloat();
