--- a/net/minecraft/world/level/block/RedstoneWallTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneWallTorchBlock.java
@@ -90,6 +_,13 @@
         return state.getValue(LIT) && state.getValue(FACING) != direction ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.getValue(LIT) && state.getValue(FACING) != direction ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected BlockState rotate(final BlockState state, final Rotation rotation) {
         return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
