--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -165,7 +_,7 @@
             return 0;
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockState relativeState = level.getBlockState(relativePos.setWithOffset(pos, direction));
             if (relativeState.is(Blocks.SCAFFOLDING)) {
                 distance = Math.min(distance, relativeState.getValue(DISTANCE) + 1);
