--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -196,6 +_,18 @@
         return direction == Direction.UP ? state.getSignal(level, pos, direction) : Redstone.SIGNAL_MIN;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.getValue(POWER);
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return direction == Direction.UP ? state.pluto$getSignal(level, x, y, z, direction) : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     public static SculkSensorPhase getPhase(final BlockState state) {
         return state.getValue(PHASE);
     }
@@ -256,10 +_,18 @@
     }
 
     public static void tryResonateVibration(final @Nullable Entity sourceEntity, final Level level, final BlockPos pos, final int vibrationFrequency) {
-        for (Direction direction : Direction.values()) {
-            BlockPos relativePos = pos.relative(direction);
-            BlockState blockState = level.getBlockState(relativePos);
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
+            final int x1 = x + direction.getStepX();
+            final int y1 = y + direction.getStepY();
+            final int z1 = z + direction.getStepZ();
+            BlockState blockState = level.pluto$getBlockState(x1, y1, z1);
             if (blockState.is(BlockTags.VIBRATION_RESONATORS)) {
+                BlockPos relativePos = new BlockPos(x1, y1, z1);
+                // Pluto end - Memory allocation optimisations
                 level.gameEvent(VibrationSystem.getResonanceEventByFrequency(vibrationFrequency), relativePos, GameEvent.Context.of(sourceEntity, blockState));
                 float pitch = RESONANCE_PITCH_BEND[vibrationFrequency];
                 level.playSound(null, relativePos, SoundEvents.AMETHYST_BLOCK_RESONATE, SoundSource.BLOCKS, 1.0F, pitch);
