--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -198,6 +_,18 @@
         return direction == Direction.UP ? state.getSignal(level, pos, direction) : 0;
     }
 
+    // Pluto start - Memory efficient block lookups
+    @Override
+    protected int pluto$getSignal(BlockState state, BlockGetter level, int x, int y, int z, Direction direction) {
+        return state.getValue(POWER);
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState state, BlockGetter level, int x, int y, int z, Direction direction) {
+        return direction == Direction.UP ? state.pluto$getSignal(level, x, y, z, direction) : 0;
+    }
+    // Pluto end - Memory efficient block lookups
+
     public static SculkSensorPhase getPhase(BlockState state) {
         return state.getValue(PHASE);
     }
