--- a/net/minecraft/world/level/block/SeaPickleBlock.java
+++ b/net/minecraft/world/level/block/SeaPickleBlock.java
@@ -51,7 +_,7 @@
         if (blockState.is(this)) {
             return blockState.setValue(PICKLES, Math.min(4, blockState.getValue(PICKLES) + 1));
         } else {
-            FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
+            FluidState fluidState = blockState.getFluidState(); // Pluto - Decrease chunk/block lookups
             boolean flag = fluidState.getType() == Fluids.WATER;
             return super.getStateForPlacement(context).setValue(WATERLOGGED, flag);
         }
@@ -122,7 +_,7 @@
 
     @Override
     public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state) {
-        return !isDead(state) && level.getBlockState(pos.below()).is(BlockTags.CORAL_BLOCKS);
+        return !isDead(state) && level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()).is(BlockTags.CORAL_BLOCKS); // Pluto - Memory efficient block lookups
     }
 
     @Override
@@ -146,7 +_,7 @@
                 for (int i9 = i8 - 2; i9 < i8; i9++) {
                     BlockPos blockPos = new BlockPos(i4 + i6, i9, pos.getZ() - i5 + i7);
                     if (blockPos != pos && random.nextInt(6) == 0 && level.getBlockState(blockPos).is(Blocks.WATER)) {
-                        BlockState blockState = level.getBlockState(blockPos.below());
+                        BlockState blockState = level.pluto$getBlockState(blockPos.getX(), blockPos.getY() - 1, blockPos.getZ()); // Pluto - Memory efficient block lookups | I'm not going to do the above since this is just for bonemeal
                         if (blockState.is(BlockTags.CORAL_BLOCKS)) {
                             level.setBlock(blockPos, Blocks.SEA_PICKLE.defaultBlockState().setValue(PICKLES, random.nextInt(4) + 1), 3);
                         }
