--- a/net/minecraft/world/level/block/SmallDripleafBlock.java
+++ b/net/minecraft/world/level/block/SmallDripleafBlock.java
@@ -52,7 +_,7 @@
     @Override
     protected boolean mayPlaceOn(final BlockState state, final BlockGetter level, final BlockPos pos) {
         return state.is(BlockTags.SUPPORTS_SMALL_DRIPLEAF)
-            || level.getFluidState(pos.above()).isSourceOfType(Fluids.WATER) && super.mayPlaceOn(state, level, pos);
+            || level.pluto$getFluidState(pos.getX(), pos.getY() + 1, pos.getZ()).isSourceOfType(Fluids.WATER) && super.mayPlaceOn(state, level, pos); // Pluto - Memory allocation optimisations
     }
 
     @Override
