--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -48,10 +_,18 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        if (level.isEmptyBlock(pos.above())) {
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        if (chunk.getBlockState(x, y + 1, z).isAir()) {
+            // Pluto end - Decrease chunk/block lookups
             int height = 1;
 
-            while (level.getBlockState(pos.below(height)).is(this)) {
+            while (chunk.getBlockState(x, y - height, z).is(this)) { // Pluto - Decrease chunk/block lookups
                 height++;
             }
 
@@ -88,17 +_,22 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
-        BlockState stateBelow = level.getBlockState(pos.below());
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockState stateBelow = level.pluto$getBlockState(x, y - 1, z);
+        // Pluto end - Memory allocation optimisations
         if (stateBelow.is(this)) {
             return true;
         }
 
         if (stateBelow.is(BlockTags.SUPPORTS_SUGAR_CANE)) {
-            BlockPos below = pos.below();
+            if (level instanceof ServerLevel serverLevel && serverLevel.plutoConfig().alwaysMoistSugarCane) return true; // Pluto - Always moist blocks
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
-                BlockState blockState = level.getBlockState(below.relative(direction));
-                FluidState fluidState = level.getFluidState(below.relative(direction));
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+                BlockState blockState = level.pluto$getBlockState(x + direction.getStepX(), (y - 1) + direction.getStepY(), z + direction.getStepZ()); // Pluto - Memory allocation optimisations
+                FluidState fluidState = blockState.getFluidState(); // Pluto - Decrease chunk/block lookups
                 if (fluidState.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockState.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY)) {
                     return true;
                 }
