--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -48,10 +_,18 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.isEmptyBlock(pos.above())) {
+        // Pluto start - Decrease chunk/block lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        if (chunk.getBlockState(x, y + 1, z).isAir()) {
+            // Pluto end - Decrease chunk/block lookups
             int i = 1;
 
-            while (level.getBlockState(pos.below(i)).is(this)) {
+            while (chunk.getBlockState(x, y - i, z).is(this)) { // Pluto - Decrease chunk/block lookups
                 i++;
             }
 
@@ -88,16 +_,21 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        BlockState blockState = level.getBlockState(pos.below());
+        // Pluto start - Memory efficient block lookups
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        BlockState blockState = level.pluto$getBlockState(x, y - 1, z);
+        // Pluto end - Memory efficient block lookups
         if (blockState.is(this)) {
             return true;
         } else {
             if (blockState.is(BlockTags.DIRT) || blockState.is(BlockTags.SAND)) {
-                BlockPos blockPos = pos.below();
+                if (level instanceof ServerLevel serverLevel && serverLevel.plutoConfig().alwaysMoistSugarCane) return true; // Pluto - Always moist blocks
 
                 for (Direction direction : Direction.Plane.HORIZONTAL) {
-                    BlockState blockState1 = level.getBlockState(blockPos.relative(direction));
-                    FluidState fluidState = level.getFluidState(blockPos.relative(direction));
+                    BlockState blockState1 = level.pluto$getBlockState(x + direction.getStepX(), (y - 1) + direction.getStepY(), z + direction.getStepZ()); // Pluto - Memory efficient block lookups
+                    FluidState fluidState = blockState1.getFluidState(); // Pluto - Decrease chunk/block lookups
                     if (fluidState.is(FluidTags.WATER) || blockState1.is(Blocks.FROSTED_ICE)) {
                         return true;
                     }
