--- a/net/minecraft/world/level/block/TargetBlock.java
+++ b/net/minecraft/world/level/block/TargetBlock.java
@@ -117,6 +_,13 @@
         return state.getValue(OUTPUT_POWER);
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.getValue(OUTPUT_POWER);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected boolean isSignalSource(final BlockState state) {
         return true;
