--- a/net/minecraft/world/level/block/TrappedChestBlock.java
+++ b/net/minecraft/world/level/block/TrappedChestBlock.java
@@ -52,4 +_,16 @@
     protected int getDirectSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
         return side == Direction.UP ? state.getSignal(level, pos, side) : 0;
     }
+
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(BlockState state, BlockGetter blockAccess, int x, int y, int z, Direction direction) {
+        return Mth.clamp(ChestBlockEntity.pluto$getOpenCount(blockAccess, x, y, z), 0, 15);
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        return side == Direction.UP ? blockState.pluto$getSignal(blockAccess, x, y, z, side) : 0;
+    }
+    // Pluto end - Memory allocation optimisations
 }
