--- a/net/minecraft/world/level/block/TrappedChestBlock.java
+++ b/net/minecraft/world/level/block/TrappedChestBlock.java
@@ -53,4 +_,16 @@
     protected int getDirectSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
         return direction == Direction.UP ? state.getSignal(level, pos, direction) : Redstone.SIGNAL_MIN;
     }
+
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return Mth.clamp(ChestBlockEntity.pluto$getOpenCount(level, x, y, z), 0, 15);
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return direction == Direction.UP ? state.pluto$getSignal(level, x, y, z, direction) : Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
 }
