--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -283,6 +_,22 @@
         }
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(BlockState blockState, BlockGetter level, int x, int y, int z, Direction direction) {
+        return blockState.getValue(POWERED) ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(BlockState blockState, BlockGetter blockAccess, int x, int y, int z, Direction side) {
+        if (!blockState.getValue(POWERED)) {
+            return Redstone.SIGNAL_MIN;
+        } else {
+            return blockState.getValue(FACING) == side ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+        }
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected boolean isSignalSource(final BlockState state) {
         return true;
