--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -273,6 +_,22 @@
         }
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return state.getValue(POWERED) ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+    }
+
+    @Override
+    protected int pluto$getDirectSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        if (!state.getValue(POWERED)) {
+            return Redstone.SIGNAL_MIN;
+        } else {
+            return state.getValue(FACING) == direction ? Redstone.SIGNAL_MAX : Redstone.SIGNAL_MIN;
+        }
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     protected boolean isSignalSource(final BlockState state) {
         return true;
