--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -125,7 +_,7 @@
 
         BlockState aboveState = null;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BooleanProperty property = getPropertyForFace(direction);
             if (state.getValue(property)) {
                 boolean canSupport = this.canSupportAtFace(level, pos, direction);
@@ -217,7 +_,7 @@
 
                             BlockState aboveState = state;
 
-                            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                 if (random.nextBoolean() || !isAcceptableNeighbour(level, abovePos.relative(direction), direction)) {
                                     aboveState = aboveState.setValue(getPropertyForFace(direction), false);
                                 }
@@ -248,7 +_,7 @@
     }
 
     private BlockState copyRandomFaces(final BlockState from, BlockState to, final RandomSource random) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (random.nextBoolean()) {
                 BooleanProperty propertyForFace = getPropertyForFace(direction);
                 if (from.getValue(propertyForFace)) {
