--- a/net/minecraft/world/level/block/WaterlilyBlock.java
+++ b/net/minecraft/world/level/block/WaterlilyBlock.java
@@ -49,8 +_,12 @@
 
     @Override
     protected boolean mayPlaceOn(BlockState state, BlockGetter level, BlockPos pos) {
-        FluidState fluidState = level.getFluidState(pos);
-        FluidState fluidState1 = level.getFluidState(pos.above());
-        return (fluidState.getType() == Fluids.WATER || state.getBlock() instanceof IceBlock) && fluidState1.getType() == Fluids.EMPTY;
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        FluidState fluidState = level.pluto$getFluidState(x, y, z);
+        return (fluidState.getType() == Fluids.WATER || state.getBlock() instanceof IceBlock) && level.pluto$getFluidState(x, y + 1, z).getType() == Fluids.EMPTY; // Pluto - Decrease chunk/block lookups
+        // Pluto end - Memory allocation optimisations
     }
 }
