--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -176,8 +_,8 @@
                 levelData.isHardcore(),
                 this.server.levelKeys(),
                 this.getMaxPlayers(),
-                serverLevel.spigotConfig.viewDistance, // Spigot - view distance
-                serverLevel.spigotConfig.simulationDistance, // Spigot - simulation distance
+                io.papermc.paper.FeatureHooks.getViewDistance(serverLevel), // Paper - view distance
+                io.papermc.paper.FeatureHooks.getSimulationDistance(serverLevel), // Paper - simulation distance
                 _boolean1,
                 !_boolean,
                 _boolean2,
@@ -984,10 +_,10 @@
         player.connection.send(new ClientboundInitializeBorderPacket(worldBorder));
         player.connection.send(new ClientboundSetTimePacket(level.getGameTime(), level.getDayTime(), level.getGameRules().getBoolean(GameRules.RULE_DAYLIGHT)));
         player.connection.send(new ClientboundSetDefaultSpawnPositionPacket(level.getRespawnData()));
-        // Paper start
-        player.connection.send(new ClientboundSetChunkCacheRadiusPacket(level.spigotConfig.viewDistance));
-        player.connection.send(new ClientboundSetSimulationDistancePacket(level.spigotConfig.simulationDistance));
-        // Paper end
+        // Paper start - view distances
+        //player.connection.send(new ClientboundSetChunkCacheRadiusPacket(io.papermc.paper.FeatureHooks.getViewDistance(level))); // Paper - rewrite chunk system
+        //player.connection.send(new ClientboundSetSimulationDistancePacket(io.papermc.paper.FeatureHooks.getSimulationDistance(level))); // Paper - rewrite chunk system
+        // Paper end - view distances
         if (level.isRaining()) {
             // CraftBukkit start - handle player weather
             // player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
