--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -19,17 +_,33 @@
         this.level = level;
     }
 
+    private static final it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, org.bukkit.GameEvent> GAME_EVENT_CACHE = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Pluto - Cache GameEvent lookups
     public void post(Holder<GameEvent> gameEvent, Vec3 pos, GameEvent.Context context) {
         int notificationRadius = gameEvent.value().notificationRadius();
         BlockPos blockPos = BlockPos.containing(pos);
+        // Pluto start - Cache GameEvent lookups
+        org.bukkit.GameEvent bukkitEvent;
+        if (this.level.plutoConfig().useGameEventCache) {
+            String name = gameEvent.getRegisteredName();
+            if (name.contains("[unregistered]")) {
+                bukkitEvent = org.bukkit.craftbukkit.CraftGameEvent.minecraftToBukkit(gameEvent.value());
+            } else {
+                bukkitEvent = GAME_EVENT_CACHE.computeIfAbsent(name, s -> org.bukkit.craftbukkit.CraftGameEvent.minecraftToBukkit(gameEvent.value()));
+            }
+        } else {
+            bukkitEvent = org.bukkit.craftbukkit.CraftGameEvent.minecraftToBukkit(gameEvent.value());
+        }
+        if (bukkitEvent != null) {
+        // Pluto end - Cache GameEvent lookups
         // CraftBukkit start
         org.bukkit.event.world.GenericGameEvent apiEvent = new org.bukkit.event.world.GenericGameEvent(
-            org.bukkit.craftbukkit.CraftGameEvent.minecraftToBukkit(gameEvent.value()), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockPos, this.level),
+            bukkitEvent, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockPos, this.level.getWorld()), // Pluto - Cache GameEvent lookups
             (context.sourceEntity() == null) ? null : context.sourceEntity().getBukkitEntity(), notificationRadius, !org.bukkit.Bukkit.isPrimaryThread());
         if (!apiEvent.callEvent()) {
             return;
         }
         notificationRadius = apiEvent.getRadius();
+        } // Pluto - Cache GameEvent lookups
         // CraftBukkit end
         int sectionPosCoord = SectionPos.blockToSectionCoord(blockPos.getX() - notificationRadius);
         int sectionPosCoord1 = SectionPos.blockToSectionCoord(blockPos.getY() - notificationRadius);
