--- a/net/minecraft/world/CompoundContainer.java
+++ b/net/minecraft/world/CompoundContainer.java
@@ -44,6 +_,13 @@
         return null; // This method won't be called since CraftInventoryDoubleChest doesn't defer to here
     }
 
+    // Pluto start - Allow toggling snapshots for certain blocks
+    @Override
+    public @javax.annotation.Nullable org.bukkit.inventory.InventoryHolder getOwner(boolean useSnapshot) {
+        return null;
+    }
+    // Pluto end - Allow toggling snapshots for certain blocks
+
     public void setMaxStackSize(int size) {
         this.container1.setMaxStackSize(size);
         this.container2.setMaxStackSize(size);
@@ -117,6 +_,8 @@
     public boolean stillValid(final Player player) {
         return this.container1.stillValid(player) && this.container2.stillValid(player);
     }
+
+    @Override public boolean stillValid() { return this.container1.stillValid() && this.container2.stillValid(); } // Pluto - Hoppers cache containers
 
     @Override
     public void startOpen(final ContainerUser containerUser) {
