--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -42,6 +_,14 @@
 
     boolean stillValid(Player player);
 
+    // Pluto start - Hoppers cache containers
+    boolean stillValid();
+    static boolean stillValidBlockEntity(BlockEntity blockEntity) {
+        Level level = blockEntity.getLevel();
+        return level != null && level.getBlockEntity(blockEntity.getBlockPos()) == blockEntity;
+    }
+    // Pluto end - Hoppers cache containers
+
     default void startOpen(final ContainerUser containerUser) {
     }
 
@@ -156,6 +_,8 @@
     java.util.List<org.bukkit.entity.HumanEntity> getViewers();
 
     org.bukkit.inventory.@Nullable InventoryHolder getOwner();
+
+    @javax.annotation.Nullable org.bukkit.inventory.InventoryHolder getOwner(boolean useSnapshot); // Pluto - Allow toggling snapshots for certain blocks
 
     void setMaxStackSize(int size);
 
