--- a/net/minecraft/world/SimpleContainer.java
+++ b/net/minecraft/world/SimpleContainer.java
@@ -63,6 +_,13 @@
         return this.bukkitOwner;
     }
 
+    // Pluto start - Allow toggling snapshots for certain blocks
+    @Override
+    public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
+        return getOwner();
+    }
+    // Pluto end - Allow toggling snapshots for certain blocks
+
     @Override
     public org.bukkit.@Nullable Location getLocation() {
         // Paper start - Fix inventories returning null Locations
@@ -243,6 +_,8 @@
     public boolean stillValid(Player player) {
         return true;
     }
+
+    @Override public boolean stillValid() { return true; } // Pluto - Hoppers cache containers
 
     @Override
     public void clearContent() {
