--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -443,7 +_,7 @@
         }
 
         this.minecart.move(MoverType.SELF, newPosition.subtract(oldPosition));
-        BlockState newBlockState = this.level().getBlockState(BlockPos.containing(newPosition));
+        BlockState newBlockState = this.level().pluto$getBlockState(newPosition.x, newPosition.y, newPosition.z); // Pluto - Memory allocation optimisations
         if (inHill) {
             if (BaseRailBlock.isRail(newBlockState)) {
                 RailShape newRailShape = newBlockState.getValue(((BaseRailBlock)newBlockState.getBlock()).getShapeProperty());
@@ -522,7 +_,14 @@
         if (this.minecart.isRideable() && !this.minecart.isVehicle()) {
             List<Entity> entities = this.level().getEntities(this.minecart, hitbox, EntitySelector.pushableBy(this.minecart));
             if (!entities.isEmpty()) {
+                // Pluto start - Option to remove excess vehicles on collision
+                int minecartCount = 0;
                 for (Entity entity : entities) {
+                    if (entity instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                        entity.discard();
+                        continue;
+                    }
+                    // Pluto end - Option to remove excess vehicles on collision
                     if (!(entity instanceof Player)
                         && !(entity instanceof IronGolem)
                         && !(entity instanceof AbstractMinecart)
@@ -549,10 +_,17 @@
 
     public boolean pushEntities(final AABB hitbox) {
         boolean pushed = false;
+        int minecartCount = 0; // Pluto - Option to remove excess vehicles on collision
         if (this.minecart.isRideable()) {
             List<Entity> entities = this.level().getEntities(this.minecart, hitbox, EntitySelector.pushableBy(this.minecart));
             if (!entities.isEmpty()) {
                 for (Entity entity : entities) {
+                    // Pluto start - Option to remove excess vehicles on collision
+                    if (entity instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                        entity.discard();
+                        continue;
+                    }
+                    // Pluto end - Option to remove excess vehicles on collision
                     if (entity instanceof Player
                         || entity instanceof IronGolem
                         || entity instanceof AbstractMinecart
@@ -576,6 +_,12 @@
             }
         } else {
             for (Entity entity : this.level().getEntities(this.minecart, hitbox)) {
+                // Pluto start - Option to remove excess vehicles on collision
+                if (entity instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                    entity.discard();
+                    continue;
+                }
+                // Pluto end - Option to remove excess vehicles on collision
                 if (!this.minecart.hasPassenger(entity) && entity.isPushable() && entity instanceof AbstractMinecart) {
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
