--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -610,7 +_,7 @@
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
             BlockState blockState = level.getBlockState(pos);
             if (blockState.is(Blocks.FARMLAND) && this.wantsToRaid && !this.canRaid) {
-                blockState = level.getBlockState(pos.above());
+                blockState = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
                 if (blockState.getBlock() instanceof CarrotBlock && ((CarrotBlock)blockState.getBlock()).isMaxAge(blockState)) {
                     this.canRaid = true;
                     return true;
