--- a/net/minecraft/world/entity/animal/Turtle.java
+++ b/net/minecraft/world/entity/animal/Turtle.java
@@ -391,7 +_,7 @@
                     posTowards = DefaultRandomPos.getPosTowards(this.turtle, 8, 7, vec3, (float) (Math.PI / 2));
                 }
 
-                if (posTowards != null && !flag && !this.turtle.level().getBlockState(BlockPos.containing(posTowards)).is(Blocks.WATER)) {
+                if (posTowards != null && !flag && !this.turtle.level().pluto$getBlockState(posTowards.asBlockPosLong()).is(Blocks.WATER)) { // Pluto - Memory allocation optimisations
                     posTowards = DefaultRandomPos.getPosTowards(this.turtle, 16, 5, vec3, (float) (Math.PI / 2));
                 }
 
@@ -489,7 +_,7 @@
 
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-            return level.isEmptyBlock(pos.above()) && TurtleEggBlock.isSand(level, pos);
+            return level.pluto$isEmptyBlock(pos.getX(), pos.getY() + 1, pos.getZ()) && TurtleEggBlock.isSand(level, pos); // Pluto - Memory allocation optimisations
         }
     }
 
