--- a/net/minecraft/world/entity/animal/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/WaterAnimal.java
@@ -78,7 +_,9 @@
         // Paper end - Make water animal spawn height configurable
         return pos.getY() >= i
             && pos.getY() <= seaLevel
-            && level.getFluidState(pos.below()).is(FluidTags.WATER)
-            && level.getBlockState(pos.above()).is(Blocks.WATER);
+            // Pluto start - Memory efficient block lookups
+            && level.pluto$getFluidState(pos.getX(), pos.getY() - 1, pos.getZ()).is(FluidTags.WATER)
+            && level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()).is(Blocks.WATER);
+            // Pluto ebd - Memory efficient block lookups
     }
 }
