--- a/src/main/java/io/papermc/paper/inventory/PaperInventoryCustomHolderContainer.java
+++ b/src/main/java/io/papermc/paper/inventory/PaperInventoryCustomHolderContainer.java
@@ -124,6 +_,13 @@
         return this.owner;
     }
 
+    // Pluto start - Allow toggling snapshots for certain blocks
+    @Override
+    public InventoryHolder getOwner(boolean useSnapshot) {
+        return getOwner();
+    }
+    // Pluto end - Allow toggling snapshots for certain blocks
+
     @Override
     public void setMaxStackSize(int size) {
         this.delegate.setMaxStackSize(size);
