--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -254,6 +_,7 @@
         final ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
         for (int i = 0, len = inRange.size(); i < len; i++) {
             ++(backingSet[i].mobCounts[index]);
+            if (entity.level().plutoConfig().useAsyncSpawningChunks) backingSet[i].mobCountsAsync.put(index, backingSet[i].mobCountsAsync.getOrDefault(index, 0) + 1); // Pluto - Async Spawning Chunks
         }
     }
 
@@ -271,10 +_,12 @@
         final ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
         for (int i = 0, len = inRange.size(); i < len; i++) {
             ++(backingSet[i].mobBackoffCounts[idx]);
+            if (this.level.plutoConfig().useAsyncSpawningChunks) backingSet[i].mobBackoffCountsAsync.put(idx, backingSet[i].mobBackoffCounts[idx]); // Pluto - Async Spawning Chunks
         }
     }
     // Paper end - per player mob count backoff
     public int getMobCountNear(final ServerPlayer player, final net.minecraft.world.entity.MobCategory mobCategory) {
+        if (player.level().plutoConfig().useAsyncSpawningChunks) return player.mobCountsAsync.getOrDefault(mobCategory.ordinal(), 0) + player.mobBackoffCountsAsync.getOrDefault(mobCategory.ordinal(), 0); // Pluto - Async Spawning Chunks
         return player.mobCounts[mobCategory.ordinal()] + player.mobBackoffCounts[mobCategory.ordinal()]; // Paper - per player mob count backoff
     }
     // Paper end - Optional per player mob spawns
@@ -748,6 +_,7 @@
         Objects.checkFromToIndex(0, size, raw.length);
         for (int i = 0; i < size; ++i) {
             final LevelChunk levelChunk = raw[i];
+            if (levelChunk == null) continue; // Pluto - Async Spawning Chunks
 
             if (!this.isChunkNearPlayer((ChunkMap)(Object)this, levelChunk.getPos(), levelChunk)) {
                 continue;
@@ -1383,6 +_,7 @@
                     visibleToPlayer = false;
                 }
                 // CraftBukkit end
+                if (visibleToPlayer && level.plutoConfig().disablePlayerTracking && this.entity instanceof ServerPlayer) visibleToPlayer = false; // Pluto - Option to disable players seeing other players.
                 if (visibleToPlayer) {
                     if (this.seenBy.add(player.connection)) {
                         // Paper start - entity tracking events
