--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -526,16 +_,20 @@
 
             // Pufferfish start - optimize mob spawning
             if (gg.pufferfish.pufferfish.PufferfishConfig.enableAsyncMobSpawning) {
-                for (ServerPlayer player : this.level.players) {
+                // Pluto start - Swap for loops around for per player mob spawns
+                for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
                     // Paper start - per player mob spawning backoff
-                    for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
+                    for (ServerPlayer player : this.level.players) {
+                        // Pluto end - Swap for loops around for per player mob spawns
                         player.mobCounts[ii] = 0;
+                        if (this.level.plutoConfig().useAsyncSpawningChunks) player.mobCountsAsync.put(ii, 0); // Pluto - Async Spawning Chunks
 
-                        int newBackoff = player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
+                        int newBackoff = this.level.plutoConfig().useAsyncSpawningChunks ? player.mobBackoffCountsAsync.getOrDefault(ii, 0) - 1 : player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm? // Pluto - Async Spawning Chunks
                         if (newBackoff < 0) {
                             newBackoff = 0;
                         }
                         player.mobBackoffCounts[ii] = newBackoff;
+                        if (this.level.plutoConfig().useAsyncSpawningChunks) player.mobBackoffCountsAsync.put(ii, newBackoff); // Pluto - Async Spawning Chunks
                     }
                     // Paper end - per player mob spawning backoff
                 }
@@ -577,6 +_,7 @@
         profiler.pop();
     }
 
+    public final java.util.concurrent.atomic.AtomicBoolean pluto$stillSpawningEntities = new java.util.concurrent.atomic.AtomicBoolean(false); // Pluto - Async Spawning Chunks
     private void tickChunks(ProfilerFiller profiler, long timeInhabited) {
         profiler.popPush("naturalSpawnCount");
         int naturalSpawnChunkCount = this.distanceManager.getNaturalSpawnChunkCount();
@@ -585,16 +_,20 @@
         if ((this.spawnFriendlies || this.spawnEnemies) && this.level.paperConfig().entities.spawning.perPlayerMobSpawns) { // don't count mobs when animals and monsters are disabled
             if (!gg.pufferfish.pufferfish.PufferfishConfig.enableAsyncMobSpawning) { // Pufferfish - moved down when async processing
             // re-set mob counts
-            for (ServerPlayer player : this.level.players) {
+            // Pluto start - Swap for loops around for per player mob spawns
+            for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
                 // Paper start - per player mob spawning backoff
-                for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
+                for (ServerPlayer player : this.level.players) {
+                    // Pluto end - Swap for loops around for per player mob spawns
                     player.mobCounts[ii] = 0;
+                    if (this.level.plutoConfig().useAsyncSpawningChunks) player.mobCountsAsync.put(ii, 0); // Pluto - Async Spawning Chunks
 
-                    int newBackoff = player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
+                    int newBackoff = this.level.plutoConfig().useAsyncSpawningChunks ? player.mobBackoffCountsAsync.getOrDefault(ii, 0) - 1 : player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm? // Pluto - Async Spawning Chunks
                     if (newBackoff < 0) {
                         newBackoff = 0;
                     }
                     player.mobBackoffCounts[ii] = newBackoff;
+                    if (this.level.plutoConfig().useAsyncSpawningChunks) player.mobBackoffCountsAsync.put(ii, newBackoff); // Pluto - Async Spawning Chunks
                 }
                 // Paper end - per player mob spawning backoff
             }
@@ -625,6 +_,34 @@
             filteredSpawningCategories = List.of();
         }
 
+        // Pluto start - Async spawning chunks
+        spawningChunks:
+        if (this.level.plutoConfig().useAsyncSpawningChunks) {
+            if (this.pluto$stillSpawningEntities.get()) break spawningChunks;
+            this.pluto$stillSpawningEntities.set(true);
+            final long next = this.level.random.nextLong(); // Considering this is unsafe, we should do this on the main thread.
+            this.level.plutoSpawningChunksExecutor().submit(() -> {
+                List<LevelChunk> list = new ObjectArrayList<>();
+
+                try {
+                    this.chunkMap.collectSpawningChunks(list);
+                    // Paper start - chunk tick iteration optimisation
+                    this.shuffleRandom.setSeed(next); // This random is only used for spawning chunks.
+                    if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(list, this.shuffleRandom); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+                    // Paper end - chunk tick iteration optimisation
+
+                    for (LevelChunk levelChunk : list) {
+                        // Probably should do this outside of the loop, but it is possible for this to become populated during the loop.
+                        if (lastSpawnState == null) continue;
+                        this.tickSpawningChunk(levelChunk, timeInhabited, filteredSpawningCategories, lastSpawnState); // Pufferfish
+                    }
+                } finally {
+                    list.clear();
+                    level.chunkSource.pluto$stillSpawningEntities.set(false);
+                }
+            });
+        } else {
+        // Pluto end - Async spawning chunks
         List<LevelChunk> list = this.spawningChunks;
 
         try {
@@ -643,6 +_,7 @@
         } finally {
             list.clear();
         }
+        } // Pluto - Async spawning chunks
 
         profiler.popPush("tickTickingChunks");
         this.iterateTickingChunksFaster(); // Paper - chunk tick iteration optimisations
@@ -662,6 +_,7 @@
 
         if (!spawnCategories.isEmpty()) {
             if (this.level.getWorldBorder().isWithinBounds(pos) && (!gg.pufferfish.pufferfish.PufferfishConfig.enableAsyncMobSpawning || _pufferfish_spawnCountsReady.get())) { // Pufferfish
+                if (this.level.plutoConfig().useAsyncSpawningChunks) NaturalSpawner.pluto$spawnForChunk(this.level, chunk, spawnState, spawnCategories); else // Pluto - Async spawning chunks
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnState, spawnCategories);
             }
         }
