--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -118,8 +_,10 @@
         }
 
         if (!this.trackedPlayers.isEmpty() && this.entity instanceof ItemFrame frame /*&& this.tickCount % 10 == 0*/) { // CraftBukkit - moved tickCount below // Paper - Perf: Only tick item frames if players can see it
-            ItemStack itemStack = frame.getItem();
-            if (this.level.paperConfig().maps.itemFrameCursorUpdateInterval > 0 && this.tickCount % this.level.paperConfig().maps.itemFrameCursorUpdateInterval == 0 && itemStack.getItem() instanceof MapItem) { // CraftBukkit - Moved this.tickCounter % 10 logic here so item frames do not enter the other blocks // Paper - Make item frame map cursor update interval configurable
+            // Pluto start - Optimise tracked item frames
+            ItemStack itemStack;
+            if (this.level.paperConfig().maps.itemFrameCursorUpdateInterval > 0 && this.tickCount % this.level.paperConfig().maps.itemFrameCursorUpdateInterval == 0 && (itemStack = frame.getItem()).getItem() instanceof MapItem) { // CraftBukkit - Moved this.tickCounter % 10 logic here so item frames do not enter the other blocks // Paper - Make item frame map cursor update interval configurable
+                // Pluto end - Optimise tracked item frames
                 MapId id = frame.cachedMapId; // Paper - Perf: Cache map ids on item frames
                 MapItemSavedData data = MapItem.getSavedData(id, this.level);
                 if (data != null) {
@@ -186,7 +_,10 @@
                     packet = ClientboundEntityPositionSyncPacket.of(this.entity);
                     sentPosition = true;
                     sentRotation = true;
-                } else if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
+                // Pufferfish start - Better checking for useless move packets
+                } else if (pos || shouldSendRotation || this.entity instanceof AbstractArrow) {
+                    if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
+                    // Pufferfish end
                     if (pos) {
                         packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround());
                         sentPosition = true;
@@ -199,6 +_,7 @@
                     sentPosition = true;
                     sentRotation = true;
                 }
+                } // Pufferfish - Better checking for useless move packets
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity livingEntity && livingEntity.isFallFlying()) {
                     Vec3 movement = this.entity.getDeltaMovement();
