--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -865,6 +_,15 @@
                 super.tick();
             }
 
+            // Pluto start - Only tick inventory slots that are important
+            // This drops the ticked inventory slots from 41 -> 2.
+            if (this.level().plutoConfig().playerInventoryOnlyUpdateImportantSlots) {
+                for (ItemStack item : this.getHandSlots()) {
+                    if (item.isEmpty()) continue;
+                    synchronizeSpecialItemUpdates(item); // FIXME: If mojang expects this to handle non-map items.
+                }
+            } else
+            // Pluto end - Only tick inventory slots that are important
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack item = this.getInventory().getItem(i);
                 if (!item.isEmpty()) {
