--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -195,7 +_,7 @@
     public final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final io.papermc.paper.world.saveddata.PaperLevelOverrides serverLevelData; // Paper - type
-    private final EntityTickList entityTickList = new EntityTickList();
+    public final EntityTickList entityTickList = new EntityTickList(); // Pufferfish
     private final ServerWaypointManager waypointManager;
     private EnvironmentAttributeSystem environmentAttributes;
     // Paper - rewrite chunk system
@@ -828,14 +_,17 @@
         if (!this.isDebug() && runs) {
             long tick = this.getGameTime();
             profiler.push("blockTicks");
+            if (plutoConfig().blockTickingModulo > 0 && (plutoConfig().blockTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().blockTickingModulo == 0)) // Pluto - Configuration for block ticks
             this.blockTicks.tick(tick, paperConfig().environment.maxBlockTicks, this::tickBlock); // Paper - configurable max block ticks
             profiler.popPush("fluidTicks");
+            if (plutoConfig().fluidTickingModulo > 0 && (plutoConfig().fluidTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().fluidTickingModulo == 0)) // Pluto - Configuration for fluid ticks
             this.fluidTicks.tick(tick, paperConfig().environment.maxFluidTicks, this::tickFluid); // Paper - configurable max fluid ticks
             profiler.pop();
         }
 
         profiler.popPush("raid");
         if (runs) {
+            if (plutoConfig().raidTickingModulo > 0 && (plutoConfig().raidTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().raidTickingModulo == 0)) // Pluto - Configuration for raid ticking
             this.raids.tick(this);
         }
 
@@ -843,6 +_,7 @@
         this.getChunkSource().tick(haveTime, true);
         profiler.popPush("blockEvents");
         if (runs) {
+            if (plutoConfig().blockEventTickingModulo > 0 && (plutoConfig().blockEventTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().blockEventTickingModulo == 0)) // Pluto - Configuration for block event ticking
             this.runBlockEvents();
         }
 
@@ -865,13 +_,16 @@
                 profiler.pop();
             }
 
+            if (!plutoConfig().disableEntityActivation && MinecraftServer.currentTick % plutoConfig().entityActivationInterval == 0) // Pluto - Configurable entity activation
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
             this.entityTickList
                 .forEach(
                     entity -> {
+                        entity.activatedPriorityReset = false; // Pufferfish - DAB
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
                                 profiler.push("checkDespawn");
+                                if (plutoConfig().entityDespawnCheckModulo > 0 && (plutoConfig().entityDespawnCheckModulo == 1 || MinecraftServer.currentTick % plutoConfig().entityDespawnCheckModulo == 0)) // Pluto - Configuration for entity despawn check
                                 entity.checkDespawn();
                                 profiler.pop();
                                 if (true) { // Paper - rewrite chunk system
@@ -885,7 +_,19 @@
                                     }
 
                                     profiler.push("tick");
-                                    this.guardEntityTick(this::tickNonPassenger, entity);
+                                    // Pufferfish start - copied from this.guardEntityTick
+                                    try {
+                                        this.tickNonPassenger(entity);
+                                    } catch (Throwable var6) {
+                                        // Paper start - Prevent block entity and entity crashes
+                                        final String msg = String.format("Entity threw exception at %s:%s,%s,%s", io.papermc.paper.util.MCUtil.getLevelName(entity.level()), entity.getX(), entity.getY(), entity.getZ());
+                                        MinecraftServer.LOGGER.error(msg, var6);
+                                        getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, var6))); // Paper - ServerExceptionEvent
+                                        entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+                                        // Paper end - Prevent block entity and entity crashes
+                                    }
+                                    this.moonrise$midTickTasks(); // Paper - rewrite chunk system
+                                    // Pufferfish end
                                     profiler.pop();
                                 }
                             }
@@ -1028,13 +_,38 @@
         int minZ = chunkPos.getMinBlockZ();
         ProfilerFiller profiler = Profiler.get();
         profiler.push("thunder");
-        if (!this.paperConfig().environment.disableThunder && raining && this.isThundering() && this.spigotConfig.thunderChance > 0 && this.random.nextInt(this.spigotConfig.thunderChance) == 0) { // Spigot // Paper - Option to disable thunder
+        if (!this.paperConfig().environment.disableThunder && raining && this.isThundering() && this.spigotConfig.thunderChance > 0 && /*simpleRandom.nextInt(this.spigotConfig.thunderChance) == 0*/ chunk.shouldDoLightning(this.random)) { // Spigot // Paper - Option to disable thunder // Paper - optimise random ticking // Pufferfish - replace random with shouldDoLightning
             BlockPos pos = this.findLightningTargetAround(this.getBlockRandomPos(minX, 0, minZ, 15));
             if (this.isRainingAt(pos)) {
                 DifficultyInstance difficulty = this.getCurrentDifficultyAt(pos);
                 boolean isTrap = this.getGameRules().get(GameRules.SPAWN_MOBS)
                     && this.random.nextDouble() < difficulty.getEffectiveDifficulty() * this.paperConfig().entities.spawning.skeletonHorseThunderSpawnChance.or(0.01) // Paper - Configurable spawn chances for skeleton horses
                     && !this.getBlockState(pos.below()).is(BlockTags.LIGHTNING_RODS);
+
+                // Pluto start - Async spawning chunks
+                if (this.plutoConfig().useAsyncSpawningChunks) {
+                    io.papermc.paper.util.MCUtil.ensureMain(() -> {
+                        if (isTrap) {
+                            SkeletonHorse horse = EntityType.SKELETON_HORSE.create(this, EntitySpawnReason.EVENT);
+                            if (horse != null) {
+                                horse.setTrap(true);
+                                horse.setAge(0);
+                                horse.setPos(pos.getX(), pos.getY(), pos.getZ());
+                                this.addFreshEntity(horse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
+                            }
+                        }
+
+                        LightningBolt bolt = EntityType.LIGHTNING_BOLT.create(this, EntitySpawnReason.EVENT);
+                        if (bolt != null) {
+                            bolt.snapTo(Vec3.atBottomCenterOf(pos));
+                            bolt.setVisualOnly(isTrap);
+                            this.strikeLightning(bolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.WEATHER); // CraftBukkit
+                        }
+                    });
+                    return;
+                }
+                // Pluto end - Async spawning chunks
+
                 if (isTrap) {
                     SkeletonHorse horse = EntityType.SKELETON_HORSE.create(this, EntitySpawnReason.EVENT);
                     if (horse != null) {
@@ -1111,7 +_,10 @@
     public @Nullable BlockPos findLightningTargetAround(final BlockPos pos, boolean returnNullWhenNoTarget) {
         // Paper end - Add methods to find targets for lightning strikes
         BlockPos center = this.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, pos);
-        Optional<BlockPos> lightningRodTarget = this.findLightningRod(center);
+        // Pluto start - Async Spawning Chunks
+        final BlockPos centerFinal = center;
+        Optional<BlockPos> lightningRodTarget = this.plutoConfig().useAsyncSpawningChunks && !ca.spottedleaf.moonrise.common.util.TickThread.isTickThread() ? dev.yive.pluto.async.AsyncUtils.completeOnMain(() -> this.findLightningRod(centerFinal)).join() : this.findLightningRod(center);
+        // Pluto end - Async Spawning Chunks
         if (lightningRodTarget.isPresent()) {
             return lightningRodTarget.get();
         } else {
