--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -817,6 +_,16 @@
                 super.tick();
             }
 
+            // Pluto start - Only tick inventory slots that are important
+            // This drops the ticked inventory slots from 41 -> 2.
+            if (this.level().plutoConfig().playerInventoryOnlyUpdateImportantSlots) { // FIXME: If mojang expects this to handle non-map items.
+                ItemStack mainHand = this.getInventory().getSelectedItem();
+                if (!mainHand.isEmpty()) synchronizeSpecialItemUpdates(mainHand);
+
+                ItemStack offHand = this.getInventory().equipment.get(EquipmentSlot.OFFHAND);
+                if (!offHand.isEmpty()) synchronizeSpecialItemUpdates(offHand);
+            } else
+            // Pluto end - Only tick inventory slots that are important
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack item = this.getInventory().getItem(i);
                 if (!item.isEmpty()) {
