--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -419,6 +_,10 @@
     // Paper end - cancellable death event
     // Paper start - Optional per player mob spawns
     public static final int MOBCATEGORY_TOTAL_ENUMS = net.minecraft.world.entity.MobCategory.values().length;
+    // Pluto start - Async Spawning Chunks
+    public final java.util.Map<Integer, Integer> mobCountsAsync = new java.util.concurrent.ConcurrentHashMap<>(MOBCATEGORY_TOTAL_ENUMS);
+    public final java.util.Map<Integer, Integer> mobBackoffCountsAsync = new java.util.concurrent.ConcurrentHashMap<>(MOBCATEGORY_TOTAL_ENUMS);
+    // Pluto end - Async Spawning Chunks
     public final int[] mobCounts = new int[MOBCATEGORY_TOTAL_ENUMS];
     // Paper end - Optional per player mob spawns
     public final int[] mobBackoffCounts = new int[MOBCATEGORY_TOTAL_ENUMS]; // Paper - per player mob count backoff
@@ -831,6 +_,16 @@
                 }
             }
 
+            // Pluto start - Only tick inventory slots that are important
+            // This drops the ticked inventory slots from 41 -> 2.
+            if (this.level().plutoConfig().playerInventoryOnlyUpdateImportantSlots) { // FIXME: If mojang expects this to handle non-map items.
+                ItemStack mainHand = this.getInventory().getSelectedItem();
+                if (!mainHand.isEmpty()) synchronizeSpecialItemUpdates(mainHand);
+
+                ItemStack offHand = this.getInventory().equipment.get(EquipmentSlot.OFFHAND);
+                if (!offHand.isEmpty()) synchronizeSpecialItemUpdates(offHand);
+            } else
+            // Pluto end - Only tick inventory slots that are important
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack itemStack = this.getInventory().getItem(i);
                 if (!itemStack.isEmpty()) {
@@ -1473,7 +_,7 @@
             }
 
             boolean freeBottom = block.isPossibleToRespawnInThis(blockState);
-            BlockState topState = level.getBlockState(pos.above());
+            BlockState topState = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
             boolean freeTop = topState.getBlock().isPossibleToRespawnInThis(topState);
             return freeBottom && freeTop
                 ? Optional.of(new ServerPlayer.RespawnPosAngle(new Vec3(pos.getX() + 0.5, pos.getY() + 0.1, pos.getZ() + 0.5), yaw, pitch, false, false, null)) // Paper - Fix SPIGOT-5989
