--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -198,6 +_,50 @@
     }
     // Paper end
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return this.getChunk(x >> 4, z >> 4).getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(final long packedPos) {
+        return pluto$getBlockState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.getBlockState(x, y, z);
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(final long packedPos) {
+        return pluto$getBlockStateIfLoaded(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        return this.getChunk(x >> 4, z >> 4).pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final long packedPos) {
+        return pluto$getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(final int x, final int y, final int z) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(final long packedPos) {
+        return pluto$getFluidStateIfLoaded(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         return this.getChunk(SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getZ())).getBlockState(pos);
