--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -205,6 +_,24 @@
     }
     // Paper end
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return this.getChunk(x >> 4, z >> 4).getBlockState(x, y, z);
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.getBlockState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        return this.getChunk(x >> 4, z >> 4).pluto$getFluidState(x, y, z);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         int chunkX = SectionPos.blockToSectionCoord(pos.getX());
