--- a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
@@ -40,7 +_,7 @@
 
     @Override
     public boolean isStableDestination(final BlockPos pos) {
-        return !this.level.getBlockState(pos.below()).isAir();
+        return !this.level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()).isAir(); // Pluto - Memory allocation optimisations
     }
 
     @Override
