--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -288,8 +_,7 @@
     }
 
     protected double getGroundY(final Vec3 target) {
-        BlockPos blockPos = BlockPos.containing(target);
-        return this.level.getBlockState(blockPos.below()).isAir() ? target.y : WalkNodeEvaluator.getFloorLevel(this.level, blockPos);
+        return this.level.pluto$getBlockState(Mth.floor(target.x), Mth.floor(target.y) - 1, Mth.floor(target.z)).isAir() ? target.y : WalkNodeEvaluator.getFloorLevel(this.level, BlockPos.containing(target)); // Pluto - Memory allocation optimisations
     }
 
     protected void followThePath() {
@@ -402,7 +_,7 @@
             for (int i = 0; i < this.path.getNodeCount(); i++) {
                 Node node = this.path.getNode(i);
                 Node nextNode = i + 1 < this.path.getNodeCount() ? this.path.getNode(i + 1) : null;
-                BlockState state = this.level.getBlockState(new BlockPos(node.x, node.y, node.z));
+                BlockState state = this.level.pluto$getBlockState(node.x, node.y, node.z); // Pluto - Memory allocation optimisations
                 if (state.is(BlockTags.CAULDRONS)) {
                     this.path.replaceNode(i, node.cloneAndMove(node.x, node.y + 1, node.z));
                     if (nextNode != null && node.y >= nextNode.y) {
@@ -430,8 +_,7 @@
     }
 
     public boolean isStableDestination(final BlockPos pos) {
-        BlockPos below = pos.below();
-        return this.level.getBlockState(below).isSolidRender();
+        return this.level.pluto$getBlockState(pos.getX(), pos.getY() - 1, pos.getZ()).isSolidRender(); // Pluto - Memory allocation optimisations
     }
 
     public NodeEvaluator getNodeEvaluator() {
