--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -199,7 +_,7 @@
             }
 
             int maxStackSize = furnace.getMaxStackSize();
-            if (!furnace.isLit() && canBurn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize)) {
+            if (!furnace.isLit() && canBurn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, furnace, maxStackSize)) { // Pluto - Cache canBurn result in Furnaces
                 // CraftBukkit start
                 org.bukkit.craftbukkit.inventory.CraftItemStack fuel = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack);
 
@@ -225,7 +_,7 @@
                 }
             }
 
-            if (furnace.isLit() && canBurn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize)) {
+            if (furnace.isLit() && canBurn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, furnace, maxStackSize)) { // Pluto - Cache canBurn result in Furnaces
                 // CraftBukkit start
                 if (recipeHolder != null && furnace.cookingTimer == 0) {
                     org.bukkit.craftbukkit.inventory.CraftItemStack source = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(furnace.items.get(0));
@@ -247,7 +_,7 @@
                 if (furnace.cookingTimer >= furnace.cookingTotalTime) { // Paper - cook speed multiplier API
                     furnace.cookingTimer = 0;
                     furnace.cookingTotalTime = getTotalCookTime(level, furnace, furnace.recipeType, furnace.cookSpeedMultiplier); // Paper - cook speed multiplier API
-                    if (burn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize, level, furnace.worldPosition)) { // CraftBukkit
+                    if (burn(level.registryAccess(), recipeHolder, singleRecipeInput, furnace.items, maxStackSize, level, furnace, furnace.worldPosition)) { // CraftBukkit // Pluto - Cache canBurn result in Furnaces
                         furnace.setRecipeUsed(recipeHolder);
                     }
 
@@ -271,27 +_,54 @@
         }
     }
 
+    // Pluto start - Cache canBurn result in Furnaces
+    protected net.minecraft.util.TriState canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        this.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    }
+
+    @Override
+    public void setBlockState(BlockState blockState) {
+        super.setBlockState(blockState);
+        this.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    }
+    // Pluto end - Cache canBurn result in Furnaces
+
     private static boolean canBurn(
         RegistryAccess registryAccess,
         @Nullable RecipeHolder<? extends AbstractCookingRecipe> recipe,
         SingleRecipeInput recipeInput,
         NonNullList<ItemStack> items,
+        AbstractFurnaceBlockEntity furnace, // Pluto - Cache canBurn result in Furnaces
         int maxStackSize
     ) {
+        // Pluto start - Cache canBurn result in Furnaces
+        if (furnace.canBurnCache != net.minecraft.util.TriState.DEFAULT) {
+            return furnace.canBurnCache == net.minecraft.util.TriState.TRUE;
+        }
+        // Pluto end - Cache canBurn result in Furnaces
         if (!items.get(0).isEmpty() && recipe != null) {
             ItemStack itemStack = recipe.value().assemble(recipeInput, registryAccess);
             if (itemStack.isEmpty()) {
+                furnace.canBurnCache = net.minecraft.util.TriState.FALSE; // Pluto - Cache canBurn result in Furnaces
                 return false;
             } else {
                 ItemStack itemStack1 = items.get(2);
-                return itemStack1.isEmpty()
+                boolean result = itemStack1.isEmpty() // Pluto - Cache canBurn result in Furnaces
                     || ItemStack.isSameItemSameComponents(itemStack1, itemStack)
                         && (
                             itemStack1.getCount() < maxStackSize && itemStack1.getCount() < itemStack1.getMaxStackSize()
                                 || itemStack1.getCount() < itemStack.getMaxStackSize()
                         );
+                // Pluto start - Cache canBurn result in Furnaces
+                furnace.canBurnCache = result ? net.minecraft.util.TriState.TRUE : net.minecraft.util.TriState.FALSE;
+                return result;
+                // Pluto end - Cache canBurn result in Furnaces
             }
         } else {
+            furnace.canBurnCache = net.minecraft.util.TriState.FALSE; // Pluto - Cache canBurn result in Furnaces
             return false;
         }
     }
@@ -303,9 +_,13 @@
         NonNullList<ItemStack> items,
         int maxStackSize,
         net.minecraft.world.level.Level level, // CraftBukkit
+        AbstractFurnaceBlockEntity furnace, // Pluto - Cache canBurn result in Furnaces
         BlockPos blockPos // CraftBukkit
     ) {
-        if (recipe != null && canBurn(registryAccess, recipe, recipeInput, items, maxStackSize)) {
+        // Pluto start - Cache canBurn result in Furnaces
+        if (recipe != null && canBurn(registryAccess, recipe, recipeInput, items, furnace, maxStackSize)) {
+            furnace.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+            // Pluto end - Cache canBurn result in Furnaces
             ItemStack itemStack = items.get(0); final ItemStack ingredient = itemStack; // Paper - OBFHELPER
             ItemStack itemStack1 = recipe.value().assemble(recipeInput, registryAccess); ItemStack result = itemStack1; // Paper - OBFHELPER
             ItemStack itemStack2 = items.get(2); final ItemStack existingResults = itemStack2; // Paper - OBFHELPER
