--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -200,7 +_,7 @@
                 if (recipe != null) {
                     int maxStackSize = entity.getMaxStackSize();
                     ItemStack burnResult = recipe.value().assemble(input);
-                    if (!burnResult.isEmpty() && canBurn(entity.items, maxStackSize, burnResult)) {
+                    if (!burnResult.isEmpty() && canBurn(entity.items, maxStackSize, burnResult, entity)) { // Pluto - Cache canBurn result in Furnaces
                         // CraftBukkit start
                         org.bukkit.event.inventory.FurnaceBurnEvent furnaceBurnEvent = new org.bukkit.event.inventory.FurnaceBurnEvent(
                             org.bukkit.craftbukkit.block.CraftBlock.at(level, pos),
@@ -240,7 +_,7 @@
                             if (entity.cookingTimer >= entity.cookingTotalTime) { // Paper - cook speed multiplier API
                                 entity.cookingTimer = 0;
                                 entity.cookingTotalTime = getTotalCookTime(level, entity, entity.recipeType, entity.cookSpeedMultiplier); // Paper - cook speed multiplier API
-                                if (burn(entity.items, ingredient, burnResult, recipe, level, entity.worldPosition)) { // CraftBukkit - add level & pos // Paper - make burn return a boolean again, add recipe
+                                if (burn(entity.items, ingredient, burnResult, recipe, level, entity.worldPosition, entity)) { // CraftBukkit - add level & pos // Paper - make burn return a boolean again, add recipe // Pluto - Cache canBurn result in Furnaces
                                 entity.setRecipeUsed(recipe);
                                 } // Paper
                                 changed = true;
@@ -279,22 +_,49 @@
         }
     }
 
-    private static boolean canBurn(final NonNullList<ItemStack> items, final int maxStackSize, final ItemStack burnResult) {
+    // Pluto start - Cache canBurn result in Furnaces
+    protected net.minecraft.util.TriState canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        this.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    }
+
+    @Override
+    public void setBlockState(BlockState blockState) {
+        super.setBlockState(blockState);
+        this.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+    }
+
+    private static boolean canBurn(final NonNullList<ItemStack> items, final int maxStackSize, final ItemStack burnResult, final AbstractFurnaceBlockEntity furnace) {
+        if (furnace.canBurnCache != net.minecraft.util.TriState.DEFAULT) {
+            return furnace.canBurnCache == net.minecraft.util.TriState.TRUE;
+        }
+        // Pluto end - Cache canBurn result in Furnaces
         ItemStack resultItemStack = items.get(2);
         if (resultItemStack.isEmpty()) {
+            furnace.canBurnCache = net.minecraft.util.TriState.TRUE; // Pluto - Cache canBurn result in Furnaces
             return true;
         }
 
         if (!ItemStack.isSameItemSameComponents(resultItemStack, burnResult)) {
+            furnace.canBurnCache = net.minecraft.util.TriState.FALSE; // Pluto - Cache canBurn result in Furnaces
             return false;
         }
 
         int resultCount = resultItemStack.getCount() + burnResult.count();
         int maxResultCount = Math.min(maxStackSize, burnResult.getMaxStackSize());
-        return resultCount <= maxResultCount;
+        // Pluto start - Cache canBurn result in Furnaces
+        boolean result = resultCount <= maxResultCount;
+        furnace.canBurnCache = net.minecraft.util.TriState.from(result);
+        return result;
+        // Pluto end - Cache canBurn result in Furnaces
     }
 
-    private static boolean burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, ItemStack result, RecipeHolder<? extends AbstractCookingRecipe> recipe, final net.minecraft.world.level.Level level, final BlockPos blockPos) { // CraftBukkit
+    // Pluto start - Cache canBurn result in Furnaces
+    private static boolean burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, ItemStack result, RecipeHolder<? extends AbstractCookingRecipe> recipe, final net.minecraft.world.level.Level level, final BlockPos blockPos, final AbstractFurnaceBlockEntity furnace) { // CraftBukkit
+        furnace.canBurnCache = net.minecraft.util.TriState.DEFAULT;
+        // Pluto end - Cache canBurn result in Furnaces
         ItemStack resultItemStack = items.get(2);
         // CraftBukkit start - fire FurnaceSmeltEvent
         org.bukkit.craftbukkit.inventory.CraftItemStack apiIngredient = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(inputItemStack);
