--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -154,6 +_,8 @@
         return Container.stillValidBlockEntity(this, player);
     }
 
+    @Override public boolean stillValid() { return !isRemoved(); } // Pluto - Hoppers cache containers
+
     @Override
     public void clearContent() {
         this.getItems().clear();
