--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -143,6 +_,21 @@
         this.items = items;
     }
 
+    // Pluto start - Cache isBrewable result in Brewing Stands
+    protected net.minecraft.util.TriState brewableCache = net.minecraft.util.TriState.DEFAULT;
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        this.brewableCache = net.minecraft.util.TriState.DEFAULT;
+    }
+
+    @Override
+    public void setBlockState(BlockState blockState) {
+        super.setBlockState(blockState);
+        this.brewableCache = net.minecraft.util.TriState.DEFAULT;
+    }
+    // Pluto end - Cache isBrewable result in Brewing Stands
+
     public static void serverTick(final Level level, final BlockPos pos, final BlockState selfState, final BrewingStandBlockEntity entity) {
         ItemStack fuel = entity.items.get(4);
         if (entity.fuel <= 0 && fuel.is(ItemTags.BREWING_FUEL)) {
@@ -164,7 +_,15 @@
             setChanged(level, pos, selfState);
         }
 
-        boolean brewable = isBrewable(level.potionBrewing(), entity.items);
+        // Pluto start - Cache isBrewable result in Brewing Stands
+        boolean brewable;
+        if (entity.brewableCache == net.minecraft.util.TriState.DEFAULT) {
+            brewable = isBrewable(level.potionBrewing(), entity.items);
+            entity.brewableCache = brewable ? net.minecraft.util.TriState.TRUE : net.minecraft.util.TriState.FALSE;
+        } else {
+            brewable = entity.brewableCache == net.minecraft.util.TriState.TRUE;
+        }
+        // Pluto end - Cache isBrewable result in Brewing Stands
         boolean isBrewing = entity.brewTime > 0;
         ItemStack ingredient = entity.items.get(3);
         if (isBrewing) {
@@ -238,11 +_,12 @@
     }
 
     private static void doBrew(final Level level, final BlockPos pos, final NonNullList<ItemStack> items, final BrewingStandBlockEntity entity) { // CraftBukkit
+        entity.brewableCache = net.minecraft.util.TriState.DEFAULT; // Pluto - Cache isBrewable result in Brewing Stands
         ItemStack ingredient = items.get(3);
         PotionBrewing potionBrewing = level.potionBrewing();
 
         // CraftBukkit start
-        org.bukkit.inventory.InventoryHolder owner = entity.getOwner();
+        org.bukkit.inventory.InventoryHolder owner = entity.getOwner(level.plutoConfig().brewingStandDoSnapshotForSelf); // Pluto - Allow toggling snapshots for certain blocks
         java.util.List<org.bukkit.inventory.ItemStack> brewResults = new java.util.ArrayList<>(3);
         for (int dest = 0; dest < 3; dest++) {
             brewResults.add(dest, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(potionBrewing.mix(ingredient, items.get(dest))));
