--- a/net/minecraft/world/level/block/entity/ChestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChestBlockEntity.java
@@ -212,6 +_,20 @@
             : 0;
     }
 
+    // Pluto start - Memory allocation optimisations
+    public static int pluto$getOpenCount(final BlockGetter level, final int x, final int y, final int z) {
+        BlockState state = level.pluto$getBlockState(x, y, z);
+        if (state.hasBlockEntity()) {
+            BlockEntity blockEntity = level.getBlockEntity(new BlockPos(x, y, z));
+            if (blockEntity instanceof ChestBlockEntity) {
+                return ((ChestBlockEntity)blockEntity).openersCounter.getOpenerCount();
+            }
+        }
+
+        return 0;
+    }
+    // Pluto end - Memory allocation optimisations
+
     public static void swapContents(final ChestBlockEntity one, final ChestBlockEntity two) {
         NonNullList<ItemStack> items = one.getItems();
         one.setItems(two.getItems());
