--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -123,6 +_,8 @@
         return retrievedItem;
     }
 
+    @Override public boolean stillValid() { return !isRemoved(); } // Pluto - Hoppers cache containers
+
     @Override
     public void setItem(final int slot, final ItemStack itemStack) {
         if (this.acceptsItemType(itemStack)) {
