--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -123,6 +_,8 @@
         return itemStack;
     }
 
+    @Override public boolean stillValid() { return !isRemoved(); } // Pluto - Hoppers cache containers
+
     @Override
     public void setItem(int slot, ItemStack stack) {
         if (this.acceptsItemType(stack)) {
