--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -170,6 +_,8 @@
         return Container.stillValidBlockEntity(this, player);
     }
 
+    @Override public boolean stillValid() { return !isRemoved(); } // Pluto - Hoppers cache containers
+
     @Override
     public boolean canPlaceItem(int slot, ItemStack stack) {
         return stack.is(ItemTags.BOOKSHELF_BOOKS) && this.getItem(slot).isEmpty() && stack.getCount() == this.getMaxStackSize();
