--- a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
@@ -90,9 +_,11 @@
                     Vec3 heartPosition = Vec3.atCenterOf(pos);
                     float progress = 0.2F + 0.8F * (100 - entity.emitter) / 100.0F;
                     Vec3 soundLocation = heartPosition.subtract(entity.emitterTarget).scale(progress).add(entity.emitterTarget);
-                    BlockPos soundPos = BlockPos.containing(soundLocation);
+                    // Pluto start - Memory allocation optimisations
+                    long soundPos = soundLocation.asBlockPosLong();
                     float volume = entity.emitter / 2.0F / 100.0F + 0.5F;
-                    serverLevel.playSound(null, soundPos, SoundEvents.CREAKING_HEART_HURT, SoundSource.BLOCKS, volume, 1.0F);
+                    serverLevel.playSound(null, BlockPos.getX(soundPos) + 0.5, BlockPos.getY(soundPos) + 0.5, BlockPos.getZ(soundPos) + 0.5, SoundEvents.CREAKING_HEART_HURT, SoundSource.BLOCKS, volume, 1.0F);
+                    // Pluto end - Memory allocation optimisations
                 }
 
                 entity.emitter--;
