--- a/net/minecraft/world/level/block/entity/DispenserBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DispenserBlockEntity.java
@@ -62,6 +_,14 @@
 
     public int getRandomSlot(RandomSource random) {
         this.unpackLootTable(null);
+        // Pluto start - Less random dispensing
+        if (this.level != null && this.level.plutoConfig().lessRandomDispensing) {
+            for (int i = 0; i < this.items.size(); i++) {
+                if (!this.items.get(i).isEmpty()) return i;
+            }
+            return -1;
+        }
+        // Pluto end - Less random dispensing
         int i = -1;
         int i1 = 1;
 
