--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -39,6 +_,11 @@
     private long tickedGameTime;
     private Direction facing;
 
+    // Pluto start - Hoppers cache containers
+    private java.lang.ref.WeakReference<Container> pushContainer = new java.lang.ref.WeakReference<>(null);
+    private java.lang.ref.WeakReference<Container> pullContainer = new java.lang.ref.WeakReference<>(null);
+    // Pluto end - Hoppers cache containers
+
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
     private int maxStack = MAX_STACK;
@@ -121,6 +_,10 @@
     public void setBlockState(final BlockState blockState) {
         super.setBlockState(blockState);
         this.facing = blockState.getValue(HopperBlock.FACING);
+        // Pluto start - Hoppers cache containers
+        this.pushContainer = new java.lang.ref.WeakReference<>(null);
+        this.pullContainer = new java.lang.ref.WeakReference<>(null);
+        // Pluto end - Hoppers cache containers
     }
 
     @Override
@@ -179,9 +_,11 @@
     }
     // Paper end - Perf: Optimize Hoppers
 
+    private boolean canAcceptItems = false; public boolean canAcceptItems() { return canAcceptItems; } // Pluto - Entities search for hoppers
     private static boolean tryMoveItems(
         final Level level, final BlockPos pos, final BlockState state, final HopperBlockEntity entity, final BooleanSupplier action
     ) {
+        entity.canAcceptItems = false; // Pluto - Entities search for hoppers
         if (level.isClientSide()) {
             return false;
         }
@@ -194,6 +_,7 @@
             }
 
             if (changed || fullState != HOPPER_IS_FULL) { // Paper - Perf: Optimize Hoppers
+                entity.canAcceptItems = true; // Pluto - Entities search for hoppers
                 changed |= action.getAsBoolean();
             }
 
@@ -418,7 +_,18 @@
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(final Level level, final BlockPos blockPos, final HopperBlockEntity self) {
-        Container container = getAttachedContainer(level, blockPos, self);
+        // Pluto start - Hoppers cache containers
+        final Container cached = self.pushContainer.get();
+        Container container = null;
+        if (cached != null) {
+            container = cached.stillValid() ? cached : null;
+        }
+
+        if (container == null) {
+            container = getAttachedContainer(level, blockPos, self);
+            self.pushContainer = new java.lang.ref.WeakReference<>(container);
+        }
+        // Pluto end - Hoppers cache containers
         if (container == null) {
             return false;
         }
@@ -475,10 +_,44 @@
         return true;
     }
 
+    // Pluto start - Hoppers cache containers
+    public static boolean suckInItems(Level level, HopperBlockEntity hopper) {
+        BlockPos blockPos = null;
+        BlockState blockState = null;
+        final Container cached = hopper.pullContainer.get();
+        Container sourceContainer = null;
+        if (cached != null) {
+            sourceContainer = cached.stillValid() ? cached : null;
+        }
+
+        if (sourceContainer == null) {
+            sourceContainer = getSourceContainer(level, hopper, blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ()), blockState = level.getBlockState(blockPos));
+            hopper.pullContainer = new java.lang.ref.WeakReference<>(sourceContainer);
+        }
+        return suckInItems(level, hopper, sourceContainer, net.minecraft.util.Mth.floor(hopper.getLevelX()), net.minecraft.util.Mth.floor(hopper.getLevelY() + 1.0), net.minecraft.util.Mth.floor(hopper.getLevelZ()), blockState);
+    }
+    // Pluto end - Hoppers cache containers
+
     public static boolean suckInItems(final Level level, final Hopper hopper) {
-        BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
-        BlockState blockState = level.getBlockState(blockPos);
-        Container container = getSourceContainer(level, hopper, blockPos, blockState);
+        // Pluto start - Entities search for hoppers
+        return suckInItems(level, hopper, null);
+    }
+
+    public static boolean suckInItems(Level level, Hopper hopper, @Nullable Container sourceContainer) {
+        // Pluto end - Entities search for hoppers
+        // Pluto start - Memory allocation optimisations
+        final int x = net.minecraft.util.Mth.floor(hopper.getLevelX());
+        final int y = net.minecraft.util.Mth.floor(hopper.getLevelY() + 1.0D);
+        final int z = net.minecraft.util.Mth.floor(hopper.getLevelZ());
+        BlockState blockState = level.pluto$getBlockState(x, y, z);
+        if (sourceContainer == null) sourceContainer = getSourceContainer(level, hopper, new BlockPos(x, y, z), blockState); // Pluto - Entities search for hoppers
+        // Pluto end - Memory allocation optimisations
+        // Pluto start - Hoppers cache containers
+        return suckInItems(level, hopper, sourceContainer, x, y, z, blockState);
+    }
+
+    public static boolean suckInItems(Level level, Hopper hopper, @Nullable Container container, int x, int y, int z, BlockState blockState) {
+        // Pluto end - Hoppers cache containers
         if (container != null) {
             Direction direction = Direction.DOWN;
             skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
@@ -491,8 +_,9 @@
 
             return false;
         } else {
+            if (level.plutoConfig().entitiesSearchForHoppers && !(hopper instanceof net.minecraft.world.entity.vehicle.minecart.MinecartHopper)) return false; // Pluto - Entities search for hoppers | Avoid doing collision check
             boolean isBlocked = hopper.isGridAligned()
-                && blockState.isCollisionShapeFullBlock(level, blockPos)
+                && (blockState == null ? blockState = level.pluto$getBlockState(x, y, z) : blockState).isCollisionShapeFullBlock(level, new BlockPos(x, y, z)) // Pluto - Memory allocation optimisations // Pluto - Hoppers cache containers
                 && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
             if (!isBlocked) {
                 for (ItemEntity entity : getItemsAtAndAbove(level, hopper)) {
@@ -652,7 +_,7 @@
 
     private static @Nullable Container getSourceContainer(final Level level, final Hopper hopper, final BlockPos pos, final BlockState state) {
         // Paper start
-        final Container inventory = HopperBlockEntity.getContainerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
+        final Container inventory = HopperBlockEntity.getContainerAt(level, hopper, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0D, hopper.getLevelZ(), false); // Pluto - Entities search for hoppers
         if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory;
 
         final BlockPos hopperPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY(), hopper.getLevelZ());
@@ -673,34 +_,37 @@
     }
 
     public static @Nullable Container getContainerAt(final Level level, final BlockPos pos) {
-        return getContainerAt(level, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, true); // Paper - Optimize hoppers
+        return getContainerAt(level, null, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, true); // Paper - Optimize hoppers // Pluto - Entities search for hoppers
     }
 
     private static @Nullable Container getContainerAt(
         final Level level, final BlockPos pos, final BlockState state, final double x, final double y, final double z
     ) {
         // Paper start - Perf: Optimize Hoppers
-        return getContainerAt(level, pos, state, x, y, z, false);
+        return HopperBlockEntity.getContainerAt(level, null, pos, state, x, y, z, false); // Pluto - Entities search for hoppers
     }
-    private static @Nullable Container getContainerAt(
-        final Level level, final BlockPos pos, final BlockState state, final double x, final double y, final double z,
-        final boolean optimizeEntities
-    ) {
+    private static @Nullable Container getContainerAt(Level level, Hopper hopper, BlockPos pos, BlockState state, double x, double y, double z, final boolean optimizeEntities) { // Pluto - Entities search for hoppers
         // Paper end - Perf: Optimize Hoppers
         Container result = getBlockContainer(level, pos, state);
         if (result == null && (!optimizeEntities || !level.paperConfig().hopper.ignoreOccludingBlocks || !state.getBukkitMaterial().isOccluding())) { // Paper - Perf: Optimize Hoppers
-            result = getEntityContainer(level, x, y, z);
+            result = getEntityContainer(level, hopper, x, y, z); // Pluto - Entities search for hoppers
         }
 
         return result;
     }
 
     private static @Nullable Container getBlockContainer(final Level level, final BlockPos pos, final BlockState state) {
-        if (!level.spigotConfig.hopperCanLoadChunks && !level.hasChunkAt(pos)) return null; // Spigot
+        // Pluto start - Small hopper optimisations
+        net.minecraft.world.level.chunk.LevelChunk chunk = null;
+        if (!level.spigotConfig.hopperCanLoadChunks) {
+            chunk = level.getChunkIfLoaded(pos);
+            if (chunk == null) return null;
+        }
+        // Pluto end - Small hopper optimisations
         Block block = state.getBlock();
         if (block instanceof WorldlyContainerHolder worldlyContainerHolder) {
             return worldlyContainerHolder.getContainer(state, level, pos);
-        } else if (state.hasBlockEntity() && level.getBlockEntity(pos) instanceof Container container) {
+        } else if (state.hasBlockEntity() && (chunk == null ? level.getBlockEntity(pos) : chunk.getBlockEntity(pos, net.minecraft.world.level.chunk.LevelChunk.EntityCreationType.IMMEDIATE)) instanceof Container container) { // Pluto - Small hopper optimisations
             if (container instanceof ChestBlockEntity && block instanceof ChestBlock chestBlock) {
                 container = ChestBlock.getContainer(chestBlock, state, level, pos, true);
             }
@@ -711,7 +_,10 @@
         }
     }
 
-    private static @Nullable Container getEntityContainer(final Level level, final double x, final double y, final double z) {
+    // Pluto start - Entities search for hoppers
+    private static @Nullable Container getEntityContainer(Level level, Hopper hopper, double x, double y, double z) {
+        if (level.plutoConfig().entitiesSearchForHoppers && !(hopper instanceof  net.minecraft.world.entity.vehicle.minecart.MinecartHopper)) return null;
+        // Pluto end - Entities search for hoppers
         List<Entity> entities = level.getEntitiesOfClass( // Paper - Perf: Optimize hoppers
             (Class) Container.class, new AABB(x - 0.5, y - 0.5, z - 0.5, x + 0.5, y + 0.5, z + 0.5), EntitySelector.CONTAINER_ENTITY_SELECTOR // Paper - Perf: Optimize hoppers
         );
