--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -39,6 +_,11 @@
     private long tickedGameTime;
     private Direction facing;
 
+    // Pluto start - Hoppers cache containers
+    private java.lang.ref.WeakReference<Container> pushContainer = new java.lang.ref.WeakReference<>(null);
+    private java.lang.ref.WeakReference<Container> pullContainer = new java.lang.ref.WeakReference<>(null);
+    // Pluto end - Hoppers cache containers
+
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
     private int maxStack = MAX_STACK;
@@ -121,6 +_,10 @@
     public void setBlockState(BlockState blockState) {
         super.setBlockState(blockState);
         this.facing = blockState.getValue(HopperBlock.FACING);
+        // Pluto start - Hoppers cache containers
+        this.pushContainer = new java.lang.ref.WeakReference<>(null);
+        this.pullContainer = new java.lang.ref.WeakReference<>(null);
+        // Pluto end - Hoppers cache containers
     }
 
     @Override
@@ -181,7 +_,9 @@
     }
     // Paper end - Perf: Optimize Hoppers
 
+    private boolean canAcceptItems = false; public boolean canAcceptItems() { return canAcceptItems; } // Pluto - Entities search for hoppers
     private static boolean tryMoveItems(Level level, BlockPos pos, BlockState state, HopperBlockEntity blockEntity, BooleanSupplier validator) {
+        blockEntity.canAcceptItems = false; // Pluto - Entities search for hoppers
         if (level.isClientSide()) {
             return false;
         } else {
@@ -193,6 +_,7 @@
                 }
 
                 if (fullState != HOPPER_IS_FULL || flag) { // Paper - Perf: Optimize Hoppers
+                    blockEntity.canAcceptItems = true; // Pluto - Entities search for hoppers
                     flag |= validator.getAsBoolean(); // Paper - note: this is not a validator, it's what adds/sucks in items
                 }
 
@@ -418,7 +_,18 @@
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(Level level, BlockPos pos, HopperBlockEntity blockEntity) {
-        Container attachedContainer = getAttachedContainer(level, pos, blockEntity);
+        // Pluto start - Hoppers cache containers
+        final Container cached = blockEntity.pushContainer.get();
+        Container attachedContainer = null;
+        if (cached != null) {
+            attachedContainer = cached.stillValid() ? cached : null;
+        }
+
+        if (attachedContainer == null) {
+            attachedContainer = getAttachedContainer(level, pos, blockEntity);
+            blockEntity.pushContainer = new java.lang.ref.WeakReference<>(attachedContainer);
+        }
+        // Pluto end - Hoppers cache containers
         if (attachedContainer == null) {
             return false;
         } else {
@@ -528,10 +_,44 @@
         return true;
     }
 
+    // Pluto start - Hoppers cache containers
+    public static boolean suckInItems(Level level, HopperBlockEntity hopper) {
+        BlockPos blockPos = null;
+        BlockState blockState = null;
+        final Container cached = hopper.pullContainer.get();
+        Container sourceContainer = null;
+        if (cached != null) {
+            sourceContainer = cached.stillValid() ? cached : null;
+        }
+
+        if (sourceContainer == null) {
+            sourceContainer = getSourceContainer(level, hopper, blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ()), blockState = level.getBlockState(blockPos));
+            hopper.pullContainer = new java.lang.ref.WeakReference<>(sourceContainer);
+        }
+        return suckInItems(level, hopper, sourceContainer, net.minecraft.util.Mth.floor(hopper.getLevelX()), net.minecraft.util.Mth.floor(hopper.getLevelY() + 1.0), net.minecraft.util.Mth.floor(hopper.getLevelZ()), blockState);
+    }
+    // Pluto end - Hoppers cache containers
+
     public static boolean suckInItems(Level level, Hopper hopper) {
-        BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
-        BlockState blockState = level.getBlockState(blockPos);
-        Container sourceContainer = getSourceContainer(level, hopper, blockPos, blockState);
+        // Pluto start - Entities search for hoppers
+        return suckInItems(level, hopper, null);
+    }
+
+    public static boolean suckInItems(Level level, Hopper hopper, @Nullable Container sourceContainer) {
+        // Pluto end - Entities search for hoppers
+        // Pluto start - Memory allocation optimisations
+        final int x = net.minecraft.util.Mth.floor(hopper.getLevelX());
+        final int y = net.minecraft.util.Mth.floor(hopper.getLevelY() + 1.0D);
+        final int z = net.minecraft.util.Mth.floor(hopper.getLevelZ());
+        BlockState blockState = level.pluto$getBlockState(x, y, z);
+        if (sourceContainer == null) sourceContainer = getSourceContainer(level, hopper, new BlockPos(x, y, z), blockState); // Pluto - Entities search for hoppers
+        // Pluto end - Memory allocation optimisations
+        // Pluto start - Hoppers cache containers
+        return suckInItems(level, hopper, sourceContainer, x, y, z, blockState);
+    }
+
+    public static boolean suckInItems(Level level, Hopper hopper, @Nullable Container sourceContainer, int x, int y, int z, BlockState blockState) {
+        // Pluto end - Hoppers cache containers
         if (sourceContainer != null) {
             Direction direction = Direction.DOWN;
             skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
@@ -544,7 +_,8 @@
 
             return false;
         } else {
-            boolean flag = hopper.isGridAligned() && blockState.isCollisionShapeFullBlock(level, blockPos) && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
+            if (level.plutoConfig().entitiesSearchForHoppers && !(hopper instanceof net.minecraft.world.entity.vehicle.minecart.MinecartHopper)) return false; // Pluto - Entities search for hoppers | Avoid doing collision check
+            boolean flag = hopper.isGridAligned() && (blockState == null ? blockState = level.pluto$getBlockState(x, y, z) : blockState).isCollisionShapeFullBlock(level, new BlockPos(x, y, z)) && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS); // Pluto - Memory allocation optimisations // Pluto - Hoppers cache containers
             if (!flag) {
                 for (ItemEntity itemEntity : getItemsAtAndAbove(level, hopper)) {
                     if (addItem(hopper, itemEntity)) {
@@ -750,7 +_,7 @@
 
     private static @Nullable Container getSourceContainer(Level level, Hopper hopper, BlockPos pos, BlockState state) {
         // Paper start
-        final Container inventory = HopperBlockEntity.getContainerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0D, hopper.getLevelZ());
+        final Container inventory = HopperBlockEntity.getContainerAt(level, hopper, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0D, hopper.getLevelZ(), false); // Pluto - Entities search for hoppers
         if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory;
 
         final BlockPos hopperPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY(), hopper.getLevelZ());
@@ -771,30 +_,36 @@
     }
 
     public static @Nullable Container getContainerAt(Level level, BlockPos pos) {
-        return getContainerAt(level, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, true); // Paper - Optimize hoppers
+        return getContainerAt(level, null, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, true); // Paper - Optimize hoppers // Pluto - Entities search for hoppers
     }
 
     @Nullable
     private static Container getContainerAt(Level level, BlockPos pos, BlockState state, double x, double y, double z) {
         // Paper start - Perf: Optimize Hoppers
-        return HopperBlockEntity.getContainerAt(level, pos, state, x, y, z, false);
+        return HopperBlockEntity.getContainerAt(level, null, pos, state, x, y, z, false); // Pluto - Entities search for hoppers
     }
-    private static @Nullable Container getContainerAt(Level level, BlockPos pos, BlockState state, double x, double y, double z, final boolean optimizeEntities) {
+    private static Container getContainerAt(Level level, Hopper hopper, BlockPos pos, BlockState state, double x, double y, double z, final boolean optimizeEntities) { // Pluto - Entities search for hoppers
         // Paper end - Perf: Optimize Hoppers
         Container blockContainer = getBlockContainer(level, pos, state);
         if (blockContainer == null && (!optimizeEntities || !level.paperConfig().hopper.ignoreOccludingBlocks || !state.getBukkitMaterial().isOccluding())) { // Paper - Perf: Optimize Hoppers
-            blockContainer = getEntityContainer(level, x, y, z);
+            blockContainer = getEntityContainer(level, hopper, x, y, z); // Pluto - Entities search for hoppers
         }
 
         return blockContainer;
     }
 
     private static @Nullable Container getBlockContainer(Level level, BlockPos pos, BlockState state) {
-        if (!level.spigotConfig.hopperCanLoadChunks && !level.hasChunkAt(pos)) return null; // Spigot
+        // Pluto start - Small hopper optimisations
+        net.minecraft.world.level.chunk.LevelChunk chunk = null;
+        if (!level.spigotConfig.hopperCanLoadChunks) {
+            chunk = level.getChunkIfLoaded(pos);
+            if (chunk == null) return null;
+        }
+        // Pluto end - Small hopper optimisations
         Block block = state.getBlock();
         if (block instanceof WorldlyContainerHolder) {
             return ((WorldlyContainerHolder)block).getContainer(state, level, pos);
-        } else if (state.hasBlockEntity() && level.getBlockEntity(pos) instanceof Container container) {
+        } else if (state.hasBlockEntity() && (chunk == null ? level.getBlockEntity(pos) : chunk.getBlockEntity(pos, net.minecraft.world.level.chunk.LevelChunk.EntityCreationType.IMMEDIATE)) instanceof Container container) { // Pluto - Small hopper optimisations
             if (container instanceof ChestBlockEntity && block instanceof ChestBlock) {
                 container = ChestBlock.getContainer((ChestBlock)block, state, level, pos, true);
             }
@@ -805,7 +_,10 @@
         }
     }
 
-    private static @Nullable Container getEntityContainer(Level level, double x, double y, double z) {
+    // Pluto start - Entities search for hoppers
+    private static Container getEntityContainer(Level level, Hopper hopper, double x, double y, double z) {
+        if (level.plutoConfig().entitiesSearchForHoppers && !(hopper instanceof  net.minecraft.world.entity.vehicle.minecart.MinecartHopper)) return null;
+        // Pluto end - Entities search for hoppers
         List<Entity> entities = level.getEntitiesOfClass(
             (Class) Container.class, new AABB(x - 0.5, y - 0.5, z - 0.5, x + 0.5, y + 0.5, z + 0.5), EntitySelector.CONTAINER_ENTITY_SELECTOR // Paper - Perf: Optimize hoppers
         );
