--- a/net/minecraft/world/level/block/entity/ShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShelfBlockEntity.java
@@ -121,6 +_,8 @@
         return Container.stillValidBlockEntity(this, player);
     }
 
+    @Override public boolean stillValid() { return Container.stillValidBlockEntity(this); } // Pluto - Hoppers cache containers
+
     public ItemStack swapItemNoUpdate(final int slot, final ItemStack heldItemStack) {
         ItemStack retrievedItem = this.removeItemNoUpdate(slot);
         this.setItemNoUpdate(slot, heldItemStack);
