--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -53,7 +_,7 @@
             level.addFreshEntity(crystal);
             level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, above);
             EnderDragonFight fight = ((ServerLevel)level).getDragonFight();
-            if (fight != null) {
+            if (fight != null && gg.pufferfish.pufferfish.PufferfishConfig.allowEndCrystalRespawn) { // Pufferfish
                 fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup
             }
         }
