--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -336,6 +_,12 @@
         return 0;
     }
 
+    // Pluto start - Memory allocation optimisations
+    protected int pluto$getSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     protected VoxelShape getShape(final BlockState state, final BlockGetter level, final BlockPos pos, final CollisionContext context) {
         return Shapes.block();
     }
@@ -396,6 +_,12 @@
         return Redstone.SIGNAL_MIN;
     }
 
+    // Pluto start - Memory allocation optimisations
+    protected int pluto$getDirectSignal(final BlockState state, final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+        return Redstone.SIGNAL_MIN;
+    }
+    // Pluto end - Memory allocation optimisations
+
     public final Optional<ResourceKey<LootTable>> getLootTable() {
         return this.drops;
     }
@@ -760,6 +_,12 @@
             return this.getBlock().getSignal(this.asState(), level, pos, direction);
         }
 
+        // Pluto start - Memory allocation optimisations
+        public int pluto$getSignal(final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+            return this.getBlock().pluto$getSignal(this.asState(), level, x, y, z, direction);
+        }
+        // Pluto end - Memory allocation optimisations
+
         public boolean hasAnalogOutputSignal() {
             return this.getBlock().hasAnalogOutputSignal(this.asState());
         }
@@ -779,6 +_,12 @@
         public int getDirectSignal(final BlockGetter level, final BlockPos pos, final Direction direction) {
             return this.getBlock().getDirectSignal(this.asState(), level, pos, direction);
         }
+
+        // Pluto start - Memory allocation optimisations
+        public int pluto$getDirectSignal(final BlockGetter level, final int x, final int y, final int z, final Direction direction) {
+            return this.getBlock().pluto$getDirectSignal(this.asState(), level, x, y, z, direction);
+        }
+        // Pluto end - Memory allocation optimisations
 
         public PushReaction getPistonPushReaction() {
             return !this.isDestroyable() ? PushReaction.BLOCK : this.pushReaction; // Paper - Protect Bedrock and End Portal/Frames from being destroyed
