--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -314,7 +_,7 @@
                 return false;
             }
 
-            FluidState fluidState = level.getFluidState(pos);
+            FluidState fluidState = state.getFluidState(); // Pluto - Decrease chunk/block lookups
             if (!fluidState.isEmpty() && (!preciseBlockCollisions || this.mob.collidedWithFluid(fluidState, pos, start, end))) {
                 if (fluidState.is(FluidTags.WATER)) {
                     return canPathThroughWater;
@@ -471,7 +_,7 @@
                 return false;
             }
 
-            for (Direction dir : Direction.values()) {
+            for (Direction dir : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
                 for (int i = 1; i < distanceToBlocks; i++) {
                     BlockPos offset = pos.relative(dir, i);
                     if (!level.getBlockState(offset).isAir()) {
