--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -379,6 +_,7 @@
     }
 
     protected boolean teleportSomewhere() {
+        if (this.level().plutoConfig().disableShulkerTeleporting) return false; // Pluto - Shulker configuration
         if (!this.isNoAi() && this.isAlive()) {
             BlockPos blockPos = this.blockPosition();
 
@@ -460,6 +_,7 @@
         Vec3 vec3 = this.position();
         AABB boundingBox = this.getBoundingBox();
         if (!this.isClosed() && this.teleportSomewhere()) {
+            if (this.level().plutoConfig().disableShulkerSplitting) return; // Pluto - Shulker configuration
             int size = this.level().getEntities(EntityType.SHULKER, boundingBox.inflate(8.0), Entity::isAlive).size();
             float f = (size - 1) / 5.0F;
             if (!(this.level().random.nextFloat() < f)) {
