--- a/net/minecraft/world/entity/monster/Silverfish.java
+++ b/net/minecraft/world/entity/monster/Silverfish.java
@@ -143,8 +_,7 @@
                 RandomSource random = this.mob.getRandom();
                 if (getServerLevel(this.mob).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && random.nextInt(reducedTickDelay(10)) == 0) {
                     this.selectedDirection = Direction.getRandom(random);
-                    BlockPos blockPos = BlockPos.containing(this.mob.getX(), this.mob.getY() + 0.5, this.mob.getZ()).relative(this.selectedDirection);
-                    BlockState blockState = this.mob.level().getBlockState(blockPos);
+                    BlockState blockState = this.mob.level().pluto$getBlockState(net.minecraft.util.Mth.floor(this.mob.getX()) + this.selectedDirection.getStepX(), net.minecraft.util.Mth.floor(this.mob.getY() + 0.5) + this.selectedDirection.getStepY(), net.minecraft.util.Mth.floor(this.mob.getZ()) + this.selectedDirection.getStepZ()); // Pluto - Memory efficient block lookups
                     if (InfestedBlock.isCompatibleHostBlock(blockState)) {
                         this.doMerge = true;
                         return true;
