--- a/net/minecraft/world/entity/monster/Zoglin.java
+++ b/net/minecraft/world/entity/monster/Zoglin.java
@@ -236,10 +_,12 @@
         this.setAggressive(this.brain.hasMemoryValue(MemoryModuleType.ATTACK_TARGET));
     }
 
+    private int behaviourTick = 0; // Pluto - Add DAB to more entities
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("zoglinBrain");
+        if (this.behaviourTick++ % this.activatedPriority == 0) // Pluto - Add DAB to more entities
         this.getBrain().tick(level, this);
         profiler.pop();
         this.updateActivity();
