--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -31,6 +_,27 @@
         super(entityType, level);
     }
 
+    // Pluto start - Entities search for hoppers
+    @Override
+    public boolean acceptItem(net.minecraft.world.level.block.entity.HopperBlockEntity hopper) {
+        return net.minecraft.world.level.block.entity.HopperBlockEntity.suckInItems(this.level(), hopper, this);
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        if (!this.level().plutoConfig().entitiesSearchForHoppers) return;
+        foundHopper();
+    }
+
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        if (!this.level().plutoConfig().entitiesSearchForHoppers) return;
+        foundHopper();
+    }
+    // Pluto end - Entities search for hoppers
+
     @Override
     public void destroy(ServerLevel level, DamageSource damageSource) {
         super.destroy(level, damageSource);
@@ -201,6 +_,13 @@
     public @Nullable org.bukkit.inventory.InventoryHolder getOwner() {
         return this.getBukkitEntity() instanceof final org.bukkit.inventory.InventoryHolder inventoryHolder ? inventoryHolder : null;
     }
+
+    // Pluto start - Allow toggling snapshots for certain blocks
+    @Override
+    public @Nullable org.bukkit.inventory.InventoryHolder getOwner(boolean useSnapshot) {
+        return getOwner();
+    }
+    // Pluto end - Allow toggling snapshots for certain blocks
 
     @Override
     public int getMaxStackSize() {
