--- a/net/minecraft/world/entity/vehicle/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/AbstractBoat.java
@@ -325,7 +_,14 @@
         if (!entities.isEmpty()) {
             boolean flag = !this.level().isClientSide && !(this.getControllingPassenger() instanceof Player);
 
+            int boatCount = 0; // Pluto - Option to remove excess vehicles on collision
             for (Entity entity : entities) {
+                // Pluto start - Option to remove excess vehicles on collision
+                if (entity instanceof Boat && this.level().plutoConfig().removeExcessBoats && boatCount++ >= this.level().plutoConfig().excessBoatsLimit) {
+                    entity.discard();
+                    continue;
+                }
+                // Pluto end - Option to remove excess vehicles on collision
                 if (!entity.hasPassenger(this)) {
                     if (flag
                         && this.getPassengers().size() < this.getMaxPassengers()
