--- a/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -328,10 +_,10 @@
         int floor2 = Mth.floor(this.getZ());
         if (useExperimentalMovement(this.level())) {
             double d = this.getY() - 0.1 - 1.0E-5F;
-            if (this.level().getBlockState(BlockPos.containing(floor, d, floor2)).is(BlockTags.RAILS)) {
+            if (this.level().pluto$getBlockState(floor, d, floor2).is(BlockTags.RAILS)) { // Pluto - Memory efficient block lookups
                 floor1 = Mth.floor(d);
             }
-        } else if (this.level().getBlockState(new BlockPos(floor, floor1 - 1, floor2)).is(BlockTags.RAILS)) {
+        } else if (this.level().pluto$getBlockState(floor, floor1 - 1, floor2).is(BlockTags.RAILS)) { // Pluto - Memory efficient block lookups
             floor1--;
         }
 
