--- a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
@@ -72,10 +_,12 @@
         return (Brain<ZombieNautilus>) super.getBrain();
     }
 
+    private int behaviourTick = 0; // Pluto - Add DAB to more entities
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("zombieNautilusBrain");
+        if (this.behaviourTick++ % this.activatedPriority == 0) // Pluto - Add DAB to more entities
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("zombieNautilusActivityUpdate");
