--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -53,6 +_,7 @@
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
+    public boolean wasApiUsed = false; // Pluto - Options for spawner settings | Just so that we don't overwrite custom spawners... hopefully
 
     public void setEntityId(EntityType<?> type, @Nullable Level level, RandomSource random, BlockPos pos) {
         this.getOrCreateNextSpawnData(level, random, pos).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(type).toString());
@@ -89,7 +_,19 @@
         tickDelay = level.paperConfig().tickRates.mobSpawner;
         if (tickDelay == -1) { return; } // If disabled
         // Paper end - Configurable mob spawner tick rate
-        if (this.isNearPlayer(level, pos) && level.isSpawnerBlockEnabled()) {
+
+        // Pluto start - Options for spawner settings
+        if (!wasApiUsed && level.plutoConfig().spawnerSettingsEnabled) {
+            this.minSpawnDelay = level.plutoConfig().spawnerMinSpawnDelay;
+            this.maxSpawnDelay = level.plutoConfig().spawnerMaxSpawnDelay;
+            this.spawnCount = level.plutoConfig().spawnerSpawnCount;
+            this.maxNearbyEntities = level.plutoConfig().spawnerMaxNearbyEntities;
+            this.requiredPlayerRange = level.plutoConfig().spawnerRequiredPlayerRange;
+            this.spawnRange = level.plutoConfig().spawnerSpawnRange;
+        }
+        // Pluto end - Options for spawner settings
+
+        if (level.isSpawnerBlockEnabled() && (!level.plutoConfig().spawnerCheckForNearbyPlayers || this.isNearPlayer(level, pos))) { // Pluto - Options for spawner settings
             if (this.spawnDelay < -tickDelay) { // Paper - Configurable mob spawner tick rate
                 this.delay(level, pos);
             }
@@ -118,7 +_,7 @@
                                     pos.getZ() + (random.nextDouble() - random.nextDouble()) * this.spawnRange + 0.5
                                 )
                             );
-                        if (level.noCollision(optional.get().getSpawnAABB(vec3.x, vec3.y, vec3.z))) {
+                        if (!level.plutoConfig().spawnerCheckForBlockCollision || level.noCollision(optional.get().getSpawnAABB(vec3.x, vec3.y, vec3.z))) { // Pluto - Options for spawner settings
                             BlockPos blockPos = BlockPos.containing(vec3);
                             if (nextSpawnData.getCustomSpawnRules().isPresent()) {
                                 if (!optional.get().getCategory().isFriendly() && level.getDifficulty() == Difficulty.PEACEFUL) {
@@ -157,7 +_,7 @@
                                 return;
                             }
 
-                            int size = level.getEntities(
+                            int size = !level.plutoConfig().spawnerCheckForNearbyEntities ? 0 : level.getEntities( // Pluto - Options for spawner settings
                                     EntityTypeTest.forExactClass(entity.getClass()),
                                     new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
                                     EntitySelector.NO_SPECTATORS
@@ -180,6 +_,15 @@
                                     ((Mob)entity).finalizeSpawn(level, level.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, null);
                                 }
 
+                                // Pluto start - Options for spawner settings
+                                // I wish I could put this higher up in the chain, but the baby
+                                // state only gets assigned after the finalizeSpawn method is called.
+                                if (level.plutoConfig().spawnerDisableBabySpawns && mob.isBaby()) {
+                                    if (mob.getVehicle() != null) mob.getVehicle().discard(); // Due to jockeys
+                                    continue;
+                                }
+                                // Pluto end - Options for spawner settings
+
                                 nextSpawnData.getEquipment().ifPresent(mob::equip);
                                 // Spigot start
                                 if (mob.level().spigotConfig.nerfSpawnerMobs) {
@@ -201,6 +_,7 @@
                                 return;
                             }
 
+                            if (level.plutoConfig().spawnerDisableParticles) continue; // Pluto - Options for spawner settings
                             level.levelEvent(LevelEvent.PARTICLES_MOBBLOCK_SPAWN, pos, 0);
                             level.gameEvent(entity, GameEvent.ENTITY_PLACE, blockPos);
                             if (entity instanceof Mob) {
@@ -247,9 +_,11 @@
         this.requiredPlayerRange = input.getIntOr("RequiredPlayerRange", 16);
         this.spawnRange = input.getIntOr("SpawnRange", 4);
         this.displayEntity = null;
+        this.wasApiUsed = input.getBooleanOr("Pluto.WasApiUsed", false); // Pluto - Options for spawner settings
     }
 
     public void save(ValueOutput output) {
+        output.putBoolean("Pluto.WasApiUsed", this.wasApiUsed); // Pluto - Options for spawner settings
         // Paper start
         if (this.spawnDelay > Short.MAX_VALUE) {
             output.putInt("Paper.Delay", this.spawnDelay);
