--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -29,6 +_,32 @@
         return blockEntity != null && blockEntity.getType() == type ? Optional.of((T)blockEntity) : Optional.empty();
     }
 
+    // Pluto start - Memory allocation optimisations
+    default BlockState pluto$getBlockState(double x, double y, double z) {
+        return pluto$getBlockState(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    BlockState pluto$getBlockState(final int x, final int y, final int z);
+    BlockState pluto$getBlockState(final long packedPos);
+
+    default @Nullable BlockState pluto$getBlockStateIfLoaded(final double x, final double y, final double z) {
+        return pluto$getBlockStateIfLoaded(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    @Nullable BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z);
+    @Nullable BlockState pluto$getBlockStateIfLoaded(final long packedPos);
+
+    default FluidState pluto$getFluidState(final double x, final double y, final double z) {
+        return pluto$getFluidState(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    FluidState pluto$getFluidState(final int x, final int y, final int z);
+    FluidState pluto$getFluidState(final long packedPos);
+
+    default @Nullable FluidState pluto$getFluidStateIfLoaded(final double x, final double y, final double z) {
+        return pluto$getFluidStateIfLoaded(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    @Nullable FluidState pluto$getFluidStateIfLoaded(final int x, final int y, final int z);
+    @Nullable FluidState pluto$getFluidStateIfLoaded(final long packedPos);
+    // Pluto end - Memory allocation optimisations
+
     BlockState getBlockState(final BlockPos pos);
 
     // Paper start - if loaded util
