--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -31,6 +_,32 @@
         return blockEntity != null && blockEntity.getType() == blockEntityType ? Optional.of((T)blockEntity) : Optional.empty();
     }
 
+    // Pluto start - Memory efficient block lookups
+    default BlockState pluto$getBlockState(double x, double y, double z) {
+        return pluto$getBlockState(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    BlockState pluto$getBlockState(int x, int y, int z);
+    BlockState pluto$getBlockState(long packedPos);
+
+    default @Nullable BlockState pluto$getBlockStateIfLoaded(double x, double y, double z) {
+        return pluto$getBlockStateIfLoaded(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    @Nullable BlockState pluto$getBlockStateIfLoaded(int x, int y, int z);
+    @Nullable BlockState pluto$getBlockStateIfLoaded(long packedPos);
+
+    default FluidState pluto$getFluidState(double x, double y, double z) {
+        return pluto$getFluidState(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    FluidState pluto$getFluidState(int x, int y, int z);
+    FluidState pluto$getFluidState(long packedPos);
+
+    default @Nullable FluidState pluto$getFluidStateIfLoaded(double x, double y, double z) {
+        return pluto$getFluidStateIfLoaded(Mth.floor(x), Mth.floor(y), Mth.floor(z));
+    }
+    @Nullable FluidState pluto$getFluidStateIfLoaded(int x, int y, int z);
+    @Nullable FluidState pluto$getFluidStateIfLoaded(long packedPos);
+    // Pluto end - Memory efficient block lookups
+
     BlockState getBlockState(BlockPos pos);
 
     // Paper start - if loaded util
