--- a/net/minecraft/world/level/EmptyBlockGetter.java
+++ b/net/minecraft/world/level/EmptyBlockGetter.java
@@ -16,6 +_,38 @@
         return null;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(final long packedPos) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public BlockState pluto$getBlockStateIfLoaded(final long packedPos) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        return Fluids.EMPTY.defaultFluidState();
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final long packedPos) {
+        return Fluids.EMPTY.defaultFluidState();
+    }
+    // Pluto end - Memory allocation optimisations
+
     // Paper start - If loaded util
     @Override
     public final FluidState getFluidIfLoaded(BlockPos pos) {
