--- a/net/minecraft/world/level/EmptyBlockAndTintGetter.java
+++ b/net/minecraft/world/level/EmptyBlockAndTintGetter.java
@@ -33,6 +_,48 @@
         return null;
     }
 
+    // Pluto start - Memory allocation optimisations
+    @Override
+    public BlockState pluto$getBlockState(int x, int y, int z) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(long packedPos) {
+        return Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(int x, int y, int z) {
+        return null;
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(long packedPos) {
+        return null;
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        return Fluids.EMPTY.defaultFluidState();
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        return Fluids.EMPTY.defaultFluidState();
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(int x, int y, int z) {
+        return null;
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(long packedPos) {
+        return null;
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     public BlockState getBlockState(BlockPos pos) {
         return Blocks.AIR.defaultBlockState();
