--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -94,6 +_,15 @@
         return new BlockPos(pos.getX(), this.getHeight(type, pos.getX(), pos.getZ()), pos.getZ());
     }
 
+    // Pluto start - Memory allocation optimisations
+    default boolean pluto$isEmptyBlock(final int x, final int y, final int z) {
+        return this.pluto$getBlockState(x, y, z).isAir();
+    }
+    default boolean pluto$isEmptyBlock(final long packedPos) {
+        return this.pluto$getBlockState(packedPos).isAir();
+    }
+    // Pluto end - Memory allocation optimisations
+
     default boolean isEmptyBlock(final BlockPos pos) {
         return this.getBlockState(pos).isAir();
     }
@@ -108,8 +_,13 @@
             return false;
         }
 
-        for (BlockPos var4 = scanPoint.below(); var4.getY() > pos.getY(); var4 = var4.below()) {
-            BlockState state = this.getBlockState(var4);
+        // Pluto start - Memory allocation optimisations
+        final int originY = pos.getY();
+        final int x = scanPoint.getX();
+        final int z = scanPoint.getZ();
+        for (int var4 = scanPoint.getY(); var4 > originY; var4--) {
+            BlockState state = this.pluto$getBlockState(x, var4, z);
+            // Pluto end - Memory allocation optimisations
             if (state.getLightDampening() > 0 && !state.liquid()) {
                 return false;
             }
@@ -157,12 +_,11 @@
         int y1 = Mth.ceil(box.maxY);
         int z0 = Mth.floor(box.minZ);
         int z1 = Mth.ceil(box.maxZ);
-        BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
 
         for (int x = x0; x < x1; x++) {
             for (int y = y0; y < y1; y++) {
                 for (int z = z0; z < z1; z++) {
-                    BlockState blockState = this.getBlockState(pos.set(x, y, z));
+                    BlockState blockState = this.pluto$getBlockState(x, y, z); // Pluto - Memory allocation optimisations
                     if (!blockState.getFluidState().isEmpty()) {
                         return true;
                     }
