--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -140,7 +_,7 @@
 
     public boolean captureBlockStates = false;
     public boolean captureTreeGeneration = false;
-    public Map<BlockPos, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new java.util.LinkedHashMap<>(); // Paper
+    public Map<Long, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<>(); // Paper // Pluto - Memory efficient block lookups
     public Map<BlockPos, BlockEntity> capturedTileEntities = new java.util.LinkedHashMap<>(); // Paper - Retain block place order when capturing blockstates
     @Nullable
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
@@ -160,6 +_,8 @@
     }
     // Paper end - add paper world config
 
+    private final dev.yive.pluto.PlutoConfig.WorldConfig plutoConfig; public dev.yive.pluto.PlutoConfig.WorldConfig plutoConfig() { return this.plutoConfig; } // Pluto - Config
+
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     private int tileTickPosition;
     public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions
@@ -840,6 +_,7 @@
         // Paper end - getblock optimisations - cache world height/sections
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.plutoConfig = new dev.yive.pluto.PlutoConfig.WorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Pluto - Config
         this.generator = gen;
         this.world = new CraftWorld((ServerLevel) this, gen, biomeProvider, env);
 
@@ -1019,7 +_,7 @@
     public final BlockState getBlockStateIfLoaded(BlockPos pos) {
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
-            CraftBlockState previous = this.capturedBlockStates.get(pos);
+            CraftBlockState previous = this.capturedBlockStates.get(pos.asLong()); // Pluto - Memory efficient block lookups
             if (previous != null) {
                 return previous.getHandle();
             }
@@ -1084,15 +_,16 @@
     @Override
     public boolean setBlock(BlockPos pos, BlockState state, int flags, int recursionLeft) {
         // CraftBukkit start - tree generation
+        long posLong = pos.asLong(); // Pluto - Memory efficient block lookups
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
             BlockState type = getBlockState(pos);
             if (!type.isDestroyable()) return false;
             // Paper end - Protect Bedrock and End Portal/Frames from being destroyed
-            CraftBlockState blockstate = this.capturedBlockStates.get(pos);
+            CraftBlockState blockstate = this.capturedBlockStates.get(posLong); // Pluto - Memory efficient block lookups
             if (blockstate == null) {
                 blockstate = org.bukkit.craftbukkit.block.CapturedBlockState.getTreeBlockState(this, pos, flags);
-                this.capturedBlockStates.put(pos.immutable(), blockstate);
+                this.capturedBlockStates.put(posLong, blockstate); // Pluto - Memory efficient block lookups
             }
             blockstate.setData(state);
             blockstate.setFlags(flags);
@@ -1110,12 +_,12 @@
             boolean captured = false;
             if (this.captureBlockStates) {
                 final CraftBlockState snapshot;
-                if (!this.capturedBlockStates.containsKey(pos)) {
+                if (!this.capturedBlockStates.containsKey(posLong)) { // Pluto - Memory efficient block lookups
                     snapshot = (CraftBlockState) org.bukkit.craftbukkit.block.CraftBlock.at(this, pos).getState(); // Paper - use CB getState to get a suitable snapshot
-                    this.capturedBlockStates.put(pos.immutable(), snapshot);
+                    this.capturedBlockStates.put(posLong, snapshot); // Pluto - Memory efficient block lookups
                     captured = true;
                 } else {
-                    snapshot = this.capturedBlockStates.get(pos);
+                    snapshot = this.capturedBlockStates.get(posLong); // Pluto - Memory efficient block lookups
                 }
                 snapshot.setFlags(flags); // Paper - always set the flag of the most recent call to mitigate issues with multiple update at the same pos with different flags
             }
@@ -1125,7 +_,7 @@
             if (blockState == null) {
                 // CraftBukkit start - remove blockstate if failed (or the same)
                 if (this.captureBlockStates && captured) {
-                    this.capturedBlockStates.remove(pos);
+                    this.capturedBlockStates.remove(posLong); // Pluto - Memory efficient block lookups
                 }
                 // CraftBukkit end
                 return false;
@@ -1323,11 +_,81 @@
         return this.getChunkSource().getLightEngine();
     }
 
+    // Pluto start - Memory efficient block lookups
+    @Override
+    public BlockState pluto$getBlockState(int x, int y, int z) {
+        if (this.captureTreeGeneration) {
+            CraftBlockState previous = this.capturedBlockStates.get(BlockPos.asLong(x, y, z)); // Paper
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+
+        if (this.isOutsideBuildHeight(y))
+            return Blocks.VOID_AIR.defaultBlockState();
+
+        return this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true).getBlockState(x, y, z);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(long packedPos) {
+        return pluto$getBlockState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(int x, int y, int z) {
+        if (this.captureTreeGeneration) {
+            CraftBlockState previous = this.capturedBlockStates.get(BlockPos.asLong(x, y, z));
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+
+        if (this.isOutsideBuildHeight(y))
+            return Blocks.VOID_AIR.defaultBlockState();
+
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.getBlockState(x, y, z);
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(long packedPos) {
+        return pluto$getBlockStateIfLoaded(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(int x, int y, int z) {
+        if (this.isOutsideBuildHeight(y))
+            return Fluids.EMPTY.defaultFluidState();
+
+        return this.getChunk(x >> 4, z >> 4).pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(long packedPos) {
+        return pluto$getFluidState(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(int x, int y, int z) {
+        if (this.isOutsideBuildHeight(y))
+            return Fluids.EMPTY.defaultFluidState();
+
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.pluto$getFluidState(x, y, z);
+    }
+
+    @Override
+    public @Nullable FluidState pluto$getFluidStateIfLoaded(long packedPos) {
+        return pluto$getFluidStateIfLoaded(BlockPos.getX(packedPos), BlockPos.getY(packedPos), BlockPos.getZ(packedPos));
+    }
+    // Pluto end - Memory efficient block lookups
+
     @Override
     public BlockState getBlockState(BlockPos pos) {
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
-            CraftBlockState previous = this.capturedBlockStates.get(pos); // Paper
+            CraftBlockState previous = this.capturedBlockStates.get(pos.asLong()); // Paper // Pluto - Memory efficient block lookups
             if (previous != null) {
                 return previous.getHandle();
             }
@@ -1863,8 +_,11 @@
     public abstract Collection<EnderDragonPart> dragonParts();
 
     public void blockEntityChanged(BlockPos pos) {
-        if (this.hasChunkAt(pos)) {
-            this.getChunkAt(pos).markUnsaved();
+        // Pluto start - Decrease chunk/block lookups
+        LevelChunk chunk = this.getChunkIfLoaded(pos);
+        if (chunk != null) {
+            chunk.markUnsaved();
+            // Pluto end - Decrease chunk/block lookups
         }
     }
 
@@ -1978,12 +_,17 @@
 
     public void updateNeighbourForOutputSignal(BlockPos pos, Block block) {
         for (Direction direction : Direction.Plane.HORIZONTAL) {
+            // Pluto start - Decrease chunk/block lookups
+            final int x = pos.getX() + direction.getStepX();
+            final int z = pos.getZ() + direction.getStepZ();
             BlockPos blockPos = pos.relative(direction);
-            if (this.hasChunkAt(blockPos)) {
-                BlockState blockState = this.getBlockState(blockPos);
+            LevelChunk chunk = this.getChunkIfLoaded(x >> 4, z >> 4);
+            if (chunk != null) {
+                BlockState blockState = chunk.getBlockState(x, pos.getY() + direction.getStepY(), z);
                 if (blockState.is(Blocks.COMPARATOR)) {
                     this.neighborChanged(blockState, blockPos, block, null, false);
-                } else if (blockState.isRedstoneConductor(this, blockPos)) {
+                } else if (blockState.isRedstoneConductor(this, blockPos = pos.relative(direction))) {
+                    // Pluto end - Decrease chunk/block lookups
                     blockPos = blockPos.relative(direction);
                     blockState = this.getBlockState(blockPos);
                     if (blockState.is(Blocks.COMPARATOR)) {
@@ -1998,9 +_,12 @@
     public DifficultyInstance getCurrentDifficultyAt(BlockPos pos) {
         long l = 0L;
         float f = 0.0F;
-        if (this.hasChunkAt(pos)) {
+        // Pluto start - Decrease chunk/block lookups
+        LevelChunk chunk = this.getChunkIfLoaded(pos);
+        if (chunk != null) {
+            // Pluto end - Decrease chunk/block lookups
             f = this.getMoonBrightness();
-            l = this.getChunkAt(pos).getInhabitedTime();
+            l = chunk.getInhabitedTime(); // Pluto - Decrease chunk/block lookups
         }
 
         return new DifficultyInstance(this.getDifficulty(), this.getDayTime(), l, f);
