--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -152,7 +_,7 @@
 
     public boolean captureBlockStates = false;
     public boolean captureTreeGeneration = false;
-    public Map<BlockPos, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new java.util.LinkedHashMap<>(); // Paper
+    public Map<Long, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<>(); // Paper // Pluto - Memory allocation optimisations
     public Map<BlockPos, BlockEntity> capturedBlockEntities = new java.util.LinkedHashMap<>(); // Paper - Retain block place order when capturing blockstates
     @Nullable
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
@@ -172,6 +_,13 @@
     }
     // Paper end - add paper world config
 
+    private final dev.yive.pluto.config.PlutoConfig.WorldConfig plutoConfig; public dev.yive.pluto.config.PlutoConfig.WorldConfig plutoConfig() { return this.plutoConfig; } // Pluto - Config
+    // Pluto start - Async spawning chunks
+    private final java.util.concurrent.ExecutorService pluto$spawningChunksExecutor; public java.util.concurrent.ExecutorService plutoSpawningChunksExecutor() { return pluto$spawningChunksExecutor; }
+    @Deprecated
+    private final RandomSource pluto$threadSafeRandom = RandomSource.createThreadSafe(); public @Deprecated RandomSource pluto$threadSafeRandom() { return this.pluto$threadSafeRandom; }
+    // Pluto end - Async spawning chunks
+
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions
     public java.util.ArrayDeque<net.minecraft.world.level.block.RedstoneTorchBlock.Toggle> redstoneUpdateInfos; // Paper - Faster redstone torch rapid clock removal; Move from Map in BlockRedstoneTorch to here
@@ -612,7 +_,7 @@
 
         final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
         final ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.LazyEntityCollisionContext collisionShape = new ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.LazyEntityCollisionContext(entity);
-        BlockPos selected = null;
+        Long selected = null; // Pluto - Memory allocation optimisations
         double selectedDistance = Double.MAX_VALUE;
         final Vec3 entityPos = entity.position();
 
@@ -686,7 +_,7 @@
                                 }
 
                                 final double distance = mutablePos.distToCenterSqr(entityPos);
-                                if (distance > selectedDistance || (distance == selectedDistance && selected.compareTo(mutablePos) >= 0)) {
+                                if (distance > selectedDistance || (distance == selectedDistance && selected.compareTo(mutablePos.asLong()) >= 0)) { // Pluto - Memory allocation optimisations
                                     continue;
                                 }
 
@@ -708,24 +_,38 @@
                                     }
 
                                     // avoid VoxelShape#move by shifting the entity collision shape instead
-                                    final AABB shiftedAABB = aabb.move(-(double)blockX, -(double)blockY, -(double)blockZ);
+                                    // Pluto start - Memory allocation optimisations
+                                    final double x1 = aabb.minX + (-(double)blockX);
+                                    final double x2 = aabb.maxX + (-(double)blockX);
+                                    final double y1 = aabb.minY + (-(double)blockY);
+                                    final double y2 = aabb.maxY + (-(double)blockY);
+                                    final double z1 = aabb.minZ + (-(double)blockZ);
+                                    final double z2 = aabb.maxZ + (-(double)blockZ);
+
+                                    final double minX = Math.min(x1, x2);
+                                    final double minY = Math.min(y1, y2);
+                                    final double minZ = Math.min(z1, z2);
+                                    final double maxX = Math.max(x1, x2);
+                                    final double maxY = Math.max(y1, y2);
+                                    final double maxZ = Math.max(z1, z2);
+                                    // Pluto end - Memory allocation optimisations
 
                                     final AABB singleAABB = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)blockCollision).moonrise$getSingleAABBRepresentation();
                                     if (singleAABB != null) {
-                                        if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersect(singleAABB, shiftedAABB)) {
+                                        if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersect(singleAABB, minX, minY, minZ, maxX, maxY, maxZ)) { // Pluto - Memory allocation optimisations
                                             continue;
                                         }
 
-                                        selected = mutablePos.immutable();
+                                        selected = mutablePos.asLong(); // Pluto - Memory allocation optimisations
                                         selectedDistance = distance;
                                         continue;
                                     }
 
-                                    if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersectNoEmpty(blockCollision, shiftedAABB)) {
+                                    if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.pluto$voxelShapeIntersectNoEmpty(blockCollision, minX, minY, minZ, maxX, maxY, maxZ)) { // Pluto - Memory allocation optimisations
                                         continue;
                                     }
 
-                                    selected = mutablePos.immutable();
+                                    selected = mutablePos.asLong(); // Pluto - Memory allocation optimisations
                                     selectedDistance = distance;
                                     continue;
                                 }
@@ -736,7 +_,7 @@
             }
         }
 
-        return java.util.Optional.ofNullable(selected);
+        return java.util.Optional.ofNullable(selected == null ? null : BlockPos.of(selected)); // Pluto - Memory allocation optimisations
     }
     // Paper end - optimise collisions
     // Paper start - getblock optimisations - cache world height/sections
@@ -853,6 +_,8 @@
         final org.bukkit.NamespacedKey worldKey = CraftNamespacedKey.fromMinecraft(dimension.identifier()); // Paper
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(bukkitName, worldKey); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.plutoConfig = new dev.yive.pluto.config.PlutoConfig.WorldConfig(bukkitName, worldKey); // Pluto - Config
+        this.pluto$spawningChunksExecutor = java.util.concurrent.Executors.newSingleThreadExecutor(new com.google.common.util.concurrent.ThreadFactoryBuilder().setNameFormat(worldKey + " - Spawning Chunks Executor").build()); // Pluto - Async spawning chunks
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, worldKey, biomeProvider, environment);
 
@@ -992,7 +_,7 @@
     public final BlockState getBlockStateIfLoaded(BlockPos pos) {
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
-            CraftBlockState previous = this.capturedBlockStates.get(pos);
+            CraftBlockState previous = this.capturedBlockStates.get(pos.asLong()); // Pluto - Memory allocation optimisations
             if (previous != null) {
                 return previous.getHandle();
             }
@@ -1055,15 +_,16 @@
     @Override
     public boolean setBlock(final BlockPos pos, final BlockState blockState, final @Block.UpdateFlags int updateFlags, final int updateLimit) {
         // CraftBukkit start - tree generation
+        long posLong = pos.asLong(); // Pluto - Memory allocation optimisations
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
             BlockState type = getBlockState(pos);
             if (!type.isDestroyable()) return false;
             // Paper end - Protect Bedrock and End Portal/Frames from being destroyed
-            CraftBlockState blockstate = this.capturedBlockStates.get(pos);
+            CraftBlockState blockstate = this.capturedBlockStates.get(posLong); // Pluto - Memory allocation optimisations
             if (blockstate == null) {
                 blockstate = org.bukkit.craftbukkit.block.CapturedBlockState.getTreeBlockState(this, pos, updateFlags);
-                this.capturedBlockStates.put(pos.immutable(), blockstate);
+                this.capturedBlockStates.put(posLong, blockstate); // Pluto - Memory allocation optimisations
             }
             blockstate.setBlock(blockState);
             blockstate.setFlags(updateFlags);
@@ -1084,12 +_,12 @@
         boolean captured = false;
         if (this.captureBlockStates) {
             final CraftBlockState snapshot;
-            if (!this.capturedBlockStates.containsKey(pos)) {
+            if (!this.capturedBlockStates.containsKey(posLong)) { // Pluto - Memory allocation optimisations
                 snapshot = (CraftBlockState) org.bukkit.craftbukkit.block.CraftBlock.at(this, pos).getState(); // Paper - use CB getState to get a suitable snapshot
-                this.capturedBlockStates.put(pos.immutable(), snapshot);
+                this.capturedBlockStates.put(posLong, snapshot); // Pluto - Memory allocation optimisations
                 captured = true;
             } else {
-                snapshot = this.capturedBlockStates.get(pos);
+                snapshot = this.capturedBlockStates.get(posLong); // Pluto - Memory allocation optimisations
             }
             snapshot.setFlags(updateFlags); // Paper - always set the flag of the most recent call to mitigate issues with multiple update at the same pos with different flags
         }
@@ -1099,7 +_,7 @@
         if (oldState == null) {
             // CraftBukkit start - remove blockstate if failed (or the same)
             if (this.captureBlockStates && captured) {
-                this.capturedBlockStates.remove(pos);
+                this.capturedBlockStates.remove(posLong); // Pluto - Memory allocation optimisations
             }
             // CraftBukkit end
             return false;
@@ -1306,11 +_,60 @@
         return this.getChunkSource().getLightEngine();
     }
 
+    // Pluto start - Memory allocation optimisations
+    public boolean isInValidBounds(int x, int y, int z) {
+        return !this.isOutsideBuildHeight(y) && isInValidBoundsHorizontal(x, z);
+    }
+
+    private static boolean isInValidBoundsHorizontal(int x, int z) {
+        return ChunkPos.isValid(x >> 4, z >> 4);
+    }
+
+    @Override
+    public BlockState pluto$getBlockState(final int x, final int y, final int z) {
+        if (this.captureTreeGeneration) {
+            CraftBlockState previous = this.capturedBlockStates.get(BlockPos.asLong(x, y, z)); // Paper
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+
+        if (!this.isInValidBounds(x, y, z))
+            return Blocks.VOID_AIR.defaultBlockState();
+
+        return this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true).getBlockState(x, y, z);
+    }
+
+    @Override
+    public @Nullable BlockState pluto$getBlockStateIfLoaded(final int x, final int y, final int z) {
+        if (this.captureTreeGeneration) {
+            CraftBlockState previous = this.capturedBlockStates.get(BlockPos.asLong(x, y, z));
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+
+        if (!this.isInValidBounds(x, y, z))
+            return Blocks.VOID_AIR.defaultBlockState();
+
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+        return chunk == null ? null : chunk.getBlockState(x, y, z);
+    }
+
+    @Override
+    public FluidState pluto$getFluidState(final int x, final int y, final int z) {
+        if (!this.isInValidBounds(x, y, z))
+            return Fluids.EMPTY.defaultFluidState();
+
+        return this.getChunk(x >> 4, z >> 4).pluto$getFluidState(x, y, z);
+    }
+    // Pluto end - Memory allocation optimisations
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
-            CraftBlockState previous = this.capturedBlockStates.get(pos); // Paper
+            CraftBlockState previous = this.capturedBlockStates.get(pos.asLong()); // Paper // Pluto - Memory allocation optimisations
             if (previous != null) {
                 return previous.getHandle();
             }
@@ -1501,9 +_,10 @@
         for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) {
             final TickingBlockEntity ticker = this.blockEntityTickers.get(tickerIndex);
         // Paper end - Fix MC-117075 use removeAll
+
             if (ticker.isRemoved()) {
                 toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
-            } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
+            } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getChunkPos())) { // Pluto - Optimise block entity ticking
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 if ((++tickedEntities & 7) == 0) {
@@ -1520,12 +_,12 @@
     public <T extends Entity> void guardEntityTick(final Consumer<T> tick, final T entity) {
         try {
             tick.accept(entity);
-        } catch (Throwable t) {
+        } catch (Throwable t) { // Pufferfish - diff on change ServerLevel.tick
             // Paper start - Prevent block entity and entity crashes
             final String msg = String.format("Entity threw exception at %s:%s,%s,%s", io.papermc.paper.util.MCUtil.getLevelName(entity.level()), entity.getX(), entity.getY(), entity.getZ());
             MinecraftServer.LOGGER.error(msg, t);
             getCraftServer().getPluginManager().callEvent(new com.destroystokyo.paper.event.server.ServerExceptionEvent(new com.destroystokyo.paper.exception.ServerInternalException(msg, t))); // Paper - ServerExceptionEvent
-            entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+            entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD); // Pufferfish - diff on change ServerLevel.tick
             // Paper end - Prevent block entity and entity crashes
         }
         this.moonrise$midTickTasks(); // Paper - rewrite chunk system
@@ -1926,8 +_,11 @@
     public abstract Collection<EnderDragonPart> dragonParts();
 
     public void blockEntityChanged(final BlockPos pos) {
-        if (this.hasChunkAt(pos)) {
-            this.getChunkAt(pos).markUnsaved();
+        // Pluto start - Decrease chunk/block lookups
+        LevelChunk chunk = this.getChunkIfLoaded(pos);
+        if (chunk != null) {
+            chunk.markUnsaved();
+            // Pluto end - Decrease chunk/block lookups
         }
     }
 
@@ -2054,17 +_,21 @@
     public abstract Scoreboard getScoreboard();
 
     public void updateNeighbourForOutputSignal(final BlockPos pos, final Block changedBlock) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
-            BlockPos relativePos = pos.relative(direction);
-            if (this.hasChunkAt(relativePos)) {
-                BlockState state = this.getBlockState(relativePos);
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+            // Pluto start - Decrease chunk/block lookups
+            final int x = pos.getX() + direction.getStepX();
+            final int z = pos.getZ() + direction.getStepZ();
+            LevelChunk chunk = this.getChunkIfLoaded(x >> 4, z >> 4);
+            if (chunk != null) {
+                BlockState state = chunk.getBlockState(x, pos.getY() + direction.getStepY(), z);
+                BlockPos relativePos = pos.relative(direction);
                 if (state.is(Blocks.COMPARATOR)) {
                     this.neighborChanged(state, relativePos, changedBlock, null, false);
                 } else if (state.isRedstoneConductor(this, relativePos)) {
-                    relativePos = relativePos.relative(direction);
-                    state = this.getBlockState(relativePos);
+                    // Pluto end - Decrease chunk/block lookups
+                    state = this.pluto$getBlockState(relativePos.getX() + direction.getStepX(), relativePos.getY() + direction.getStepY(), relativePos.getZ() + direction.getStepZ()); // Pluto - Memory allocation optimisations
                     if (state.is(Blocks.COMPARATOR)) {
-                        this.neighborChanged(state, relativePos, changedBlock, null, false);
+                        this.neighborChanged(state, relativePos.relative(direction), changedBlock, null, false); // Pluto - Memory allocation optimisations
                     }
                 }
             }
