--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -160,6 +_,8 @@
     }
     // Paper end - add paper world config
 
+    private final dev.yive.pluto.PlutoConfig.WorldConfig plutoConfig; public dev.yive.pluto.PlutoConfig.WorldConfig plutoConfig() { return this.plutoConfig; } // Pluto - Config
+
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     private int tileTickPosition;
     public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions
@@ -840,6 +_,7 @@
         // Paper end - getblock optimisations - cache world height/sections
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.plutoConfig = new dev.yive.pluto.PlutoConfig.WorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Pluto - Config
         this.generator = gen;
         this.world = new CraftWorld((ServerLevel) this, gen, biomeProvider, env);
 
@@ -1863,8 +_,11 @@
     public abstract Collection<EnderDragonPart> dragonParts();
 
     public void blockEntityChanged(BlockPos pos) {
-        if (this.hasChunkAt(pos)) {
-            this.getChunkAt(pos).markUnsaved();
+        // Pluto start - Decrease chunk/block lookups
+        LevelChunk chunk = this.getChunkIfLoaded(pos);
+        if (chunk != null) {
+            chunk.markUnsaved();
+            // Pluto end - Decrease chunk/block lookups
         }
     }
 
@@ -1979,8 +_,11 @@
     public void updateNeighbourForOutputSignal(BlockPos pos, Block block) {
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockPos blockPos = pos.relative(direction);
-            if (this.hasChunkAt(blockPos)) {
-                BlockState blockState = this.getBlockState(blockPos);
+            // Pluto start - Decrease chunk/block lookups
+            LevelChunk chunk = this.getChunkIfLoaded(blockPos);
+            if (chunk != null) {
+                BlockState blockState = chunk.getBlockState(blockPos);
+                // Pluto end - Decrease chunk/block lookups
                 if (blockState.is(Blocks.COMPARATOR)) {
                     this.neighborChanged(blockState, blockPos, block, null, false);
                 } else if (blockState.isRedstoneConductor(this, blockPos)) {
@@ -1998,9 +_,12 @@
     public DifficultyInstance getCurrentDifficultyAt(BlockPos pos) {
         long l = 0L;
         float f = 0.0F;
-        if (this.hasChunkAt(pos)) {
+        // Pluto start - Decrease chunk/block lookups
+        LevelChunk chunk = this.getChunkIfLoaded(pos);
+        if (chunk != null) {
+            // Pluto end - Decrease chunk/block lookups
             f = this.getMoonBrightness();
-            l = this.getChunkAt(pos).getInhabitedTime();
+            l = chunk.getInhabitedTime(); // Pluto - Decrease chunk/block lookups
         }
 
         return new DifficultyInstance(this.getDifficulty(), this.getDayTime(), l, f);
