--- a/net/minecraft/world/level/SignalGetter.java
+++ b/net/minecraft/world/level/SignalGetter.java
@@ -11,33 +_,57 @@
 public interface SignalGetter extends BlockGetter {
     Direction[] DIRECTIONS = Direction.values();
 
+    // Pluto start - Memory allocation optimisations
+    default int pluto$getDirectSignal(final int x, final int y, final int z, final Direction direction) {
+        return this.pluto$getBlockState(x, y, z).pluto$getDirectSignal(this, x, y, z, direction);
+    }
+
+    default int pluto$getControlInputSignal(final int x, final int y, final int z, final Direction direction, final boolean diodesOnly) {
+        BlockState blockState = this.pluto$getBlockState(x, y, z);
+        if (diodesOnly) {
+            return DiodeBlock.isDiode(blockState) ? this.pluto$getDirectSignal(x, y, z, direction) : Redstone.SIGNAL_MIN;
+        } else if (blockState.is(Blocks.REDSTONE_BLOCK)) {
+            return Redstone.SIGNAL_MAX;
+        } else if (blockState.is(Blocks.REDSTONE_WIRE)) {
+            return blockState.getValue(RedStoneWireBlock.POWER);
+        } else {
+            return blockState.isSignalSource() ? this.pluto$getDirectSignal(x, y, z, direction) : Redstone.SIGNAL_MIN;
+        }
+    }
+    // Pluto end - Memory allocation optimisations
+
     default int getDirectSignal(final BlockPos pos, final Direction direction) {
         return this.getBlockState(pos).getDirectSignal(this, pos, direction);
     }
 
     default int getDirectSignalTo(final BlockPos pos) {
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        // Pluto end - Memory allocation optimisations
         int result = 0;
-        result = Math.max(result, this.getDirectSignal(pos.below(), Direction.DOWN));
+        result = Math.max(result, this.pluto$getDirectSignal(x, y - 1, z, Direction.DOWN)); // Pluto - Memory allocation optimisations
         if (result >= Redstone.SIGNAL_MAX) {
             return result;
         } else {
-            result = Math.max(result, this.getDirectSignal(pos.above(), Direction.UP));
+            result = Math.max(result, this.pluto$getDirectSignal(x, y + 1, z, Direction.UP)); // Pluto - Memory allocation optimisations
             if (result >= Redstone.SIGNAL_MAX) {
                 return result;
             } else {
-                result = Math.max(result, this.getDirectSignal(pos.north(), Direction.NORTH));
+                result = Math.max(result, this.pluto$getDirectSignal(x, y, z - 1, Direction.NORTH)); // Pluto - Memory allocation optimisations
                 if (result >= Redstone.SIGNAL_MAX) {
                     return result;
                 } else {
-                    result = Math.max(result, this.getDirectSignal(pos.south(), Direction.SOUTH));
+                    result = Math.max(result, this.pluto$getDirectSignal(x, y, z + 1, Direction.SOUTH)); // Pluto - Memory allocation optimisations
                     if (result >= Redstone.SIGNAL_MAX) {
                         return result;
                     } else {
-                        result = Math.max(result, this.getDirectSignal(pos.west(), Direction.WEST));
+                        result = Math.max(result, this.pluto$getDirectSignal(x - 1, y, z, Direction.WEST)); // Pluto - Memory allocation optimisations
                         if (result >= Redstone.SIGNAL_MAX) {
                             return result;
                         } else {
-                            result = Math.max(result, this.getDirectSignal(pos.east(), Direction.EAST));
+                            result = Math.max(result, this.pluto$getDirectSignal(x + 1, y, z, Direction.EAST)); // Pluto - Memory allocation optimisations
                             return result >= Redstone.SIGNAL_MAX ? result : result;
                         }
                     }
