--- a/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java
+++ b/src/main/java/gg/pufferfish/pufferfish/PufferfishVersionFetcher.java
@@ -31,8 +_,10 @@
 	private static final Logger LOGGER = Logger.getLogger("PufferfishVersionFetcher");
 	private static final HttpClient client = HttpClient.newHttpClient();
 	
-	private static final URI JENKINS_URI = URI.create("https://ci.yive.dev/job/Pufferfish/job/1.21.6/lastSuccessfulBuild/buildNumber");
-	private static final String GITHUB_FORMAT = "https://api.github.com/repos/Yive/Tetraodontidae/compare/ver/1.21.6...%s";
+	// Pluto start - Branding
+	private static final URI JENKINS_URI = URI.create("https://ci.yive.dev/job/Pluto/job/1.21.6/lastSuccessfulBuild/buildNumber");
+	private static final String GITHUB_FORMAT = "https://api.github.com/repos/Yive/Pluto/compare/ver/1.21.6...%s";
+	// Pluto end - Branding
 	
 	private static final HttpResponse.BodyHandler<JsonObject> JSON_OBJECT_BODY_HANDLER = responseInfo -> HttpResponse.BodySubscribers
 			.mapping(
@@ -108,7 +_,7 @@
 	
 	private @NotNull Component getResponseMessage(final int versionDiff) {
 		return switch (Math.max(-1, Math.min(1, versionDiff))) {
-			case -1 -> text("You are running an unsupported version of Pufferfish.", RED);
+			case -1 -> text("You are running an unsupported version of Pluto.", RED); // Pluto - Branding
 			case 0 -> text("You are on the latest version!", GREEN);
 			default -> text("You are running " + versionDiff + " version" + (versionDiff == 1 ? "" : "s") + " behind. " +
 							"Please update your server when possible to maintain stability and security, and to receive the latest optimizations.",
