--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -995,6 +_,19 @@
      */
     <T> void setMemory(@NotNull MemoryKey<T> memoryKey, @Nullable T memoryValue);
 
+    // Pluto start - Expose some sound related functions
+    /**
+     * Gets entity voice pitch
+     *
+     * @deprecated use {@link #getSoundPitch()}
+     * @return entity voice pitch
+     */
+    @Deprecated
+    default float getVoicePitch() {
+        return getSoundPitch();
+    }
+    // Pluto end - Expose some sound related functions
+
     /**
      * Get the sound volume at which this entity plays its sounds with.
      *
