--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -750,7 +_,7 @@
             double x = slope == 0.0 ? xdiff * (i / 6.0F) : z / slope;
 
             for (int j = 1; j < 4; j++) {
-                if (!fox.level().getBlockState(BlockPos.containing(fox.getX() + x, fox.getY() + j, fox.getZ() + z)).canBeReplaced()) {
+                if (!fox.level().pluto$getBlockState(Mth.floor(fox.getX() + x), Mth.floor(fox.getY() + j), Mth.floor(fox.getZ() + z)).canBeReplaced()) { // Pluto - Memory allocation optimisations
                     return false;
                 }
             }
