--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -917,6 +_,18 @@
         this.getHandle().getBrain().setMemory(CraftMemoryKey.bukkitToMinecraft(memoryKey), CraftMemoryMapper.toNms(t));
     }
 
+    // Pluto start - Expose some sound related functions
+    @Override
+    public float getSoundVolume() {
+        return this.getHandle().plutoGetSoundVolume();
+    }
+
+    @Override
+    public float getVoicePitch() {
+        return this.getHandle().plutoGetVoicePitch();
+    }
+    // Pluto end - Expose some sound related functions
+
     @Override
     public Sound getHurtSound() {
         SoundEvent sound = this.getHandle().getHurtSound(this.getHandle().damageSources().generic());
