--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -705,6 +_,13 @@
             name = name.substring(0, 256);
         }
 
+        // Pluto start - Improve performance of custom name setting
+        if (this.getHandle().level().plutoConfig().useCustomNameOptimisation) {
+            this.getHandle().setCustomName(CraftChatMessage.fromStringOrNull(name, false, true));
+            return;
+        }
+        // Pluto end - Improve performance of custom name setting
+
         this.getHandle().setCustomName(CraftChatMessage.fromStringOrNull(name));
     }
 
