--- a/net/minecraft/core/cauldron/CauldronInteractions.java
+++ b/net/minecraft/core/cauldron/CauldronInteractions.java
@@ -373,7 +_,7 @@
     }
 
     private static boolean isUnderWater(final Level level, final BlockPos pos) {
-        FluidState fluidState = level.getFluidState(pos.above());
+        FluidState fluidState = level.pluto$getFluidState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
         return fluidState.is(FluidTags.WATER);
     }
 }
