--- a/net/minecraft/world/entity/animal/turtle/Turtle.java
+++ b/net/minecraft/world/entity/animal/turtle/Turtle.java
@@ -386,7 +_,7 @@
                     nextPos = DefaultRandomPos.getPosTowards(this.turtle, 8, 7, homePosVec, (float) (Math.PI / 2));
                 }
 
-                if (nextPos != null && !closeToHome && !this.turtle.level().getBlockState(BlockPos.containing(nextPos)).is(Blocks.WATER)) {
+                if (nextPos != null && !closeToHome && !this.turtle.level().pluto$getBlockState(nextPos.asBlockPosLong()).is(Blocks.WATER)) { // Pluto - Memory allocation optimisations
                     nextPos = DefaultRandomPos.getPosTowards(this.turtle, 16, 5, homePosVec, (float) (Math.PI / 2));
                 }
 
@@ -484,7 +_,7 @@
 
         @Override
         protected boolean isValidTarget(final LevelReader level, final BlockPos pos) {
-            return level.isEmptyBlock(pos.above()) && TurtleEggBlock.isSand(level, pos);
+            return level.pluto$isEmptyBlock(pos.getX(), pos.getY() + 1, pos.getZ()) && TurtleEggBlock.isSand(level, pos); // Pluto - Memory allocation optimisations
         }
     }
 
