--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -186,11 +_,13 @@
             }
 
             if (this.isAlive()) {
-                BlockPos currentPos = BlockPos.containing(this.position());
+                long currentPos = this.position().asBlockPosLong(); // Pluto - Memory allocation optimisations
                 if ((
                         --this.ticketTimer <= 0L
-                            || var7 != SectionPos.blockToSectionCoord(currentPos.getX())
-                            || previousChunkZ != SectionPos.blockToSectionCoord(currentPos.getZ())
+                            // Pluto start - Memory allocation optimisations
+                            || var7 != SectionPos.blockToSectionCoord(BlockPos.getX(currentPos))
+                            || previousChunkZ != SectionPos.blockToSectionCoord(BlockPos.getZ(currentPos))
+                            // Pluto end - Memory allocation optimisations
                     )
                     && owner instanceof ServerPlayer serverPlayer) {
                     this.ticketTimer = serverPlayer.registerAndUpdateEnderPearlTicket(this);
