--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -343,7 +_,7 @@
     @Override
     protected void playStepSound(BlockPos pos, BlockState block) {
         if (!block.liquid()) {
-            BlockState blockState = this.level().getBlockState(pos.above());
+            BlockState blockState = this.level().pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
             SoundType soundType = block.getSoundType();
             if (blockState.is(Blocks.SNOW)) {
                 soundType = blockState.getSoundType();
