--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -343,7 +_,7 @@
     @Override
     protected void playStepSound(final BlockPos pos, final BlockState blockState) {
         if (!blockState.liquid()) {
-            BlockState aboveState = this.level().getBlockState(pos.above());
+            BlockState aboveState = this.level().pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
             SoundType soundType = blockState.getSoundType();
             if (aboveState.is(Blocks.SNOW)) {
                 soundType = aboveState.getSoundType();
