--- a/net/minecraft/world/inventory/MerchantContainer.java
+++ b/net/minecraft/world/inventory/MerchantContainer.java
@@ -51,6 +_,13 @@
         return (this.merchant instanceof net.minecraft.world.entity.npc.AbstractVillager abstractVillager) ? (org.bukkit.craftbukkit.entity.CraftAbstractVillager) abstractVillager.getBukkitEntity() : null;
     }
 
+    // Pluto start - Allow toggling snapshots for certain blocks
+    @Override
+    public @javax.annotation.Nullable org.bukkit.inventory.InventoryHolder getOwner(boolean useSnapshot) {
+        return getOwner();
+    }
+    // Pluto end - Allow toggling snapshots for certain blocks
+
     @Override
     public @javax.annotation.Nullable org.bukkit.Location getLocation() {
         return (this.merchant instanceof net.minecraft.world.entity.npc.AbstractVillager abstractVillager) ? abstractVillager.getBukkitEntity().getLocation() : null; // Paper - Fix inventories returning null Locations
@@ -119,6 +_,8 @@
     public boolean stillValid(Player player) {
         return this.merchant.getTradingPlayer() == player;
     }
+
+    @Override public boolean stillValid() { return this.merchant.stillValid(); } // Pluto - Hoppers cache containers
 
     @Override
     public void setChanged() {
