From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 16:13:47 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index aae422978887a0f2cf531b2dc862705148569676..f97f6777f171d106b93a2b1ce2946615a3d08107 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -169,6 +169,7 @@ tasks.jar {
 
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Pluto - Build fixes
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
