From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 16:13:47 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index 80833c13a2d19e31d6f648e7ef5b3456025e767e..b17afe8d2fe8b3577a0468a6fc08f82b214cffc7 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -79,7 +79,7 @@ dependencies {
 }
 
 // Paper start
-val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("paper-api-generator/generated")
+val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("pluto-api-generator/generated") // Pluto - Build fixes
 idea {
     module {
         generatedSourceDirs.add(generatedApiPath.toFile())
@@ -130,6 +130,7 @@ tasks.jar {
 
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Pluto - Build fixes
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
