From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 16:13:47 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index b9c75a190dbd7a90ac5ef0fbc6e6fe34806acc4e..249efc2cc2e7a7b57fa94f6ed4dc29ce9b04fe56 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -76,7 +76,7 @@ dependencies {
 }
 
 // Paper start
-val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("paper-api-generator/generated")
+val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("pluto-api-generator/generated") // Pluto - Build fixes
 idea {
     module {
         generatedSourceDirs.add(generatedApiPath.toFile())
@@ -127,6 +127,7 @@ tasks.jar {
 
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Pluto - Build fixes
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
