From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 6 Aug 2024 22:31:20 -0700
Subject: [PATCH] Branding


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..64786b5167feba7a5e8b0db58a8289973cfe8c87 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,13 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // Pluto start - Branding
+    /**
+     * The brand id for Pluto.
+     */
+    Key BRAND_PLUTO_ID = Key.key("yive", "pluto");
+    // Pluto end - Branding
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
