From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 20 Aug 2024 00:47:52 -0700
Subject: [PATCH] Add method to obtain current entity id


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 10c87b7c19ed3eab28fdce5f225df3767292ee0a..c00db88910f2538868c7daa96026ff3662962483 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -219,6 +219,13 @@ public interface UnsafeValues {
      */
     public int nextEntityId();
 
+    // Pluto start - Add method to obtain current entity id
+    /**
+     * Returns the latest entity id.
+     */
+    public int currentEntityId();
+    // Pluto end - Add method to obtain current entity id
+
     /**
      * Just don't use it.
      */
