From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 20 Nov 2024 18:26:45 -0800
Subject: [PATCH] Expose some sound related functions


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 68c08e7a212bc3e3885f9b5a4d9aef85fcb3b029..df530675fb7d7f6fbbbda2cbb52f56c1674cb77c 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1468,4 +1468,21 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     boolean canUseEquipmentSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
     // Paper end - Expose canUseSlot
+
+    // Pluto start - Expose some sound related functions
+    /**
+     * Gets entity sound volume
+     *
+     * @return entity sound volume
+     */
+    float getSoundVolume();
+
+    /**
+     * Gets entity voice pitch
+     *
+     * @return entity voice pitch
+     */
+    float getVoicePitch();
+    // Pluto end - Expose some sound related functions
+
 }
