From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 18:28:51 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index 0668648a49be0a999e739ddad66e6e101f166598..cef43a2fa6f3c3b6a33e9e45ca05b8cd21494c68 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -99,6 +99,13 @@ tasks.jar {
     }
 }
 
+// Pluto start - Build fixes
+// Probably a terrible way to do this, but it works
+tasks.withType<PublishToMavenRepository>().configureEach {
+    enabled = false
+}
+// Pluto end - Build fixes
+
 publishing {
     publications.create<MavenPublication>("maven") {
         artifact(tasks.shadowJar)
