From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 18:28:51 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index 77f76653c22c44ac9ab7f396b686affc228502c9..a5da0461a8e3e337ecc24420211b6cc7563cd0f1 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -116,6 +116,13 @@ tasks.compileTestJava {
 }
 // Paper end
 
+// Pluto start - Build fixes
+// Probably a terrible way to do this, but it works
+tasks.withType<PublishToMavenRepository>().configureEach {
+    enabled = false
+}
+// Pluto end - Build fixes
+
 publishing {
     publications.create<MavenPublication>("maven") {
         artifact(tasks.shadowJar)
