From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 16 Apr 2023 18:28:51 -0700
Subject: [PATCH] Build fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index 18bd50340f9e511781dae123cdb82bc16da97d55..e2c67612317ec06a315503bf5c41b1530aace10c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -111,6 +111,12 @@ tasks.compileJava {
     // incremental compilation is currently broken due to patched files having compiled counterparts already on the compile classpath
     options.setIncremental(false)
 }
+// Pluto start - Build fixes
+// Probably a terrible way to do this, but it works
+tasks.withType<PublishToMavenRepository>().configureEach {
+    enabled = false
+}
+// Pluto end - Build fixes
 
 publishing {
     publications.create<MavenPublication>("maven") {
