From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Fri, 6 May 2022 15:29:09 -0700
Subject: [PATCH] Cache GameEvent lookups


diff --git a/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java b/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
index f2d10d58617644a589ecec3e17358c1277795e5d..4ca9a2fdbef3eb186fa5290190071f353ac3b71e 100644
--- a/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -26,17 +26,28 @@ public class GameEventDispatcher {
         this.level = world;
     }
 
+    // Pluto start - Cache GameEvent lookups
+    private static final it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<net.minecraft.resources.ResourceLocation, org.bukkit.GameEvent> GAME_EVENT_CACHE = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
     public void post(GameEvent event, Vec3 emitterPos, GameEvent.Context emitter) {
         int i = event.getNotificationRadius();
         BlockPos blockposition = BlockPos.containing(emitterPos);
+        net.minecraft.resources.ResourceLocation key = BuiltInRegistries.GAME_EVENT.getKey(event);
+        org.bukkit.GameEvent bukkitEvent = GAME_EVENT_CACHE.get(key);
+        if (bukkitEvent == null) {
+            bukkitEvent = org.bukkit.GameEvent.getByKey(CraftNamespacedKey.fromMinecraft(key));
+            GAME_EVENT_CACHE.put(key, bukkitEvent);
+        }
+        if (bukkitEvent != null) {
         // CraftBukkit start
-        GenericGameEvent event1 = new GenericGameEvent(org.bukkit.GameEvent.getByKey(CraftNamespacedKey.fromMinecraft(BuiltInRegistries.GAME_EVENT.getKey(event))), CraftLocation.toBukkit(blockposition, this.level.getWorld()), (emitter.sourceEntity() == null) ? null : emitter.sourceEntity().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+        GenericGameEvent event1 = new GenericGameEvent(bukkitEvent, CraftLocation.toBukkit(blockposition, this.level.getWorld()), (emitter.sourceEntity() == null) ? null : emitter.sourceEntity().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+        // Pluto end - Cache GameEvent lookups
         this.level.getCraftServer().getPluginManager().callEvent(event1);
         if (event1.isCancelled()) {
             return;
         }
         i = event1.getRadius();
         // CraftBukkit end
+        } // Pluto - Cache GameEvent lookups
         int j = SectionPos.blockToSectionCoord(blockposition.getX() - i);
         int k = SectionPos.blockToSectionCoord(blockposition.getY() - i);
         int l = SectionPos.blockToSectionCoord(blockposition.getZ() - i);
