From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Fri, 17 Feb 2023 16:31:50 -0800
Subject: [PATCH] Optimize map id lookups


diff --git a/src/main/java/net/minecraft/world/item/MapItem.java b/src/main/java/net/minecraft/world/item/MapItem.java
index 586852e347cfeb6e52d16e51b3f193e814036e81..86e7ee5e486af227cd10b17b931b5b2f2f83bc18 100644
--- a/src/main/java/net/minecraft/world/item/MapItem.java
+++ b/src/main/java/net/minecraft/world/item/MapItem.java
@@ -73,7 +73,14 @@ public class MapItem extends ComplexItem {
     public static Integer getMapId(ItemStack stack) {
         CompoundTag nbttagcompound = stack.getTag();
 
-        return nbttagcompound != null && nbttagcompound.contains("map", 99) ? nbttagcompound.getInt("map") : null; // CraftBukkit - make new maps for no tag // Paper - don't return invalid ID
+        // Pluto start - Optimize map id lookups
+        if (nbttagcompound == null) return null;
+
+        net.minecraft.nbt.Tag tag = nbttagcompound.get("map");
+        if (tag == null) return null;
+
+        return tag instanceof net.minecraft.nbt.NumericTag map ? map.getAsInt() : 0;
+        // Pluto end - Optimize map id lookups
     }
 
     public static int createNewSavedData(Level world, int x, int z, int scale, boolean showIcons, boolean unlimitedTracking, ResourceKey<Level> dimension) {
