From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 13 Jul 2023 12:00:00 -0700
Subject: [PATCH] Toggle inventory item move event for droppers


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index b616444c252b0760b47b7a4abf4ea7ee3974a56f..d827dbf660786b84180d4b7218d4041221b00002 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -239,5 +239,10 @@ public class PlutoConfig {
         private void lessRandomDispensing() {
             lessRandomDispensing = getBoolean("blocks.dispenser.less-random-dispensing", lessRandomDispensing);
         }
+
+        public boolean disableDropperInventoryMoveEvent = false;
+        private void disableDropperInventoryMoveEvent() {
+            disableDropperInventoryMoveEvent = getBoolean("blocks.dropper.disable-move-event", disableDropperInventoryMoveEvent);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/DropperBlock.java b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
index 9a8a0fb958e8ec782111507bae957f854b2aac72..f6f22e4dca99c3a9be73224d8ee8f6416b83aa81 100644
--- a/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
@@ -73,6 +73,19 @@ public class DropperBlock extends DispenserBlock {
                         if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(world, pos, itemstack, i)) return; // Paper - Add BlockPreDispenseEvent
                         itemstack1 = DropperBlock.DISPENSE_BEHAVIOUR.dispense(sourceblock, itemstack);
                     } else {
+                        // Pluto start - Toggle inventory item move event for droppers
+                        if (world.plutoConfig().disableDropperInventoryMoveEvent) {
+                            itemstack1 = HopperBlockEntity.addItem(tileentitydispenser, iinventory, itemstack.copy().split(1), enumdirection.getOpposite());
+                            if (itemstack1.isEmpty()) {
+                                itemstack1 = itemstack.copy();
+                                itemstack1.shrink(1);
+                            } else {
+                                itemstack1 = itemstack.copy();
+                            }
+                            tileentitydispenser.setItem(i, itemstack1);
+                            return;
+                        }
+                        // Pluto end - Toggle inventory item move event for droppers
                         // CraftBukkit start - Fire event when pushing items into other inventories
                         CraftItemStack oitemstack = CraftItemStack.asCraftMirror(itemstack.copy().split(1));
 
