From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 5 Oct 2023 17:05:01 -0700
Subject: [PATCH] Option to remove excess vehicles on collision


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 9bea6cb2908cb9809832883ea500fd079ca31b3f..35345e3b31e7774f77ffb3135f7a73096b292d53 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -246,5 +246,17 @@ public class PlutoConfig {
             disableAllayGameEventListening = getBoolean("entities.allay.disable-game-event-listener", disableAllayGameEventListening);
             disableAllayDuplication = getBoolean("entities.allay.disable-duplication", disableAllayDuplication);
         }
+
+        public boolean removeExcessMinecarts = false;
+        public boolean removeExcessBoats = false;
+        public int excessMinecartsLimit = 10;
+        public int excessBoatsLimit = 10;
+
+        private void removeExcessVehicles() {
+            removeExcessMinecarts = getBoolean("entities.minecart.remove-excess.enabled", removeExcessMinecarts);
+            removeExcessBoats = getBoolean("entities.boat.remove-excess.enabled", removeExcessBoats);
+            excessMinecartsLimit = getInt("entities.minecart.remove-excess.limit", excessMinecartsLimit);
+            excessBoatsLimit = getInt("entities.boat.remove-excess.limit", excessBoatsLimit);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index a38ce400550893f63640e3bb5bb801ab40f06266..10d31ed526187b12d00194fd858c66f8e9f9efa9 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -378,6 +378,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
                 this.level().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, from, to));
             }
             // CraftBukkit end
+            int minecartCount = 0; // Pluto - Option to remove excess vehicles on collision
             if (this.getMinecartType() == AbstractMinecart.Type.RIDEABLE && this.getDeltaMovement().horizontalDistanceSqr() > 0.01D) {
                 List<Entity> list = this.level().getEntities((Entity) this, this.getBoundingBox().inflate(0.20000000298023224D, 0.0D, 0.20000000298023224D), EntitySelector.pushableBy(this));
 
@@ -386,6 +387,12 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
                     while (iterator.hasNext()) {
                         Entity entity = (Entity) iterator.next();
+                        // Pluto start - Option to remove excess vehicles on collision
+                        if (entity instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                            entity.discard();
+                            continue;
+                        }
+                        // Pluto end - Option to remove excess vehicles on collision
 
                         if (!(entity instanceof Player) && !(entity instanceof IronGolem) && !(entity instanceof AbstractMinecart) && !this.isVehicle() && !entity.isPassenger()) {
                             // CraftBukkit start
@@ -417,6 +424,12 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
                 while (iterator1.hasNext()) {
                     Entity entity1 = (Entity) iterator1.next();
+                    // Pluto start - Option to remove excess vehicles on collision
+                    if (entity1 instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                        entity1.discard();
+                        continue;
+                    }
+                    // Pluto end - Option to remove excess vehicles on collision
 
                     if (!this.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
                         // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index f1955afc8e367f80ead85bd5ad3b8d66c255565a..39b1faad13e45f860ee0b5639334a8afc3b66474 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -409,8 +409,15 @@ public class Boat extends VehicleEntity implements Leashable, VariantHolder<Boat
             boolean flag = !this.level().isClientSide && !(this.getControllingPassenger() instanceof Player);
             Iterator iterator = list.iterator();
 
+            int boatCount = 0; // Pluto - Option to remove excess vehicles on collision
             while (iterator.hasNext()) {
                 Entity entity = (Entity) iterator.next();
+                // Pluto start - Option to remove excess vehicles on collision
+                if (entity instanceof Boat && this.level().plutoConfig().removeExcessBoats && boatCount++ >= this.level().plutoConfig().excessBoatsLimit) {
+                    entity.discard();
+                    continue;
+                }
+                // Pluto end - Option to remove excess vehicles on collision
 
                 if (!entity.hasPassenger((Entity) this)) {
                     if (flag && this.getPassengers().size() < this.getMaxPassengers() && !entity.isPassenger() && this.hasEnoughSpaceFor(entity) && entity instanceof LivingEntity && !(entity instanceof WaterAnimal) && !(entity instanceof Player)) {
