From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 13 Jul 2023 12:00:00 -0700
Subject: [PATCH] Toggle inventory item move event for droppers


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 9e36ece3d99373e8dba0891db31374bfbafd6fc9..08ce93cc7318673913bad2255ecb9432002df964 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -244,5 +244,10 @@ public class PlutoConfig {
         private void lessRandomDispensing() {
             lessRandomDispensing = getBoolean("blocks.dispenser.less-random-dispensing", lessRandomDispensing);
         }
+
+        public boolean disableDropperInventoryMoveEvent = false;
+        private void disableDropperInventoryMoveEvent() {
+            disableDropperInventoryMoveEvent = getBoolean("blocks.dropper.disable-move-event", disableDropperInventoryMoveEvent);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/DropperBlock.java b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
index 801c4c120e98584bcf218a4ef9bd66d7d18c1097..6bf527aeb888f974c68c25392b201d75c5c074c2 100644
--- a/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
@@ -57,6 +57,19 @@ public class DropperBlock extends DispenserBlock {
                 if (iinventory == null) {
                     itemstack1 = DropperBlock.DISPENSE_BEHAVIOUR.dispense(sourceblock, itemstack);
                 } else {
+                    // Pluto start - Toggle inventory item move event for droppers
+                    if (world.plutoConfig().disableDropperInventoryMoveEvent) {
+                        itemstack1 = HopperBlockEntity.addItem(tileentitydispenser, iinventory, itemstack.copy().split(1), enumdirection.getOpposite());
+                        if (itemstack1.isEmpty()) {
+                            itemstack1 = itemstack.copy();
+                            itemstack1.shrink(1);
+                        } else {
+                            itemstack1 = itemstack.copy();
+                        }
+                        tileentitydispenser.setItem(i, itemstack1);
+                        return;
+                    }
+                    // Pluto end - Toggle inventory item move event for droppers
                     // CraftBukkit start - Fire event when pushing items into other inventories
                     CraftItemStack oitemstack = CraftItemStack.asCraftMirror(itemstack.copy().split(1));
 
