From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 5 Oct 2023 17:05:01 -0700
Subject: [PATCH] Option to remove excess vehicles on collision


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index c62344f9cd099e7e32c0dd6e0a9bdde4d0e6df14..346235c8c7d1583217f268c81544829063a1cd42 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -251,5 +251,17 @@ public class PlutoConfig {
             disableAllayGameEventListening = getBoolean("entities.allay.disable-game-event-listener", disableAllayGameEventListening);
             disableAllayDuplication = getBoolean("entities.allay.disable-duplication", disableAllayDuplication);
         }
+
+        public boolean removeExcessMinecarts = false;
+        public boolean removeExcessBoats = false;
+        public int excessMinecartsLimit = 10;
+        public int excessBoatsLimit = 10;
+
+        private void removeExcessVehicles() {
+            removeExcessMinecarts = getBoolean("entities.minecart.remove-excess.enabled", removeExcessMinecarts);
+            removeExcessBoats = getBoolean("entities.boat.remove-excess.enabled", removeExcessBoats);
+            excessMinecartsLimit = getInt("entities.minecart.remove-excess.limit", excessMinecartsLimit);
+            excessBoatsLimit = getInt("entities.boat.remove-excess.limit", excessBoatsLimit);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index d514ec1e4cbdc579c3a61533998437903afdc8b6..41bed500799131ae1f0443ad97cdef97447a0092 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -381,6 +381,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
                 this.level().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, from, to));
             }
             // CraftBukkit end
+            int minecartCount = 0; // Pluto - Option to remove excess vehicles on collision
             if (this.getMinecartType() == AbstractMinecart.Type.RIDEABLE && this.getDeltaMovement().horizontalDistanceSqr() > 0.01D) {
                 List<Entity> list = this.level().getEntities((Entity) this, this.getBoundingBox().inflate(0.20000000298023224D, 0.0D, 0.20000000298023224D), EntitySelector.pushableBy(this));
 
@@ -389,6 +390,12 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
                     while (iterator.hasNext()) {
                         Entity entity = (Entity) iterator.next();
+                        // Pluto start - Option to remove excess vehicles on collision
+                        if (entity instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                            entity.discard();
+                            continue;
+                        }
+                        // Pluto end - Option to remove excess vehicles on collision
 
                         if (!(entity instanceof Player) && !(entity instanceof IronGolem) && !(entity instanceof AbstractMinecart) && !this.isVehicle() && !entity.isPassenger()) {
                             // CraftBukkit start
@@ -420,6 +427,12 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
                 while (iterator1.hasNext()) {
                     Entity entity1 = (Entity) iterator1.next();
+                    // Pluto start - Option to remove excess vehicles on collision
+                    if (entity1 instanceof AbstractMinecart && this.level().plutoConfig().removeExcessMinecarts && minecartCount++ >= this.level().plutoConfig().excessMinecartsLimit) {
+                        entity1.discard();
+                        continue;
+                    }
+                    // Pluto end - Option to remove excess vehicles on collision
 
                     if (!this.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
                         // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index db6aa75d642f4a7258f197933671907faf79c8f2..9bb51b6c3f78e177aea4931484d203f6fee95549 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -412,8 +412,15 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
             boolean flag = !this.level().isClientSide && !(this.getControllingPassenger() instanceof Player);
             Iterator iterator = list.iterator();
 
+            int boatCount = 0; // Pluto - Option to remove excess vehicles on collision
             while (iterator.hasNext()) {
                 Entity entity = (Entity) iterator.next();
+                // Pluto start - Option to remove excess vehicles on collision
+                if (entity instanceof Boat && this.level().plutoConfig().removeExcessBoats && boatCount++ >= this.level().plutoConfig().excessBoatsLimit) {
+                    entity.discard();
+                    continue;
+                }
+                // Pluto end - Option to remove excess vehicles on collision
 
                 if (!entity.hasPassenger((Entity) this)) {
                     if (flag && this.getPassengers().size() < this.getMaxPassengers() && !entity.isPassenger() && this.hasEnoughSpaceFor(entity) && entity instanceof LivingEntity && !(entity instanceof WaterAnimal) && !(entity instanceof Player)) {
