From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:37:13 -0800
Subject: [PATCH] Toggle for nether portal generation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 82eda2ec5660725b63929cafc60726e9b1ce17ae..0ea8b59e4d4b17d40f2d511ce1bc206aa796abce 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -270,5 +270,10 @@ public class PlutoConfig {
         private void disableBlockGenerationFromFluids() {
             disableBlockGenerationFromFluids = getBoolean("blocks.fluids.disable-block-generation", disableBlockGenerationFromFluids);
         }
+
+        public boolean disableNetherPortalGeneration = false;
+        private void disableNetherPortalGeneration() {
+            disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
index bdf432850f8228ed42b09afea79ecc5c1055c930..93dc070e27d7a1ebcc937a5c5dd3c328afe17161 100644
--- a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
@@ -157,7 +157,7 @@ public abstract class BaseFireBlock extends Block {
     protected void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify, @javax.annotation.Nullable net.minecraft.world.item.context.UseOnContext context) {
         // Paper end - UseOnContext param
         if (!oldState.is(state.getBlock())) {
-            if (BaseFireBlock.inPortalDimension(world)) {
+            if (!world.plutoConfig().disableNetherPortalGeneration && BaseFireBlock.inPortalDimension(world)) { // Pluto - Toggle for nether portal generation
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(world, pos, Direction.Axis.X);
 
                 if (optional.isPresent()) {
