From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:57:47 -0800
Subject: [PATCH] Configurable entity activation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 704a856f3af76a6d8ce09b8f30586a32b2f5fb1c..484d77bbd05afe3ef895d0bc4b664d0a827bff39 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -268,5 +268,12 @@ public class PlutoConfig {
         private void disableNetherPortalGeneration() {
             disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
         }
+
+        public int entityActivationInterval = 1;
+        public boolean disableEntityActivation = false;
+        private void entityActivationConfiguration() {
+            entityActivationInterval = Math.min(1, getInt("entities.global.activation-interval", entityActivationInterval));
+            disableEntityActivation = getBoolean("entities.global.disable-activation", disableEntityActivation);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 2a8ffcdd262ea73844500846c6401cdda7153d61..c9299bfe82796dc9a8001c9ab54966a8a3ce58cd 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -712,7 +712,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
                 gameprofilerfiller.pop();
             }
 
-            org.spigotmc.ActivationRange.activateEntities(this); // Spigot
+            if (!plutoConfig().disableEntityActivation && MinecraftServer.currentTick % plutoConfig().entityActivationInterval == 0) org.spigotmc.ActivationRange.activateEntities(this); // Spigot // Pluto - Configurable entity activation
             this.timings.entityTick.startTiming(); // Spigot
             this.entityTickList.forEach((entity) -> {
                 entity.activatedPriorityReset = false; // Pufferfish - DAB
