From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:37:13 -0800
Subject: [PATCH] Toggle for nether portal generation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 82eda2ec5660725b63929cafc60726e9b1ce17ae..0ea8b59e4d4b17d40f2d511ce1bc206aa796abce 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -270,5 +270,10 @@ public class PlutoConfig {
         private void disableBlockGenerationFromFluids() {
             disableBlockGenerationFromFluids = getBoolean("blocks.fluids.disable-block-generation", disableBlockGenerationFromFluids);
         }
+
+        public boolean disableNetherPortalGeneration = false;
+        private void disableNetherPortalGeneration() {
+            disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
index 8b5918dc07f17ae5001c03dc743128fd9520b819..f02b667887567cb9a64b193d52ca5ed1b1431aec 100644
--- a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
@@ -151,7 +151,7 @@ public abstract class BaseFireBlock extends Block {
     @Override
     protected void onPlace(BlockState iblockdata, Level world, BlockPos blockposition, BlockState iblockdata1, boolean flag, UseOnContext context) { // CraftBukkit - context
         if (!iblockdata1.is(iblockdata.getBlock())) {
-            if (BaseFireBlock.inPortalDimension(world)) {
+            if (!world.plutoConfig().disableNetherPortalGeneration && BaseFireBlock.inPortalDimension(world)) { // Pluto - Toggle for nether portal generation
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(world, blockposition, Direction.Axis.X);
 
                 if (optional.isPresent()) {
