From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:37:13 -0800
Subject: [PATCH] Toggle for nether portal generation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 7ec0458650955ffe3ed12bc14569ac7d91d9b93a..f5f460a112f497f90a3a3ac33c0c33cad3827a5b 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -268,5 +268,10 @@ public class PlutoConfig {
         private void disableBlockGenerationFromFluids() {
             disableBlockGenerationFromFluids = getBoolean("blocks.fluids.disable-block-generation", disableBlockGenerationFromFluids);
         }
+
+        public boolean disableNetherPortalGeneration = false;
+        private void disableNetherPortalGeneration() {
+            disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
index bd7cb2ddcac5cdc1d61899d12fc3383fd00598b1..e2664dfc00055fdaf5b6bc6614bf015f91a9bdb8 100644
--- a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
@@ -157,7 +157,7 @@ public abstract class BaseFireBlock extends Block {
     public void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify, net.minecraft.world.item.context.UseOnContext context) {
         // Paper end - UseOnContext param
         if (!oldState.is(state.getBlock())) {
-            if (BaseFireBlock.inPortalDimension(world)) {
+            if (!world.plutoConfig().disableNetherPortalGeneration && BaseFireBlock.inPortalDimension(world)) { // Pluto - Toggle for nether portal generation
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(world, pos, Direction.Axis.X);
 
                 if (optional.isPresent()) {
