From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:37:13 -0800
Subject: [PATCH] Toggle for nether portal generation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 958ecf15752e546cc51479168c1cbabaa3c4b70e..22ac1ba801306fe21eb95d51b389bc695410c987 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -266,5 +266,10 @@ public class PlutoConfig {
         private void disableBlockGenerationFromFluids() {
             disableBlockGenerationFromFluids = getBoolean("blocks.fluids.disable-block-generation", disableBlockGenerationFromFluids);
         }
+
+        public boolean disableNetherPortalGeneration = false;
+        private void disableNetherPortalGeneration() {
+            disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
index 116dc2ac4c0176244eb8fc22a471d2b9ede72822..5bff5440cece14da3f9b88358f76e75b073e07d9 100644
--- a/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BaseFireBlock.java
@@ -153,7 +153,7 @@ public abstract class BaseFireBlock extends Block {
     public void onPlace(BlockState iblockdata, Level world, BlockPos blockposition, BlockState iblockdata1, boolean flag, net.minecraft.world.item.context.UseOnContext itemActionContext) {
         // Paper end
         if (!iblockdata1.is(iblockdata.getBlock())) {
-            if (BaseFireBlock.inPortalDimension(world)) {
+            if (!world.plutoConfig().disableNetherPortalGeneration && BaseFireBlock.inPortalDimension(world)) { // Pluto - Toggle for nether portal generation
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(world, blockposition, Direction.Axis.X);
 
                 if (optional.isPresent()) {
