From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:57:47 -0800
Subject: [PATCH] Configurable entity activation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 704a856f3af76a6d8ce09b8f30586a32b2f5fb1c..484d77bbd05afe3ef895d0bc4b664d0a827bff39 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -268,5 +268,12 @@ public class PlutoConfig {
         private void disableNetherPortalGeneration() {
             disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
         }
+
+        public int entityActivationInterval = 1;
+        public boolean disableEntityActivation = false;
+        private void entityActivationConfiguration() {
+            entityActivationInterval = Math.min(1, getInt("entities.global.activation-interval", entityActivationInterval));
+            disableEntityActivation = getBoolean("entities.global.disable-activation", disableEntityActivation);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 59f90878ec9f110d78bce998deb1dcd2830b0f08..7e9888a9f047205d8be9b437c44ea5a6c8abcc68 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -790,6 +790,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                 gameprofilerfiller.pop();
             }
 
+            if (!plutoConfig().disableEntityActivation && MinecraftServer.currentTick % plutoConfig().entityActivationInterval == 0) // Pluto - Configurable entity activation
             org.spigotmc.ActivationRange.activateEntities(this); // Spigot
             this.entityTickList.forEach((entity) -> {
                 entity.activatedPriorityReset = false; // Pufferfish - DAB
