From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 13:57:47 -0800
Subject: [PATCH] Configurable entity activation

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 0ea8b59e4d4b17d40f2d511ce1bc206aa796abce..afde9883093366b82c629a9691b43a2adc50d421 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -275,5 +275,12 @@ public class PlutoConfig {
         private void disableNetherPortalGeneration() {
             disableNetherPortalGeneration = getBoolean("blocks.fire.disable-nether-portal-generation", disableNetherPortalGeneration);
         }
+
+        public int entityActivationInterval = 1;
+        public boolean disableEntityActivation = false;
+        private void entityActivationConfiguration() {
+            entityActivationInterval = Math.min(1, getInt("entities.global.activation-interval", entityActivationInterval));
+            disableEntityActivation = getBoolean("entities.global.disable-activation", disableEntityActivation);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index dbe612427b894df9da1335b94163ba9b89b090c0..9536fee892a4147892dc7eabdd9179405bd942f9 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -891,7 +891,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 gameprofilerfiller.pop();
             }
 
-            org.spigotmc.ActivationRange.activateEntities(this); // Spigot
+            if (!plutoConfig().disableEntityActivation && MinecraftServer.currentTick % plutoConfig().entityActivationInterval == 0) org.spigotmc.ActivationRange.activateEntities(this); // Spigot // Pluto - Configurable entity activation
             this.timings.entityTick.startTiming(); // Spigot
             this.entityTickList.forEach((entity) -> {
                 entity.activatedPriorityReset = false; // Pufferfish - DAB
