From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 14:34:32 -0800
Subject: [PATCH] Configuration for item merging

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index afde9883093366b82c629a9691b43a2adc50d421..ea2d0e800724569d8e641989dc9205a5ac24f7a6 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -282,5 +282,14 @@ public class PlutoConfig {
             entityActivationInterval = Math.min(1, getInt("entities.global.activation-interval", entityActivationInterval));
             disableEntityActivation = getBoolean("entities.global.disable-activation", disableEntityActivation);
         }
+
+        public boolean disableItemMergeCheckWhileMoving = false;
+        public int itemMergeMovingInterval = 2;
+        public int itemMergeStaticInterval = 40;
+        private void itemMergeConfiguration() {
+            disableItemMergeCheckWhileMoving = getBoolean("entities.item.disable-merge-check-when-moving", disableItemMergeCheckWhileMoving);
+            itemMergeMovingInterval = getInt("entities.item.merge-check-moving-interval", itemMergeMovingInterval);
+            itemMergeStaticInterval = getInt("entities.item.merge-check-static-interval", itemMergeStaticInterval);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index c7f06c3cfb737bd17a706798bf9cf0e1af5f0cc0..7504b52c6ec85cf5106176a3726ea616b2c1ec24 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -198,10 +198,12 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 }
             }
 
-            boolean flag = Mth.floor(this.xo) != Mth.floor(this.getX()) || Mth.floor(this.yo) != Mth.floor(this.getY()) || Mth.floor(this.zo) != Mth.floor(this.getZ());
-            int i = flag ? 2 : 40;
+            // Pluto start - Configuration for item merging
+            boolean flag = this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ(); // Flooring makes the movement check inconsistent
+            int i = flag ? this.level().plutoConfig().itemMergeMovingInterval : this.level().plutoConfig().itemMergeStaticInterval;
 
-            if (this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable()) {
+            if ((!flag || !this.level().plutoConfig().disableItemMergeCheckWhileMoving) && this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable()) {
+            // Pluto end - Configuration for item merging
                 this.mergeWithNeighbours();
             }
 
