From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 23 Jan 2024 14:34:32 -0800
Subject: [PATCH] Configuration for item merging

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 4ba7d0669804bc84a24e0c2bb8c178322b79a34a..1505bb360194172a485447fcabda8b4ce4e25c4b 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -280,5 +280,14 @@ public class PlutoConfig {
             entityActivationInterval = Math.min(1, getInt("entities.global.activation-interval", entityActivationInterval));
             disableEntityActivation = getBoolean("entities.global.disable-activation", disableEntityActivation);
         }
+
+        public boolean disableItemMergeCheckWhileMoving = false;
+        public int itemMergeMovingInterval = 2;
+        public int itemMergeStaticInterval = 40;
+        private void itemMergeConfiguration() {
+            disableItemMergeCheckWhileMoving = getBoolean("entities.item.disable-merge-check-when-moving", disableItemMergeCheckWhileMoving);
+            itemMergeMovingInterval = getInt("entities.item.merge-check-moving-interval", itemMergeMovingInterval);
+            itemMergeStaticInterval = getInt("entities.item.merge-check-static-interval", itemMergeStaticInterval);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index eb0d6238588efa35fa868f26290547574a08eca2..0b8ddbbf7c9aa6d16e374237078978e67d71fdc6 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -197,10 +197,12 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 }
             }
 
-            boolean flag = Mth.floor(this.xo) != Mth.floor(this.getX()) || Mth.floor(this.yo) != Mth.floor(this.getY()) || Mth.floor(this.zo) != Mth.floor(this.getZ());
-            int i = flag ? 2 : 40;
+            // Pluto start - Configuration for item merging
+            boolean flag = this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ(); // Flooring makes the movement check inconsistent
+            int i = flag ? this.level().plutoConfig().itemMergeMovingInterval : this.level().plutoConfig().itemMergeStaticInterval;
 
-            if (this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable()) {
+            if ((!flag || !this.level().plutoConfig().disableItemMergeCheckWhileMoving) && this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable()) {
+            // Pluto end - Configuration for item merging
                 this.mergeWithNeighbours();
             }
 
