From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 25 Jan 2024 18:18:58 -0800
Subject: [PATCH] Option to disable TNT chain reactions

Pulled from my old 1.8/1.12 forks

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 8c2790fb1e320b63668d791d42f4465401721151..156035b25e316fb5be83812b0b28be8ddf01e2e9 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -340,5 +340,10 @@ public class PlutoConfig {
             spawnerRequiredPlayerRange = (short) Math.min(Short.MAX_VALUE, getInt("blocks.spawner.required-player-range", spawnerRequiredPlayerRange));
             spawnerSpawnRange = (short) Math.min(Short.MAX_VALUE, getInt("blocks.spawner.spawn-range", spawnerSpawnRange));
         }
+
+        public boolean disableTntChainReaction = false;
+        private void disableTntChainReaction() {
+            disableTntChainReaction = getBoolean("blocks.tnt.disable-chain-reaction", disableTntChainReaction);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/TntBlock.java b/src/main/java/net/minecraft/world/level/block/TntBlock.java
index 5fa5e8c838720eb1491aea73d462f4bc7d779956..b1b64418740b864481df1a23e25e038ebbf76df0 100644
--- a/src/main/java/net/minecraft/world/level/block/TntBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TntBlock.java
@@ -88,7 +88,7 @@ public class TntBlock extends Block {
 
     @Override
     public void wasExploded(Level world, BlockPos pos, Explosion explosion) {
-        if (!world.isClientSide) {
+        if (!world.isClientSide && !world.plutoConfig().disableTntChainReaction) { // Pluto - Option to disable TNT chain reactions
             // Paper start - TNTPrimeEvent
             org.bukkit.block.Block tntBlock = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
             org.bukkit.entity.Entity source = explosion.source != null ? explosion.source.getBukkitEntity() : null;
