From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 15 Feb 2024 22:21:04 -0800
Subject: [PATCH] Option to disable flapping movement game event


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index e7c48c0a8773390844d2663072ec5de54d8e4e83..ac22413df335736439996c27799d71ab04b1f0ac 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -342,5 +342,10 @@ public class PlutoConfig {
         private void disableTntChainReaction() {
             disableTntChainReaction = getBoolean("blocks.tnt.disable-chain-reaction", disableTntChainReaction);
         }
+
+        public boolean disableFlapEvents = false;
+        private void disableFlapEvents() {
+            disableFlapEvents = getBoolean("entities.global.disable-flapping-game-event", false);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 8492421ed2186c0eab517a67f3140b9988f65250..75a80820f9ab4dfa7236645cebbc836fac7b45b2 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1369,7 +1369,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level.plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
