From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 15 Feb 2024 22:21:04 -0800
Subject: [PATCH] Option to disable flapping movement game event


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 156035b25e316fb5be83812b0b28be8ddf01e2e9..84b9793df4b4fd8109e00fa5fec67061ab7e7cf1 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -345,5 +345,10 @@ public class PlutoConfig {
         private void disableTntChainReaction() {
             disableTntChainReaction = getBoolean("blocks.tnt.disable-chain-reaction", disableTntChainReaction);
         }
+
+        public boolean disableFlapEvents = false;
+        private void disableFlapEvents() {
+            disableFlapEvents = getBoolean("entities.global.disable-flapping-game-event", false);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0f2af6968f6bb47f5ef7e3f63cbfdc2f6df53999..ea04ca9664d34c41c6431494d3c5a3c8e2c96709 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1366,7 +1366,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level.plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
