From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 15 Feb 2024 22:21:04 -0800
Subject: [PATCH] Option to disable flapping movement game event


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index a4f3002832eb103ef62123f67d2e29e3dc1e8b51..674f3aaaa6916bbb143cbbe506fb88792cb21504 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -349,5 +349,10 @@ public class PlutoConfig {
         private void disableTntChainReaction() {
             disableTntChainReaction = getBoolean("blocks.tnt.disable-chain-reaction", disableTntChainReaction);
         }
+
+        public boolean disableFlapEvents = false;
+        private void disableFlapEvents() {
+            disableFlapEvents = getBoolean("entities.global.disable-flapping-game-event", false);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 5c5c8fbd562b46f8699be97353447eaab36c007a..3ab5648d71391f91b452f07a9687071136e85bd5 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1379,7 +1379,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level.plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
