From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Thu, 15 Feb 2024 22:21:04 -0800
Subject: [PATCH] Option to disable flapping movement game event


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index b1675a7fddd2063a2bb66c683dc49044ab6e8201..8a7780669a09666a4b1fad3297ac2b8082f644fa 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -347,5 +347,10 @@ public class PlutoConfig {
         private void disableTntChainReaction() {
             disableTntChainReaction = getBoolean("blocks.tnt.disable-chain-reaction", disableTntChainReaction);
         }
+
+        public boolean disableFlapEvents = false;
+        private void disableFlapEvents() {
+            disableFlapEvents = getBoolean("entities.global.disable-flapping-game-event", false);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e12cbbe9eea3f243808fad547c91bd546fd4bbc8..82ed95fabf106bfa00b93d66aa4a669b8c7e6f96 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1367,7 +1367,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     protected void processFlappingMovement() {
         if (this.isFlapping()) {
             this.onFlap();
-            if (this.getMovementEmission().emitsEvents()) {
+            if (!this.level.plutoConfig().disableFlapEvents && this.getMovementEmission().emitsEvents()) { // Pluto - Option to disable flapping movement game event
                 this.gameEvent(GameEvent.FLAP);
             }
         }
