From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 18 Feb 2024 03:23:27 -0800
Subject: [PATCH] Configuration for raid ticking

A value below 1 will stop all raid ticking

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index f867010621a22ca310b35d78704ac0feceacf706..db54c74fcf72e64ad4b2aa1a9cc9a010329cb720 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -357,5 +357,10 @@ public class PlutoConfig {
         private void fluidTicking() {
             fluidTickingModulo = getInt("blocks.fluids.fluid-ticking-modulo", fluidTickingModulo);
         }
+
+        public int raidTickingModulo = 1;
+        private void raidTicking() {
+            raidTickingModulo = getInt("misc.raids.raid-ticking-modulo", raidTickingModulo);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index d8ddf2ffbedb6479c3ffa6b82e8137219301f708..93ade924c9fbb2b7b614ba579044401518385f68 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -766,6 +766,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
         gameprofilerfiller.popPush("raid");
         if (flag) {
+            if (plutoConfig().raidTickingModulo > 0 && (plutoConfig().raidTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().raidTickingModulo == 0)) // Pluto - Configuration for raid ticking
             this.raids.tick();
         }
 
