From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 18 Feb 2024 05:09:30 -0800
Subject: [PATCH] Configuration for block event ticking

Block event ticking cannot be disabled currently.

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 488a25e1cec14882d09495c3a28c19fdfe8a3862..219f5936a0f08bd08055f03b7924a9d4903c46cc 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -369,5 +369,11 @@ public class PlutoConfig {
         private void raidTicking() {
             raidTickingModulo = getInt("misc.raids.raid-ticking-modulo", raidTickingModulo);
         }
+
+        public int blockEventTickingModulo = 1;
+        private void blockEventTicking() {
+            // TODO: Allow for disabling block even ticking
+            blockEventTickingModulo = Math.min(1, getInt("blocks.global.block-event-ticking-modulo", blockEventTickingModulo));
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 0f3b49be81ac37ec2eace70cdc9fa2154d044f6d..0e50b19c9ca1fc627eb0240b2b132ae5319884e0 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -694,6 +694,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         gameprofilerfiller.popPush("blockEvents");
         if (flag) {
             this.timings.doSounds.startTiming(); // Spigot
+            if (plutoConfig().blockEventTickingModulo > 0 && (plutoConfig().blockEventTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().blockEventTickingModulo == 0)) // Pluto - Configuration for block event ticking
             this.runBlockEvents();
             this.timings.doSounds.stopTiming(); // Spigot
         }
