From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Sun, 18 Feb 2024 05:09:30 -0800
Subject: [PATCH] Configuration for block event ticking

Block event ticking cannot be disabled currently.

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 422c8748c82a344768624aae90a658ac1bb2b790..b150822b9d6da6b69e4f22de4231fd473cd43bfa 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -365,5 +365,11 @@ public class PlutoConfig {
         private void raidTicking() {
             raidTickingModulo = getInt("misc.raids.raid-ticking-modulo", raidTickingModulo);
         }
+
+        public int blockEventTickingModulo = 1;
+        private void blockEventTicking() {
+            // TODO: Allow for disabling block even ticking
+            blockEventTickingModulo = Math.min(1, getInt("blocks.global.block-event-ticking-modulo", blockEventTickingModulo));
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 42f1870354168af1d11048d9f10c5da18db4bccc..b7937696c32d0fc1b380f1cf88fd91c2d21ce70b 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -873,6 +873,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         gameprofilerfiller.popPush("blockEvents");
         if (flag) {
             this.timings.doSounds.startTiming(); // Spigot
+            if (plutoConfig().blockEventTickingModulo > 0 && (plutoConfig().blockEventTickingModulo == 1 || MinecraftServer.currentTick % plutoConfig().blockEventTickingModulo == 0)) // Pluto - Configuration for block event ticking
             this.runBlockEvents();
             this.timings.doSounds.stopTiming(); // Spigot
         }
