From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 21 Feb 2024 13:28:34 -0800
Subject: [PATCH] Configuration for entity inside block check frequency


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 1ec90416909889276645eb6daef7292e7fee60af..fbb52a4c68f8bfadd0132536adef680b98aef66e 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -387,5 +387,10 @@ public class PlutoConfig {
         private void entityPushing() {
             entityPushingModulo = getInt("entities.global.pushing-modulo", entityPushingModulo);
         }
+
+        public int entityInsideBlockCheck = 1;
+        private void entityInsideBlockCheck() {
+            entityInsideBlockCheck = getInt("entities.global.inside-block-check-modulo", entityInsideBlockCheck);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e95a75ecf446c3d1c603d726d228fc413ea73e39..7a7aac520ff4dd234f93117c9827080c2b36a218 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1355,6 +1355,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             }
 
             this.movementThisTick.add(new Entity.Movement(lastRenderPos, pos));
+            if (this.level().plutoConfig().entityInsideBlockCheck > 0 && (this.level().plutoConfig().entityInsideBlockCheck == 1 || this.tickCount % this.level().plutoConfig().entityInsideBlockCheck == 0)) // Pluto - Configuration for entity inside block check frequency
             this.checkInsideBlocks(this.movementThisTick, this.blocksInside);
             boolean flag = Iterables.any(this.blocksInside, (iblockdata1) -> {
                 return iblockdata1.is(BlockTags.FIRE) || iblockdata1.is(Blocks.LAVA);
