From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 21 Feb 2024 13:28:34 -0800
Subject: [PATCH] Configuration for entity inside block check frequency


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index d36477fb53709d160edb8c7bf64731d7347e60a0..d130bea050fbb861d26900ac89c5c9b56d3dfb99 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -394,5 +394,10 @@ public class PlutoConfig {
         private void entityPushing() {
             entityPushingModulo = getInt("entities.global.pushing-modulo", entityPushingModulo);
         }
+
+        public int entityInsideBlockCheck = 1;
+        private void entityInsideBlockCheck() {
+            entityInsideBlockCheck = getInt("entities.global.inside-block-check-modulo", entityInsideBlockCheck);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0aa71448a6cd55e2c5a50d785012c3a0fc38bff1..2822e923606a906749821e1fff6815bdb8575e1e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1278,6 +1278,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     }
                 }
 
+                if (this.level().plutoConfig().entityInsideBlockCheck > 0 && (this.level().plutoConfig().entityInsideBlockCheck == 1 || this.tickCount % this.level().plutoConfig().entityInsideBlockCheck == 0)) // Pluto - Configuration for entity inside block check frequency
                 this.tryCheckInsideBlocks();
                 float f = this.getBlockSpeedFactor();
 
