From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 20 Aug 2024 00:47:59 -0700
Subject: [PATCH] Add method to obtain current entity id


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index a13bf7cb89400e00b1c05ea97d7dee6f00bd8421..dc466f9508a2b4fe6eff1a9c10f105cfd3283159 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -5346,4 +5346,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return ((net.minecraft.server.level.ServerLevel) this.level).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    // Pluto start - Add method to obtain current entity id
+    public static int currentEntityId() {
+        return ENTITY_COUNTER.get();
+    }
+    // Pluto end - Add method to obtain current entity id
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index da306aab94697a86ac052f4536c9eff1ff23a92f..7e06cd0d44bd1be4582b14415f58f80dab6035f1 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -619,6 +619,13 @@ public final class CraftMagicNumbers implements UnsafeValues {
         return net.minecraft.world.entity.Entity.nextEntityId();
     }
 
+    // Pluto start - Add method to obtain current entity id
+    @Override
+    public int currentEntityId() {
+        return net.minecraft.world.entity.Entity.currentEntityId();
+    }
+    // Pluto end - Add method to obtain current entity id
+
     @Override
     public String getMainLevelName() {
         return ((net.minecraft.server.dedicated.DedicatedServer) net.minecraft.server.MinecraftServer.getServer()).getProperties().levelName;
