From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 20 Aug 2024 00:47:59 -0700
Subject: [PATCH] Add method to obtain current entity id


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 698a1bdbb8dd2f5a480feacd9a138db654ee6163..dff19bd87a5b5807c716ea73fd094a6a2fdd0cdc 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -5067,4 +5067,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return ((net.minecraft.server.level.ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
     }
     // Paper end - Expose entity id counter
+
+    // Pluto start - Add method to obtain current entity id
+    public static int currentEntityId() {
+        return ENTITY_COUNTER.get();
+    }
+    // Pluto end - Add method to obtain current entity id
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 83766d4a8e7c5bfd32689430c70b24741368ec00..99a57ed3546e01c571faf1f5f861c16969a99a43 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -597,6 +597,13 @@ public final class CraftMagicNumbers implements UnsafeValues {
         return net.minecraft.world.entity.Entity.nextEntityId();
     }
 
+    // Pluto start - Add method to obtain current entity id
+    @Override
+    public int currentEntityId() {
+        return net.minecraft.world.entity.Entity.currentEntityId();
+    }
+    // Pluto end - Add method to obtain current entity id
+
     @Override
     public String getMainLevelName() {
         return ((net.minecraft.server.dedicated.DedicatedServer) net.minecraft.server.MinecraftServer.getServer()).getProperties().levelName;
