From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 2 Oct 2024 16:37:57 -0700
Subject: [PATCH] Option to skip firing level events on failed dispense


diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 725fcababb084f73eb5a7bb7515c24e9817af2e4..88047c33e85893f3eed5fe5f1a8dfecd28d1c9b0 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -419,5 +419,10 @@ public class PlutoConfig {
                 type.spawnerTTL = this.config.getInt(path, this.worldDefaults.getInt(path, -1));
             }
         }
+
+        public boolean skipFailedDispenseLevelEvent = false;
+        private void skipFailedDispenseLevelEvent() {
+            skipFailedDispenseLevelEvent = getBoolean("blocks.global.skip-failed-block-dispense-level-event", skipFailedDispenseLevelEvent);
+        }
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 4632eb883e9f5efde520ee543bcad25827c0da2c..0cfa6729a51faa33d6343b56284310d766eb0c60 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2208,7 +2208,7 @@ public class CraftEventFactory {
     public static boolean handleBlockFailedDispenseEvent(ServerLevel serverLevel, BlockPos pos) {
         org.bukkit.block.Block block = CraftBlock.at(serverLevel, pos);
         io.papermc.paper.event.block.BlockFailedDispenseEvent event = new io.papermc.paper.event.block.BlockFailedDispenseEvent(block);
-        return event.callEvent();
+        return event.callEvent() && !serverLevel.plutoConfig().skipFailedDispenseLevelEvent; // Pluto - Option to skip firing level events on failed dispense
     }
 
     public static boolean handleBlockPreDispenseEvent(ServerLevel serverLevel, BlockPos pos, ItemStack itemStack, int slot) {
