From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 8 Oct 2024 23:21:15 -0700
Subject: [PATCH] Finish anything blocking before syncing configuration change

syncAfterConfigurationChange Has an insane rate of causing server crashes

diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 3e550f8e7cd4f4e16f499a8a2a4b95420270f07a..23f6f85a36540cb1ace4c7c8703c7cca963bc0bc 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -345,6 +345,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     private static void syncAfterConfigurationChange(ChannelFuture future) {
         try {
+            if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) net.minecraft.server.MinecraftServer.getServer().managedBlock(future::isDone); // Pluto - Finish anything blocking before syncing configuration change
             future.syncUninterruptibly();
         } catch (Exception exception) {
             if (exception instanceof ClosedChannelException) {
