From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 3 Sep 2024 16:13:43 -0700
Subject: [PATCH] Inline world height


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 7148e852c653be9ec60c0d35a9654cb85ffe97ab..2619573e4cf2a57007019fe31f4155c1695ac02e 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -685,6 +685,15 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
     // Paper end - optimise random ticking
 
+    // Pluto start - Inline world height
+    private final int minBuildHeight, maxBuildHeight, worldMinSection, worldMaxSection, height;
+    @Override public final int getMinBuildHeight() { return minBuildHeight; }
+    @Override public final int getMaxBuildHeight() { return maxBuildHeight; }
+    @Override public final int getMinSection() { return worldMinSection; }
+    @Override public final int getMaxSection() { return worldMaxSection; }
+    @Override public final int getHeight() { return height; }
+    // Pluto end - Inline world height
+
     protected Level(WritableLevelData worlddatamutable, ResourceKey<Level> resourcekey, RegistryAccess iregistrycustom, Holder<DimensionType> holder, Supplier<ProfilerFiller> supplier, boolean flag, boolean flag1, long i, int j, org.bukkit.generator.ChunkGenerator gen, org.bukkit.generator.BiomeProvider biomeProvider, org.bukkit.World.Environment env, java.util.function.Function<org.spigotmc.SpigotWorldConfig, io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, java.util.concurrent.Executor executor) { // Paper - create paper world config & Anti-Xray
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
@@ -705,6 +714,14 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.dimensionTypeRegistration = holder;
         final DimensionType dimensionmanager = (DimensionType) holder.value();
 
+        // Pluto start - Inline world height
+        this.minBuildHeight = dimensionmanager.minY();
+        this.height = dimensionmanager.height();
+        this.maxBuildHeight = dimensionmanager.minY() + dimensionmanager.height();
+        this.worldMinSection = SectionPos.blockToSectionCoord(dimensionmanager.minY());
+        this.worldMaxSection = SectionPos.blockToSectionCoord((dimensionmanager.minY() + dimensionmanager.height()) - 1) + 1;
+        // Pluto end - Inline world height
+
         this.dimension = resourcekey;
         this.isClientSide = flag;
         if (dimensionmanager.coordinateScale() != 1.0D) {
