From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 20 Nov 2024 18:26:50 -0800
Subject: [PATCH] Expose some sound related functions


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 76d2ede135f33fda260896d3da6937a0e8cd19a7..03de8f15efb45768e28c5dc5ea5d8a1d81f0cdd9 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2901,6 +2901,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
     }
 
+    // Pluto start - Expose some sound related functions
+    public float getSoundVolume0() { return getSoundVolume(); }
+    public float getVoicePitch0() { return getVoicePitch(); }
+    // Pluto end - Expose some sound related functions
+
     public float getSoundVolume() {
         return 1.0F;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index d0c409f4efad289e3e325f44b500fc72589d89d4..3cff3066944a87b411792b72483e4e7f341fc0c6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1211,4 +1211,17 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return this.getHandle().canUseSlot(org.bukkit.craftbukkit.CraftEquipmentSlot.getNMS(slot));
     }
     // Paper end - Expose canUseSlot
+
+    // Pluto start - Expose some sound related functions
+    @Override
+    public float getSoundVolume() {
+        return this.getHandle().getSoundVolume0();
+    }
+
+    @Override
+    public float getVoicePitch() {
+        return this.getHandle().getVoicePitch0();
+    }
+    // Pluto end - Expose some sound related functions
+
 }
