--- a/net/minecraft/world/level/biome/Biome.java
+++ b/net/minecraft/world/level/biome/Biome.java
@@ -137,7 +_,7 @@
 
         if (level.isInsideBuildHeight(pos.getY()) && level.getBrightness(LightLayer.BLOCK, pos) < 10) {
             BlockState blockState = level.getBlockState(pos);
-            FluidState fluidState = level.getFluidState(pos);
+            FluidState fluidState = blockState.getFluidState(); // Pluto start - Decrease chunk/block lookups
             if (fluidState.is(Fluids.WATER) && blockState.getBlock() instanceof LiquidBlock) {
                 if (!checkNeighbors) {
                     return true;
