--- a/net/minecraft/world/level/biome/Biome.java
+++ b/net/minecraft/world/level/biome/Biome.java
@@ -131,7 +_,7 @@
         } else {
             if (level.isInsideBuildHeight(water.getY()) && level.getBrightness(LightLayer.BLOCK, water) < 10) {
                 BlockState blockState = level.getBlockState(water);
-                FluidState fluidState = level.getFluidState(water);
+                FluidState fluidState = blockState.getFluidState(); // Pluto start - Decrease chunk/block lookups
                 if (fluidState.getType() == Fluids.WATER && blockState.getBlock() instanceof LiquidBlock) {
                     if (!mustBeAtEdge) {
                         return true;
