--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -658,7 +_,7 @@
         protected boolean isValidTarget(final LevelReader level, final BlockPos pos) {
             BlockState state = level.getBlockState(pos);
             if (state.is(BlockTags.SUPPORTS_CROPS) && this.wantsToRaid && !this.canRaid) {
-                state = level.getBlockState(pos.above());
+                state = level.pluto$getBlockState(pos.getX(), pos.getY() + 1, pos.getZ()); // Pluto - Memory allocation optimisations
                 if (state.getBlock() instanceof CarrotBlock carrotBlock && carrotBlock.isMaxAge(state)) {
                     this.canRaid = true;
                     return true;
