--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -500,6 +_,7 @@
         this.updateSwingTime();
         this.yHeadRot = this.getYRot();
         this.setSpeed((float)this.getAttributeValue(Attributes.MOVEMENT_SPEED));
+        if (this.level().plutoConfig().playersDoNearbyEntityLookups) // Pluto - Toggle players doing nearby entity lookups
         if (this.getHealth() > 0.0F && !this.isSpectator()) {
             AABB aabb;
             if (this.isPassenger() && !this.getVehicle().isRemoved()) {
@@ -1122,8 +_,10 @@
                             if (flag3) {
                                 float f5 = 1.0F + (float)this.getAttributeValue(Attributes.SWEEPING_DAMAGE_RATIO) * f;
 
+                                int entityCount = 1; // Pluto - Implement entity limit for sweeping attacks | Start at one due to original target
                                 for (LivingEntity livingEntity2 : this.level()
                                     .getEntitiesOfClass(LivingEntity.class, target.getBoundingBox().inflate(1.0, 0.25, 1.0))) {
+                                    if (this.level().plutoConfig().sweepingAttackLimit != -1 && entityCount++ >= this.level().plutoConfig().sweepingAttackLimit) continue; // Pluto - Implement entity limit for sweeping attacks
                                     if (livingEntity2 != this
                                         && livingEntity2 != target
                                         && !this.isAlliedTo(livingEntity2)
@@ -1413,7 +_,7 @@
             if (this.isSwimming()) {
                 double d = this.getLookAngle().y;
                 double d1 = d < -0.2 ? 0.085 : 0.06;
-                if (d <= 0.0 || this.jumping || !this.level().getFluidState(BlockPos.containing(this.getX(), this.getY() + 1.0 - 0.1, this.getZ())).isEmpty()) {
+                if (d <= 0.0 || this.jumping || !this.level().pluto$getFluidState(this.getX(), this.getY() + 1.0 - 0.1, this.getZ()).isEmpty()) { // Pluto - Memory allocation optimisations
                     Vec3 deltaMovement = this.getDeltaMovement();
                     this.setDeltaMovement(deltaMovement.add(0.0, (d - deltaMovement.y) * d1, 0.0));
                 }
