--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -492,6 +_,7 @@
         this.updateSwingTime();
         this.yHeadRot = this.getYRot();
         this.setSpeed((float)this.getAttributeValue(Attributes.MOVEMENT_SPEED));
+        if (this.level().plutoConfig().playersDoNearbyEntityLookups) // Pluto - Toggle players doing nearby entity lookups
         if (this.getHealth() > 0.0F && !this.isSpectator()) {
             AABB pickupArea;
             if (this.isPassenger() && !this.getVehicle().isRemoved()) {
@@ -1249,7 +_,9 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             float var12 = 1.0F + (float)this.getAttributeValue(Attributes.SWEEPING_DAMAGE_RATIO) * baseDamage;
 
+            int entityCount = 1; // Pluto - Implement entity limit for sweeping attacks | Start at one due to original target
             for (LivingEntity nearby : this.level().getEntitiesOfClass(LivingEntity.class, entity.getBoundingBox().inflate(1.0, 0.25, 1.0))) {
+                if (this.level().plutoConfig().sweepingAttackLimit != -1 && entityCount++ >= this.level().plutoConfig().sweepingAttackLimit) continue; // Pluto - Implement entity limit for sweeping attacks
                 if (nearby != this
                     && nearby != entity
                     && !this.isAlliedTo(nearby)
@@ -1503,7 +_,7 @@
                 double multiplier = lookAngleY < -0.2 ? 0.085 : 0.06;
                 if (lookAngleY <= 0.0
                     || this.jumping
-                    || !this.level().getFluidState(BlockPos.containing(this.getX(), this.getY() + 1.0 - 0.1, this.getZ())).isEmpty()) {
+                    || !this.level().pluto$getFluidState(this.getX(), this.getY() + 1.0 - 0.1, this.getZ()).isEmpty()) { // Pluto - Memory allocation optimisations
                     Vec3 movement = this.getDeltaMovement();
                     this.setDeltaMovement(movement.add(0.0, (lookAngleY - movement.y) * multiplier, 0.0));
                 }
