--- a/pufferfish-server/build.gradle.kts
+++ b/pufferfish-server/build.gradle.kts
@@ -23,6 +_,7 @@
 
     // Pufferfish start
     val pufferfish = forks.register("pufferfish") {
+        rootDirectory = upstreamsDirectory().map { it.dir("pufferfish") } // Pluto - Branding
         upstream.patchDir("paperServer") {
             upstreamPath = "paper-server"
             excludes = setOf("src/minecraft", "patches", "build.gradle.kts")
@@ -31,13 +_,30 @@
         }
     }
 
-    activeFork = pufferfish
-
     paper {
         paperServerDir = upstreamsDirectory().map { it.dir("paper/paper-server") }
     }
     // Pufferfish end
 
+    // Pluto start - Branding
+    val pluto = forks.register("pluto") {
+        forks = pufferfish
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = pufferfish.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("pluto-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchDir("pufferfishServer") {
+            upstreamPath = "pufferfish-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("pluto-server/pufferfish-patches")
+            outputDir = rootDirectory.dir("pufferfish-server")
+        }
+    }
+
+    activeFork = pluto
+    // Pluto end - Branding
+
     spigot {
         buildDataRef = "3edaf46ec1eed4115ce1b18d2846cded42577e42"
         packageVersion = "v1_21_R3" // also needs to be updated in MappingEnvironment
@@ -123,10 +_,18 @@
     main {
         java { srcDir("../paper-server/src/main/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        // Pluto start - Branding
+        java { srcDir("../pufferfish-server/src/main/java") }
+        resources { srcDir("../pufferfish-server/src/main/resources") }
+        // Pluto end - Branding
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        // Pluto start - Branding
+        java { srcDir("../pufferfish-server/src/test/java") }
+        resources { srcDir("../pufferfish-server/src/test/resources") }
+        // Pluto end - Branding
     }
 }
 
@@ -151,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":pufferfish-api")) // Pufferfish
+    implementation(project(":pluto-api")) // Pufferfish // Pluto - Branding
     implementation("ca.spottedleaf:concurrentutil:0.0.3")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -234,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Pufferfish", // Pufferfish
+            "Implementation-Title" to "Pluto", // Pufferfish // Pluto - Branding
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Pufferfish", // Pufferfish
+            "Specification-Title" to "Pluto", // Pufferfish // Pluto - Branding
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Pufferfish Studios LLC", // Pufferfish
-            "Brand-Id" to "pufferfish:pufferfish", // Pufferfish
-            "Brand-Name" to "Pufferfish", // Pufferfish
+            "Specification-Vendor" to "Yive", // Pufferfish // Pluto - Branding
+            "Brand-Id" to "yive:pluto", // Pufferfish // Pluto - Branding
+            "Brand-Name" to "Yive", // Pufferfish // Pluto - Branding
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
