--- a/alternate/current/wire/WireHandler.java
+++ b/alternate/current/wire/WireHandler.java
@@ -489,7 +_,7 @@
             }
         }
 
-        updateOrder = UpdateOrder.values()[level.paperConfig().misc.alternateCurrentUpdateOrder.ordinal()];
+        updateOrder = UpdateOrder.VALUES[level.paperConfig().misc.alternateCurrentUpdateOrder.ordinal()]; // Pluto - Avoid excessive calls to Enum#values
     }
 
     /**
