--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -99,7 +_,7 @@
         double flowZ = 0.0;
         BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             blockPos.setWithOffset(pos, direction);
             FluidState neighbourFluid = level.getFluidState(blockPos);
             if (this.affectsFlow(neighbourFluid)) {
@@ -107,8 +_,7 @@
                 float distance = 0.0F;
                 if (neighborHeight == 0.0F) {
                     if (!level.getBlockState(blockPos).blocksMotion()) {
-                        BlockPos neighborPos = blockPos.below();
-                        FluidState belowNeighborState = level.getFluidState(neighborPos);
+                        FluidState belowNeighborState = level.pluto$getFluidState(blockPos.getX(), blockPos.getY() - 1, blockPos.getZ()); // Pluto - Memory allocation optimisations
                         if (this.affectsFlow(belowNeighborState)) {
                             neighborHeight = belowNeighborState.getOwnHeight();
                             if (neighborHeight > 0.0F) {
@@ -129,7 +_,7 @@
 
         Vec3 flow = new Vec3(flowX, 0.0, flowZ);
         if (fluidState.getValue(FALLING)) {
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 blockPos.setWithOffset(pos, direction);
                 if (this.isSolidFace(level, blockPos, direction) || this.isSolidFace(level, blockPos.above(), direction)) {
                     flow = flow.normalize().add(0.0, -6.0, 0.0);
@@ -147,7 +_,7 @@
 
     protected boolean isSolidFace(final BlockGetter level, final BlockPos pos, final Direction direction) {
         BlockState state = level.getBlockState(pos);
-        FluidState fluidState = level.getFluidState(pos);
+        FluidState fluidState = state.getFluidState(); // Pluto - Decrease chunk/block lookups
         return !fluidState.getType().isSame(this)
             && (direction == Direction.UP || !(state.getBlock() instanceof IceBlock) && state.isFaceSturdy(level, pos, direction));
     }
@@ -220,7 +_,7 @@
         int neighbourSources = 0;
         BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos relativePos = mutablePos.setWithOffset(pos, direction);
             BlockState blockState = level.getBlockStateIfLoaded(relativePos); // Paper - Prevent chunk loading from fluid flowing
             if (blockState == null) continue; // Paper - Prevent chunk loading from fluid flowing
@@ -342,7 +_,7 @@
     ) {
         int lowest = 1000;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (direction != from) {
                 BlockPos testPos = pos.relative(direction);
                 BlockState testState = context.getBlockStateIfLoaded(testPos); // Paper - Prevent chunk loading from fluid flowing
@@ -410,9 +_,13 @@
     private int sourceNeighborCount(final LevelReader level, final BlockPos pos) {
         int count = 0;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
-            BlockPos testPos = pos.relative(direction);
-            FluidState testFluidState = level.getFluidState(testPos);
+        // Pluto start - Memory allocation optimisations
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
+            FluidState testFluidState = level.pluto$getFluidState(x + direction.getStepX(), y + direction.getStepY(), z + direction.getStepZ());
+            // Pluto end - Memory allocation optimisations
             if (this.isSourceBlockOfThisType(testFluidState)) {
                 count++;
             }
@@ -426,7 +_,7 @@
         Map<Direction, FluidState> result = Maps.newEnumMap(Direction.class);
         FlowingFluid.SpreadContext context = null;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos testPos = pos.relative(direction);
             BlockState testState = level.getBlockStateIfLoaded(testPos); // Paper - Prevent chunk loading from fluid flowing
             if (testState == null) continue; // Paper - Prevent chunk loading from fluid flowing
@@ -531,7 +_,7 @@
     }
 
     private static boolean hasSameAbove(final FluidState fluidState, final BlockGetter level, final BlockPos pos) {
-        return fluidState.getType().isSame(level.getFluidState(pos.above()).getType());
+        return fluidState.getType().isSame(level.pluto$getFluidState(pos.getX(), pos.getY() + 1, pos.getZ()).getType()); // Pluto - Memory allocation optimisations
     }
 
     @Override
