--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -106,8 +_,7 @@
                 float f = 0.0F;
                 if (ownHeight == 0.0F) {
                     if (!blockReader.getBlockState(mutableBlockPos).blocksMotion()) {
-                        BlockPos blockPos = mutableBlockPos.below();
-                        FluidState fluidState2 = blockReader.getFluidState(blockPos);
+                        FluidState fluidState2 = blockReader.pluto$getFluidState(mutableBlockPos.asLong()); // Pluto - Memory allocation optimisations
                         if (this.affectsFlow(fluidState2)) {
                             ownHeight = fluidState2.getOwnHeight();
                             if (ownHeight > 0.0F) {
@@ -146,7 +_,7 @@
 
     protected boolean isSolidFace(BlockGetter level, BlockPos neighborPos, Direction side) {
         BlockState blockState = level.getBlockState(neighborPos);
-        FluidState fluidState = level.getFluidState(neighborPos);
+        FluidState fluidState = blockState.getFluidState(); // Pluto - Decrease chunk/block lookups
         return !fluidState.getType().isSame(this)
             && (side == Direction.UP || !(blockState.getBlock() instanceof IceBlock) && blockState.isFaceSturdy(level, neighborPos, side));
     }
@@ -513,7 +_,7 @@
     }
 
     private static boolean hasSameAbove(FluidState fluidState, BlockGetter level, BlockPos pos) {
-        return fluidState.getType().isSame(level.getFluidState(pos.above()).getType());
+        return fluidState.getType().isSame(level.pluto$getFluidState(pos.getX(), pos.getY() + 1, pos.getZ()).getType()); // Pluto - Memory allocation optimisations
     }
 
     @Override
