--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -167,6 +_,8 @@
         return this.owner.builtInRegistryHolder().tags();
     }
 
+    public java.util.Collection<TagKey<Fluid>> pluto$tags() { return this.owner.builtInRegistryHolder().pluto$tags(); } // Pluto - Streamless tags
+
     public void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
         this.getType().entityInside(level, pos, entity, effectApplier);
     }
