--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -150,6 +_,8 @@
         return this.getType().getAABB(this, level, pos);
     }
 
+    public java.util.Collection<net.minecraft.tags.TagKey<Fluid>> pluto$tags() { return this.owner.builtInRegistryHolder().pluto$tags(); } // Pluto - Streamless tags
+
     public void entityInside(final Level level, final BlockPos pos, final Entity entity, final InsideBlockEffectApplier effectApplier) {
         this.getType().entityInside(level, pos, entity, effectApplier);
     }
