--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -141,7 +_,7 @@
     }
 
     private boolean hasFlammableNeighbours(final LevelReader level, final BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Pluto - Avoid excessive calls to Enum#values
             if (this.isFlammable(level, pos.relative(direction))) {
                 return true;
             }
@@ -221,9 +_,8 @@
     @Override
     protected void spreadTo(final LevelAccessor level, final BlockPos pos, final BlockState state, final Direction direction, final FluidState target) {
         if (direction == Direction.DOWN) {
-            FluidState fluidState = level.getFluidState(pos);
-            if (this.is(FluidTags.LAVA) && fluidState.is(FluidTags.WATER)) {
-                if (state.getBlock() instanceof LiquidBlock) {
+            if (this.is(FluidTags.LAVA) && level.getFluidState(pos).is(FluidTags.WATER)) { // Pluto - Decrease chunk/block lookups
+                if (state.getBlock() instanceof LiquidBlock && !level.getMinecraftWorld().plutoConfig().disableBlockGenerationFromFluids) { // Pluto - Toggle for block generation from fluids
                     // CraftBukkit start
                     if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level.getMinecraftWorld(), pos, Blocks.STONE.defaultBlockState(), Block.UPDATE_ALL)) {
                         return;
