--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -325,6 +_,7 @@
 
     private static void syncAfterConfigurationChange(ChannelFuture future) {
         try {
+            if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) net.minecraft.server.MinecraftServer.getServer().managedBlock(future::isDone); // Pluto - Finish anything blocking before syncing configuration change
             future.syncUninterruptibly();
         } catch (Exception var2) {
             if (var2 instanceof ClosedChannelException) {
