--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -167,7 +_,7 @@
         Vec3 targetPos = DefaultRandomPos.getPos(body, maxHorizontalDistance, maxVerticalDistance);
         int count = 0;
 
-        while (targetPos != null && !body.level().getBlockState(BlockPos.containing(targetPos)).isPathfindable(PathComputationType.WATER) && count++ < 10) {
+        while (targetPos != null && !body.level().pluto$getBlockState(targetPos.asBlockPosLong()).isPathfindable(PathComputationType.WATER) && count++ < 10) { // Pluto - Memory allocation optimisations
             targetPos = DefaultRandomPos.getPos(body, maxHorizontalDistance, maxVerticalDistance);
         }
 
