--- a/net/minecraft/world/entity/ai/behavior/RamTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/RamTarget.java
@@ -121,9 +_,15 @@
 
     private boolean hasRammedHornBreakingBlock(final ServerLevel level, final Goat body) {
         Vec3 horizontalMovementNormalized = body.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize();
-        BlockPos facingBlockPosition = BlockPos.containing(body.position().add(horizontalMovementNormalized));
-        return level.getBlockState(facingBlockPosition).is(BlockTags.SNAPS_GOAT_HORN)
-            || level.getBlockState(facingBlockPosition.above()).is(BlockTags.SNAPS_GOAT_HORN);
+        // Pluto start - Decrease chunk/block lookups
+        final int x = Mth.floor(body.position().x + horizontalMovementNormalized.x);
+        final int z = Mth.floor(body.position().z + horizontalMovementNormalized.z);
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(x >> 4, z >> 4);
+        if (chunk == null) return false;
+
+        final int y = Mth.floor(body.position().z + horizontalMovementNormalized.z);
+        return chunk.getBlockState(x, y, z).is(BlockTags.SNAPS_GOAT_HORN) || chunk.getBlockState(x, y + 1, z).is(BlockTags.SNAPS_GOAT_HORN);
+        // Pluto end - Decrease chunk/block lookups
     }
 
     protected void finishRam(final ServerLevel level, final Goat body) {
