--- a/net/minecraft/world/entity/ai/behavior/RamTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/RamTarget.java
@@ -116,8 +_,15 @@
 
     private boolean hasRammedHornBreakingBlock(ServerLevel level, Goat owner) {
         Vec3 vec3 = owner.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize();
-        BlockPos blockPos = BlockPos.containing(owner.position().add(vec3));
-        return level.getBlockState(blockPos).is(BlockTags.SNAPS_GOAT_HORN) || level.getBlockState(blockPos.above()).is(BlockTags.SNAPS_GOAT_HORN);
+        // Pluto start - Decrease chunk/block lookups
+        final int x = Mth.floor(owner.position().x + vec3.x);
+        final int z = Mth.floor(owner.position().z + vec3.z);
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(x >> 4, z >> 4);
+        if (chunk == null) return false;
+
+        final int y = Mth.floor(owner.position().z + vec3.z);
+        return chunk.getBlockState(x, y, z).is(BlockTags.SNAPS_GOAT_HORN) || chunk.getBlockState(x, y + 1, z).is(BlockTags.SNAPS_GOAT_HORN);
+        // Pluto end - Decrease chunk/block lookups
     }
 
     protected void finishRam(ServerLevel level, Goat owner) {
