--- a/net/minecraft/world/entity/ai/behavior/TryFindLand.java
+++ b/net/minecraft/world/entity/ai/behavior/TryFindLand.java
@@ -37,7 +_,7 @@
                                     BlockState state = level.getBlockState(pos);
                                     BlockState belowState = level.getBlockState(belowPos.setWithOffset(pos, Direction.DOWN));
                                     if (!state.is(Blocks.WATER)
-                                        && level.getFluidState(pos).isEmpty()
+                                        && state.getFluidState().isEmpty() // Pluto - Decrease chunk/block lookups
                                         && state.getCollisionShape(level, pos, context).isEmpty()
                                         && belowState.isFaceSturdy(level, belowPos, Direction.UP)) {
                                         BlockPos targetPos = pos.immutable();
